<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Predictive Analysis of Massive SNP Arrays • bigsnpr</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="jquery.sticky-kit.min.js"></script><script src="pkgdown.js"></script><!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">bigsnpr</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="reference/index.html">Manual</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/pruning-vs-clumping.html">Why clumping should be preferred over pruning</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://privefl.github.io/about.html">
    <span class="fa fa-user"></span>
     
    About
  </a>
</li>
<li>
  <a href="https://github.com/privefl/bigsnpr">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="contents col-md-9">
    

<div id="bigsnpr" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#bigsnpr" class="anchor"></a>bigsnpr</h1></div>
<p><strong>bignspr</strong> is an R Package for the analysis of massive SNP arrays. It enhances the features of <a href="https://privefl.github.io/bigstatsr">package <strong>bigstatsr</strong></a> for the purpose of analysing genotype data.</p>
<p><a href="https://privefl.github.io/bigsnpr/reference/index.html"><strong>LIST OF FEATURES</strong></a></p>
<div id="this-package-is-in-beta-testing" class="section level2">
<h2 class="hasAnchor">
<a href="#this-package-is-in-beta-testing" class="anchor"></a>This package is in beta testing</h2>
<p>Any bug report is welcomed.</p>
</div>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h2>
<p>For now, you can install this package using</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">devtools::<span class="kw">install_github</span>(<span class="st">"privefl/bigsnpr"</span>)</code></pre></div>
</div>
<div id="input-format" class="section level2">
<h2 class="hasAnchor">
<a href="#input-format" class="anchor"></a>Input format</h2>
<p>For now, this package only read <em>bed</em>/<em>bim</em>/<em>fam</em> files (PLINK preferred format) using <code>snp_readBed</code>. Before reading into this package’s special format, quality control and conversion can be done using PLINK, which can be called directly from R using <code>snp_plinkQC</code> and <code>snp_plinkIBDQC</code>.</p>
<p>I use a class called <code>bigSNP</code> for representing infos on massive SNP arrays. One <code>bigSNP</code> has at least 3 elements:</p>
<ul>
<li>
<code>genotypes</code>: A <code>BM.code.descriptor</code> (see <a href="https://privefl.github.io/bigstatsr/#input-format">package <strong>bigstatsr</strong></a>) which describes a special <code>big.matrix</code> encoded with type <code>raw</code> (one byte unsigned integer), representing genotype calls and possibly imputed allele dosages. Rows are samples and columns are SNPs.</li>
<li>
<code>fam</code>: A <code>data.frame</code> containing some information on the SNPs (read from the “.fam” file).</li>
<li>
<code>map</code>: A <code>data.frame</code> giving some information on the individuals (read from a “.bim” file).</li>
</ul>
</div>
<div id="possible-upcoming-features" class="section level2">
<h2 class="hasAnchor">
<a href="#possible-upcoming-features" class="anchor"></a>Possible upcoming features</h2>
<ul>
<li>Support for other input formats. Note that there is room for coding <strong>allele dosages</strong> (rounded to two decimal places). See <code>bigsnpr:::CODE_DOSAGE</code>.</li>
<li>Fast imputation algorithm which doesn’t require reference panels.</li>
<li>Imputation probabilities and multiple imputation.</li>
<li>An interactive QC procedure (call rates, difference of missingness between cases and controls, MAF cutoff, relatedness, HWE, autosomal only, others?).</li>
<li>proper integration of haploid species.</li>
</ul>
</div>
<div id="code-of-conduct" class="section level2">
<h2 class="hasAnchor">
<a href="#code-of-conduct" class="anchor"></a>Code of conduct</h2>
<p>Please note that this project is released with a <a href="https://github.com/privefl/bigsnpr/blob/master/code_of_conduct.md">Contributor Code of Conduct</a>. By participating in this project you agree to abide by its terms.</p>
</div>
</div>

  </div>

  <div class="col-md-3" id="sidebar">
    <h2>Links</h2>
<ul class="list-unstyled">
<li>Browse source code at <br><a href="https://privefl.github.io/bigsnpr">https://​privefl.github.io/​bigsnpr</a>
</li>
<li>Report a bug at <br><a href="https://github.com/privefl/bigsnpr/issues">https://​github.com/​privefl/​bigsnpr/​issues</a>
</li>
</ul>
<h2>License</h2>
<p><a href="https://www.r-project.org/Licenses/GPL-3">GPL-3</a> + file <a href="LICENSE">LICENSE</a></p>
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Florian Privé <br><small class="roles"> Author, maintainer </small> </li>
<li><a href="authors.html">All authors...</a></li>
</ul>
<h2>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://travis-ci.org/privefl/bigsnpr"><img src="https://travis-ci.org/privefl/bigsnpr.svg?branch=master" alt="Travis-CI Build Status"></a></li>
<li><a href="https://codecov.io/github/privefl/bigsnpr?branch=master"><img src="https://img.shields.io/codecov/c/github/privefl/bigsnpr/master.svg" alt="Coverage Status"></a></li>
<li><a href="http://cran.r-project.org/package=bigsnpr"><img src="http://www.r-pkg.org/badges/version/bigsnpr" alt="CRAN_Status_Badge"></a></li>
</ul>
</div>
</div>


      <footer><div class="copyright">
  <p>Developed by Florian Privé.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  </body>
</html>
