<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Demonstration of packages bigsnpr and bigstatsr • bigsnpr</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Demonstration of packages bigsnpr and bigstatsr">
<meta property="og:description" content="bigsnpr">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">bigsnpr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.4.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Manual</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/demo.html">Quick demo</a>
    </li>
    <li>
      <a href="../articles/exo.html">Exercise</a>
    </li>
    <li>
      <a href="../articles/pruning-vs-clumping.html">Why clumping should be preferred over pruning</a>
    </li>
    <li>
      <a href="../articles/how-to-PCA.html">How to capture Population Structure with PCA (LD problem explained)</a>
    </li>
    <li>
      <a href="../articles/bedpca.html">How to capture Population Structure with PCA (directly on PLINK bed files)</a>
    </li>
    <li>
      <a href="../articles/SCT.html">Computing polygenic scores using Stacked Clumping and Thresholding (SCT)</a>
    </li>
    <li>
      <a href="../articles/LDpred2.html">Computing polygenic scores using LDpred2</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://privefl.github.io/about.html">
    <span class="fa fa-user"></span>
     
    About
  </a>
</li>
<li>
  <a href="https://github.com/privefl/bigsnpr">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Demonstration of packages bigsnpr and bigstatsr</h1>
                        <h4 class="author">Florian Privé</h4>
            
            <h4 class="date">June 28, 2018</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/privefl/bigsnpr/blob/master/vignettes/demo.Rmd"><code>vignettes/demo.Rmd</code></a></small>
      <div class="hidden name"><code>demo.Rmd</code></div>

    </div>

    
    
<div id="get-genotype-file" class="section level2">
<h2 class="hasAnchor">
<a href="#get-genotype-file" class="anchor"></a>Get genotype file</h2>
<div class="sourceCode" id="cb1"><html><body><pre class="r"><span class="co"># Get the example bedfile from package bigsnpr</span>
<span class="no">bedfile</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"example.bed"</span>, <span class="kw">package</span> <span class="kw">=</span> <span class="st">"bigsnpr"</span>)</pre></body></html></div>
</div>
<div id="read-from-the-plink-files" class="section level2">
<h2 class="hasAnchor">
<a href="#read-from-the-plink-files" class="anchor"></a>Read from the PLINK files</h2>
<div class="sourceCode" id="cb2"><html><body><pre class="r"><span class="co"># Load packages bigsnpr and bigstatsr</span>
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">bigsnpr</span>)</pre></body></html></div>
<pre><code>## Loading required package: bigstatsr</code></pre>
<div class="sourceCode" id="cb4"><html><body><pre class="r"><span class="co"># Read from bed/bim/fam, it will create new files.</span>
<span class="co"># Let's put them in an temporary directory for this demo.</span>
<span class="no">tmpfile</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html">tempfile</a></span>()
<span class="fu"><a href="../reference/snp_readBed.html">snp_readBed</a></span>(<span class="no">bedfile</span>, <span class="kw">backingfile</span> <span class="kw">=</span> <span class="no">tmpfile</span>)</pre></body></html></div>
<pre><code>## [1] "C:\\Users\\au639593\\AppData\\Local\\Temp\\Rtmp8eEz1e\\file12345dbe2ad6.rds"</code></pre>
<div class="sourceCode" id="cb6"><html><body><pre class="r"><span class="co"># Attach the "bigSNP" object in R session</span>
<span class="no">obj.bigSNP</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/snp_attach.html">snp_attach</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(<span class="no">tmpfile</span>, <span class="st">".rds"</span>))
<span class="co"># See how it looks like</span>
<span class="fu"><a href="https://rdrr.io/r/utils/str.html">str</a></span>(<span class="no">obj.bigSNP</span>, <span class="kw">max.level</span> <span class="kw">=</span> <span class="fl">2</span>, <span class="kw">strict.width</span> <span class="kw">=</span> <span class="st">"cut"</span>)</pre></body></html></div>
<pre><code>## List of 3
##  $ genotypes:Reference class 'FBM.code256' [package "bigstatsr"..
##   ..and 26 methods, of which 12 are  possibly relevant:
##   ..  add_columns, as.FBM, bm, bm.desc, check_dimensions,
##   ..  check_write_permissions, copy#envRefClass, initialize,
##   ..  initialize#FBM, save, show#envRefClass, show#FBM
##  $ fam      :'data.frame':   517 obs. of  6 variables:
##   ..$ family.ID  : chr [1:517] "POP1" "POP1" "POP1" "POP1" ...
##   ..$ sample.ID  : chr [1:517] "IND0" "IND1" "IND2" "IND3" ...
##   ..$ paternal.ID: int [1:517] 0 0 0 0 0 0 0 0 0 0 ...
##   ..$ maternal.ID: int [1:517] 0 0 0 0 0 0 0 0 0 0 ...
##   ..$ sex        : int [1:517] 0 0 0 0 0 0 0 0 0 0 ...
##   ..$ affection  : int [1:517] 1 1 2 1 1 1 1 1 1 1 ...
##  $ map      :'data.frame':   4542 obs. of  6 variables:
##   ..$ chromosome  : int [1:4542] 1 1 1 1 1 1 1 1 1 1 ...
##   ..$ marker.ID   : chr [1:4542] "SNP0" "SNP1" "SNP2" "SNP3" ...
##   ..$ genetic.dist: int [1:4542] 0 0 0 0 0 0 0 0 0 0 ...
##   ..$ physical.pos: int [1:4542] 112 1098 2089 3107 4091 5091 6..
##   ..$ allele1     : chr [1:4542] "A" "T" "T" "T" ...
##   ..$ allele2     : chr [1:4542] "T" "A" "A" "A" ...
##  - attr(*, "class")= chr "bigSNP"</code></pre>
<div class="sourceCode" id="cb8"><html><body><pre class="r"><span class="co"># Get aliases for useful slots</span>
<span class="no">G</span>   <span class="kw">&lt;-</span> <span class="no">obj.bigSNP</span>$<span class="no">genotypes</span>
<span class="no">CHR</span> <span class="kw">&lt;-</span> <span class="no">obj.bigSNP</span>$<span class="no">map</span>$<span class="no">chromosome</span>
<span class="no">POS</span> <span class="kw">&lt;-</span> <span class="no">obj.bigSNP</span>$<span class="no">map</span>$<span class="no">physical.pos</span>
<span class="co"># Check some counts for the 10 first SNPs</span>
<span class="fu">big_counts</span>(<span class="no">G</span>, <span class="kw">ind.col</span> <span class="kw">=</span> <span class="fl">1</span>:<span class="fl">10</span>)</pre></body></html></div>
<pre><code>##      [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8] [,9] [,10]
## 0     226  325  309  346  158  215  395  134  143   159
## 1     228  171  171  151  246  236  112  269  250   235
## 2      63   21   37   20  113   66   10  114  124   123
## &lt;NA&gt;    0    0    0    0    0    0    0    0    0     0</code></pre>
</div>
<div id="principal-component-analysis" class="section level2">
<h2 class="hasAnchor">
<a href="#principal-component-analysis" class="anchor"></a>Principal Component Analysis</h2>
<div class="sourceCode" id="cb10"><html><body><pre class="r"><span class="co"># Half of the cores you have on your computer</span>
<span class="no">NCORES</span> <span class="kw">&lt;-</span> <span class="fu">nb_cores</span>()
<span class="co"># Exclude Long-Range Linkage Disequilibrium Regions of the human genome</span>
<span class="co"># based on an online table. </span>
<span class="no">ind.excl</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/snp_clumping.html">snp_indLRLDR</a></span>(<span class="kw">infos.chr</span> <span class="kw">=</span> <span class="no">CHR</span>, <span class="kw">infos.pos</span> <span class="kw">=</span> <span class="no">POS</span>)
<span class="co"># Use clumping (on the MAF) to keep SNPs weakly correlated with each other.</span>
<span class="co"># See https://privefl.github.io/bigsnpr/articles/pruning-vs-clumping.html</span>
<span class="co"># to know why I prefer using clumping than standard pruning.</span>
<span class="no">ind.keep</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/snp_clumping.html">snp_clumping</a></span>(<span class="no">G</span>, <span class="kw">infos.chr</span> <span class="kw">=</span> <span class="no">CHR</span>,
                         <span class="kw">exclude</span> <span class="kw">=</span> <span class="no">ind.excl</span>,
                         <span class="kw">ncores</span> <span class="kw">=</span> <span class="no">NCORES</span>)
<span class="co"># Get the first 10 PCs, corresponding to pruned SNPs</span>
<span class="no">obj.svd</span> <span class="kw">&lt;-</span> <span class="fu">big_randomSVD</span>(<span class="no">G</span>, <span class="kw">fun.scaling</span> <span class="kw">=</span> <span class="fu"><a href="../reference/snp_scaleBinom.html">snp_scaleBinom</a></span>(),
                         <span class="kw">ind.col</span> <span class="kw">=</span> <span class="no">ind.keep</span>,
                         <span class="kw">ncores</span> <span class="kw">=</span> <span class="no">NCORES</span>)
<span class="co"># As `obj.svd` has a class and a method `plot`.</span>
<span class="co"># Scree plot by default</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.html">plot</a></span>(<span class="no">obj.svd</span>)</pre></body></html></div>
<p><img src="demo_files/figure-html/unnamed-chunk-3-1.png" width="692.708333333333" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb11"><html><body><pre class="r"><span class="co"># Score (PCs) plot</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.html">plot</a></span>(<span class="no">obj.svd</span>, <span class="kw">type</span> <span class="kw">=</span> <span class="st">"scores"</span>)</pre></body></html></div>
<p><img src="demo_files/figure-html/unnamed-chunk-3-2.png" width="692.708333333333" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb12"><html><body><pre class="r"><span class="co"># As plot returns an ggplot2 object, you can easily modify it.</span>
<span class="co"># For example, you can add colors based on the population.</span>
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">ggplot2</span>)</pre></body></html></div>
<pre><code>## Warning: package 'ggplot2' was built under R version 3.6.3</code></pre>
<div class="sourceCode" id="cb14"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/graphics/plot.html">plot</a></span>(<span class="no">obj.svd</span>, <span class="kw">type</span> <span class="kw">=</span> <span class="st">"scores"</span>) +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="kw">color</span> <span class="kw">=</span> <span class="no">pop</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"POP1"</span>, <span class="st">"POP2"</span>, <span class="st">"POP3"</span>), <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">143</span>, <span class="fl">167</span>, <span class="fl">207</span>))) +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span>(<span class="kw">color</span> <span class="kw">=</span> <span class="st">"Population"</span>)</pre></body></html></div>
<p><img src="demo_files/figure-html/unnamed-chunk-3-3.png" width="692.708333333333" style="display: block; margin: auto;"></p>
</div>
<div id="genome-wide-association-study" class="section level2">
<h2 class="hasAnchor">
<a href="#genome-wide-association-study" class="anchor"></a>Genome-Wide Association Study</h2>
<div class="sourceCode" id="cb15"><html><body><pre class="r"><span class="co"># Fit a logistic model between the phenotype and each SNP separately</span>
<span class="co"># while adding PCs as covariates to each model</span>
<span class="no">y01</span> <span class="kw">&lt;-</span> <span class="no">obj.bigSNP</span>$<span class="no">fam</span>$<span class="no">affection</span> - <span class="fl">1</span>
<span class="no">obj.gwas</span> <span class="kw">&lt;-</span> <span class="fu">big_univLogReg</span>(<span class="no">G</span>, <span class="kw">y01.train</span> <span class="kw">=</span> <span class="no">y01</span>,
                           <span class="kw">covar.train</span> <span class="kw">=</span> <span class="no">obj.svd</span>$<span class="no">u</span>,
                           <span class="kw">ncores</span> <span class="kw">=</span> <span class="no">NCORES</span>)
<span class="co"># Q-Q plot of the object</span>
<span class="fu"><a href="../reference/snp_qq.html">snp_qq</a></span>(<span class="no">obj.gwas</span>)</pre></body></html></div>
<p><img src="demo_files/figure-html/unnamed-chunk-4-1.png" width="692.708333333333" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb16"><html><body><pre class="r"><span class="co"># You can easily apply genomic control to this object</span>
<span class="no">obj.gwas.gc</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/snp_gc.html">snp_gc</a></span>(<span class="no">obj.gwas</span>)
<span class="co"># Redo the Q-Q plot</span>
<span class="fu"><a href="../reference/snp_qq.html">snp_qq</a></span>(<span class="no">obj.gwas.gc</span>)</pre></body></html></div>
<p><img src="demo_files/figure-html/unnamed-chunk-4-2.png" width="692.708333333333" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb17"><html><body><pre class="r"><span class="co"># Manhattan plot, not quite sexy because there are only 1 chromosome here</span>
<span class="fu"><a href="../reference/snp_manhattan.html">snp_manhattan</a></span>(<span class="no">obj.gwas.gc</span>, <span class="kw">infos.chr</span> <span class="kw">=</span> <span class="no">CHR</span>, <span class="kw">infos.pos</span> <span class="kw">=</span> <span class="no">POS</span>)</pre></body></html></div>
<p><img src="demo_files/figure-html/unnamed-chunk-4-3.png" width="692.708333333333" style="display: block; margin: auto;"></p>
</div>
<div id="joint-prs" class="section level2">
<h2 class="hasAnchor">
<a href="#joint-prs" class="anchor"></a>Joint PRS</h2>
<div class="sourceCode" id="cb18"><html><body><pre class="r"><span class="co"># Divide the indices in training/test sets</span>
<span class="no">ind.train</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span>(<span class="no">G</span>), <span class="fl">400</span>)
<span class="no">ind.test</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sets.html">setdiff</a></span>(<span class="fu">rows_along</span>(<span class="no">G</span>), <span class="no">ind.train</span>)
<span class="co"># Train the model</span>
<span class="no">cmsa.logit</span> <span class="kw">&lt;-</span> <span class="fu">big_spLogReg</span>(<span class="kw">X</span> <span class="kw">=</span> <span class="no">G</span>, <span class="kw">y01.train</span> <span class="kw">=</span> <span class="no">y01</span>[<span class="no">ind.train</span>],
                           <span class="kw">ind.train</span> <span class="kw">=</span> <span class="no">ind.train</span>,
                           <span class="kw">covar.train</span> <span class="kw">=</span> <span class="no">obj.svd</span>$<span class="no">u</span>[<span class="no">ind.train</span>, ],
                           <span class="kw">alphas</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">1</span>, <span class="fl">0.5</span>, <span class="fl">0.05</span>, <span class="fl">0.001</span>),
                           <span class="kw">ncores</span> <span class="kw">=</span> <span class="no">NCORES</span>)
<span class="co"># Get predictions for the test set</span>
<span class="no">preds</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/predict.html">predict</a></span>(<span class="no">cmsa.logit</span>, <span class="kw">X</span> <span class="kw">=</span> <span class="no">G</span>, <span class="kw">ind.row</span> <span class="kw">=</span> <span class="no">ind.test</span>,
                 <span class="kw">covar.row</span> <span class="kw">=</span> <span class="no">obj.svd</span>$<span class="no">u</span>[<span class="no">ind.test</span>, ])
<span class="co"># Check AUC</span>
<span class="fu">AUC</span>(<span class="no">preds</span>, <span class="no">y01</span>[<span class="no">ind.test</span>])</pre></body></html></div>
<pre><code>## [1] 0.7113997</code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Florian Privé.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
