<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Dosage data: proof of concept • bigsnpr</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script><script src="../pkgdown.js"></script><!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-vignette">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">bigsnpr</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Manual</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/demo.html">Quick demo</a>
    </li>
    <li>
      <a href="../articles/pruning-vs-clumping.html">Why clumping should be preferred over pruning</a>
    </li>
    <li>
      <a href="../artciles/dosage.html">Dosage data: proof of concept</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://privefl.github.io/about.html">
    <span class="fa fa-user"></span>
     
    About
  </a>
</li>
<li>
  <a href="https://github.com/privefl/bigsnpr">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9">
    <div class="page-header toc-ignore">
      <h1>Dosage data: proof of concept</h1>
                        <h4 class="author">Florian Privé</h4>
            
            <h4 class="date">January 3, 2018</h4>
          </div>

    
    
<div class="contents">
<p>In this vignette, I show how to get dosage data from a <a href="http://www.well.ox.ac.uk/~gav/bgen_format/bgen_format_v1.2.html">BGEN</a> file. This solution is just a proof of concept. I plan to better develop this when I get the access to the UK Biobank dataset.</p>
<div id="the-r-package-rbgen" class="section level2">
<h2 class="hasAnchor">
<a href="#the-r-package-rbgen" class="anchor"></a>The R package rbgen</h2>
<p>Gavin Band developed an R package to load data in BGEN format. Let us use it in this vignette. To install this package, follow the instructions <a href="https://bitbucket.org/gavinband/bgen/wiki/rbgen">there</a>. In summary:</p>
<ol style="list-style-type: decimal">
<li>download <a href="http://bitbucket.org/gavinband/bgen/get/master.tar.gz" class="uri">http://bitbucket.org/gavinband/bgen/get/master.tar.gz</a>
</li>
<li>run <code>./waf-1.8.13 configure</code> and then <code>./waf-1.8.13</code> in a shell</li>
<li>verify that it is working by running <code>./build/test/test_bgen</code> and <code>./build/example/bgen_to_vcf example/example.8bits.bgen</code>
</li>
<li>run <code>sudo ./waf-1.8.13 install</code>
</li>
<li>run <code>R CMD INSTALL build/R/rbgen</code>
</li>
</ol>
</div>
<div id="the-fbm-code256-format" class="section level2">
<h2 class="hasAnchor">
<a href="#the-fbm-code256-format" class="anchor"></a>The FBM.code256 format</h2>
<p>As the name says, FBM are filebackeg big matrices which store elements on 8 bits and use a code of 256 values in order to map each possible byte with a corresponding access code. For example, I defined the following code:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">bigsnpr<span class="op">:::</span>CODE_DOSAGE</code></pre></div>
<pre><code>##   [1] 0.00 1.00 2.00   NA 0.00 1.00 2.00 0.00 0.01 0.02 0.03 0.04
##  [13] 0.05 0.06 0.07 0.08 0.09 0.10 0.11 0.12 0.13 0.14 0.15 0.16
##  [25] 0.17 0.18 0.19 0.20 0.21 0.22 0.23 0.24 0.25 0.26 0.27 0.28
##  [37] 0.29 0.30 0.31 0.32 0.33 0.34 0.35 0.36 0.37 0.38 0.39 0.40
##  [49] 0.41 0.42 0.43 0.44 0.45 0.46 0.47 0.48 0.49 0.50 0.51 0.52
##  [61] 0.53 0.54 0.55 0.56 0.57 0.58 0.59 0.60 0.61 0.62 0.63 0.64
##  [73] 0.65 0.66 0.67 0.68 0.69 0.70 0.71 0.72 0.73 0.74 0.75 0.76
##  [85] 0.77 0.78 0.79 0.80 0.81 0.82 0.83 0.84 0.85 0.86 0.87 0.88
##  [97] 0.89 0.90 0.91 0.92 0.93 0.94 0.95 0.96 0.97 0.98 0.99 1.00
## [109] 1.01 1.02 1.03 1.04 1.05 1.06 1.07 1.08 1.09 1.10 1.11 1.12
## [121] 1.13 1.14 1.15 1.16 1.17 1.18 1.19 1.20 1.21 1.22 1.23 1.24
## [133] 1.25 1.26 1.27 1.28 1.29 1.30 1.31 1.32 1.33 1.34 1.35 1.36
## [145] 1.37 1.38 1.39 1.40 1.41 1.42 1.43 1.44 1.45 1.46 1.47 1.48
## [157] 1.49 1.50 1.51 1.52 1.53 1.54 1.55 1.56 1.57 1.58 1.59 1.60
## [169] 1.61 1.62 1.63 1.64 1.65 1.66 1.67 1.68 1.69 1.70 1.71 1.72
## [181] 1.73 1.74 1.75 1.76 1.77 1.78 1.79 1.80 1.81 1.82 1.83 1.84
## [193] 1.85 1.86 1.87 1.88 1.89 1.90 1.91 1.92 1.93 1.94 1.95 1.96
## [205] 1.97 1.98 1.99 2.00   NA   NA   NA   NA   NA   NA   NA   NA
## [217]   NA   NA   NA   NA   NA   NA   NA   NA   NA   NA   NA   NA
## [229]   NA   NA   NA   NA   NA   NA   NA   NA   NA   NA   NA   NA
## [241]   NA   NA   NA   NA   NA   NA   NA   NA   NA   NA   NA   NA
## [253]   NA   NA   NA   NA</code></pre>
<p>The use the first 4 elements to code for a genotype call (<code>0</code>, <code>1</code>, <code>2</code> or missing), the 3 following ones to store an imputed genotype call (you may wonder what is the purpose of this but I can assure you that this can be cleverly used in some algorithms). Then the 201 following values codes for dosages, rounded to two decimal places, which is similar to the 8-bit encoding of the BGEN format used for the UK Biobank. The 209-th to 256-th values are currently not used.</p>
<p>A similar data structure called FBM.code65636 (2 bytes) may be implemented to store dosages with a precision of at least <code>0.00005</code>. Yet, we think such a precision is not useful and a precision of <code>0.01</code> is sufficient for dosage data.</p>
</div>
<div id="converting-to-a-fbm-code256" class="section level2">
<h2 class="hasAnchor">
<a href="#converting-to-a-fbm-code256" class="anchor"></a>Converting to a FBM.code256</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">bgen2FBM &lt;-<span class="st"> </span><span class="cf">function</span>(bgen_list) {
  
  DIM &lt;-<span class="st"> </span><span class="kw">dim</span>(bgen_list<span class="op">$</span>data)
  
  <span class="co"># constructing a temporary genotype big.matrix</span>
  bigGeno &lt;-<span class="st"> </span><span class="kw">FBM.code256</span>(DIM[<span class="dv">2</span>], DIM[<span class="dv">1</span>], <span class="dt">code =</span> bigsnpr<span class="op">:::</span>CODE_DOSAGE)
  
  <span class="co"># compute dosages from probabilities</span>
  <span class="kw">dim</span>(bgen_list<span class="op">$</span>data) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="kw">prod</span>(DIM[<span class="dv">1</span><span class="op">:</span><span class="dv">2</span>]), DIM[<span class="dv">3</span>])
  dosages &lt;-<span class="st"> </span>bgen_list<span class="op">$</span>data <span class="op">%*%</span><span class="st"> </span><span class="dv">0</span><span class="op">:</span><span class="dv">2</span>
  <span class="co"># transform them in raw codes</span>
  nonas &lt;-<span class="st"> </span><span class="op">!</span><span class="kw">is.na</span>(dosages)
  dosages_raw &lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="kw">as.raw</span>(<span class="dv">3</span>), <span class="dt">nrow =</span> DIM[<span class="dv">1</span>], <span class="dt">ncol =</span> DIM[<span class="dv">2</span>])
  dosages_raw[nonas] &lt;-<span class="st"> </span><span class="kw">as.raw</span>(<span class="kw">round</span>(<span class="dv">100</span> <span class="op">*</span><span class="st"> </span>dosages[nonas]) <span class="op">+</span><span class="st"> </span><span class="dv">7</span>)
  <span class="co"># put these dosages in the FBM</span>
  bigGeno[] &lt;-<span class="st"> </span><span class="kw">t</span>(dosages_raw)
  
  <span class="kw">list</span>(
    <span class="dt">genotypes =</span> bigGeno,
    <span class="dt">variants =</span> bgen_list<span class="op">$</span>variants,
    <span class="dt">samples =</span> bgen_list<span class="op">$</span>samples
  )
}</code></pre></div>
<p>Use the <strong>rbgen</strong> package to read BGEN data into R:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ranges =<span class="st"> </span><span class="kw">data.frame</span>(
  <span class="dt">chromosome =</span> <span class="st">"01"</span>,
  <span class="dt">start =</span> <span class="dv">1</span>,
  <span class="dt">end =</span> <span class="fl">2e9</span>
)
bgen_file &lt;-<span class="st"> </span><span class="kw">path.expand</span>(<span class="st">"~/Téléchargements/gavinband-bgen-798eca81c0fa/example/example.8bits.bgen"</span>)
<span class="cf">if</span> (<span class="op">!</span><span class="kw">file.exists</span>(<span class="kw">paste0</span>(bgen_file, <span class="st">".bgi"</span>)))
  <span class="kw">system</span>(glue<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/glue/topics/glue">glue</a></span>(<span class="st">"bgenix -g {bgen_file} -index"</span>))
bgen_list &lt;-<span class="st"> </span>rbgen<span class="op">::</span><span class="kw">bgen.load</span>(bgen_file, ranges)</code></pre></div>
<p>Let’s see the dosages we get for some given probabilities:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(bigstatsr)
test &lt;-<span class="st"> </span><span class="kw">bgen2FBM</span>(bgen_list)
<span class="kw">t</span>(bgen_list<span class="op">$</span>data[<span class="dv">1</span><span class="op">:</span><span class="dv">12</span>, <span class="dv">1</span>, ])</code></pre></div>
<pre><code>##        RSID_101 RSID_2    RSID_102      RSID_3    RSID_103
## g=0 0.000000000     NA 0.917647059 0.003921569 0.992156863
## g=1 0.003921569     NA 0.007843137 0.003921569 0.003921569
## g=2 0.996078431     NA 0.074509804 0.992156863 0.003921569
##          RSID_4    RSID_104      RSID_5    RSID_105      RSID_6
## g=0 0.992156863 0.007843137 0.003921569 0.968627451 0.003921569
## g=1 0.000000000 0.000000000 0.027450980 0.027450980 0.996078431
## g=2 0.007843137 0.992156863 0.968627451 0.003921569 0.000000000
##        RSID_106 RSID_7
## g=0 0.000000000     NA
## g=1 0.996078431     NA
## g=2 0.003921569     NA</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">test<span class="op">$</span>genotypes[<span class="dv">1</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">12</span>]</code></pre></div>
<pre><code>##  [1] 2.00   NA 0.16 1.99 0.01 0.02 1.98 1.96 0.04 1.00 1.00   NA</code></pre>
<p>BGEN files store twice information as FBM files (probabilities vs dosages) but are compressed so that FBMs are just slightly smaller files.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">file.size</span>(test<span class="op">$</span>genotypes<span class="op">$</span>backingfile)</code></pre></div>
<pre><code>## [1] 99500</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">file.size</span>(bgen_file)</code></pre></div>
<pre><code>## [1] 128746</code></pre>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#the-r-package-rbgen">The R package rbgen</a></li>
      <li><a href="#the-fbm-code256-format">The FBM.code256 format</a></li>
      <li><a href="#converting-to-a-fbm-code256">Converting to a FBM.code256</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Florian Privé.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  </body>
</html>
