<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Dosage data: proof of concept • bigsnpr</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script><script src="../pkgdown.js"></script><!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-vignette">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">bigsnpr</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Manual</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/demo.html">Quick demo</a>
    </li>
    <li>
      <a href="../articles/how-to-PCA.html">How to capture Population Structure with PCA</a>
    </li>
    <li>
      <a href="../articles/pruning-vs-clumping.html">Why clumping should be preferred over pruning</a>
    </li>
    <li>
      <a href="../articles/dosage.html">Dosage data: proof of concept</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://privefl.github.io/about.html">
    <span class="fa fa-user"></span>
     
    About
  </a>
</li>
<li>
  <a href="https://github.com/privefl/bigsnpr">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9">
    <div class="page-header toc-ignore">
      <h1>Dosage data: proof of concept</h1>
                        <h4 class="author">Florian Privé</h4>
            
            <h4 class="date">January 3, 2018</h4>
          </div>

    
    
<div class="contents">
<p>In this vignette, I show how to get dosage data from a <a href="http://www.well.ox.ac.uk/~gav/bgen_format/bgen_format_v1.2.html">BGEN</a> file. I provide two different solutions, as a proof of concept. Even if the second solution might be used for very large datasets, I plan to better develop the reading of BGEN files when I get the access to the full UK Biobank dataset. Indeed, it would be better to directly read probabilies from the BGEN files, without intermediaries.</p>
<div id="the-fbm-code256-format" class="section level2">
<h2 class="hasAnchor">
<a href="#the-fbm-code256-format" class="anchor"></a>The FBM.code256 format</h2>
<p>This is the format used in package <strong>bigsnpr</strong>. As the name says, FBM are filebackeg big matrices which store elements on 8 bits and use a code of 256 values in order to map each possible byte with a corresponding access code. For example, I defined the following code:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">bigsnpr<span class="op">:::</span>CODE_DOSAGE</code></pre></div>
<pre><code>##   [1] 0.00 1.00 2.00   NA 0.00 1.00 2.00 0.00 0.01 0.02 0.03 0.04
##  [13] 0.05 0.06 0.07 0.08 0.09 0.10 0.11 0.12 0.13 0.14 0.15 0.16
##  [25] 0.17 0.18 0.19 0.20 0.21 0.22 0.23 0.24 0.25 0.26 0.27 0.28
##  [37] 0.29 0.30 0.31 0.32 0.33 0.34 0.35 0.36 0.37 0.38 0.39 0.40
##  [49] 0.41 0.42 0.43 0.44 0.45 0.46 0.47 0.48 0.49 0.50 0.51 0.52
##  [61] 0.53 0.54 0.55 0.56 0.57 0.58 0.59 0.60 0.61 0.62 0.63 0.64
##  [73] 0.65 0.66 0.67 0.68 0.69 0.70 0.71 0.72 0.73 0.74 0.75 0.76
##  [85] 0.77 0.78 0.79 0.80 0.81 0.82 0.83 0.84 0.85 0.86 0.87 0.88
##  [97] 0.89 0.90 0.91 0.92 0.93 0.94 0.95 0.96 0.97 0.98 0.99 1.00
## [109] 1.01 1.02 1.03 1.04 1.05 1.06 1.07 1.08 1.09 1.10 1.11 1.12
## [121] 1.13 1.14 1.15 1.16 1.17 1.18 1.19 1.20 1.21 1.22 1.23 1.24
## [133] 1.25 1.26 1.27 1.28 1.29 1.30 1.31 1.32 1.33 1.34 1.35 1.36
## [145] 1.37 1.38 1.39 1.40 1.41 1.42 1.43 1.44 1.45 1.46 1.47 1.48
## [157] 1.49 1.50 1.51 1.52 1.53 1.54 1.55 1.56 1.57 1.58 1.59 1.60
## [169] 1.61 1.62 1.63 1.64 1.65 1.66 1.67 1.68 1.69 1.70 1.71 1.72
## [181] 1.73 1.74 1.75 1.76 1.77 1.78 1.79 1.80 1.81 1.82 1.83 1.84
## [193] 1.85 1.86 1.87 1.88 1.89 1.90 1.91 1.92 1.93 1.94 1.95 1.96
## [205] 1.97 1.98 1.99 2.00   NA   NA   NA   NA   NA   NA   NA   NA
## [217]   NA   NA   NA   NA   NA   NA   NA   NA   NA   NA   NA   NA
## [229]   NA   NA   NA   NA   NA   NA   NA   NA   NA   NA   NA   NA
## [241]   NA   NA   NA   NA   NA   NA   NA   NA   NA   NA   NA   NA
## [253]   NA   NA   NA   NA</code></pre>
<p>The use the first 4 elements to code for a genotype call (<code>0</code>, <code>1</code>, <code>2</code> or missing), the 3 following ones to store an imputed genotype call (you may wonder what is the purpose of this but I can assure you that this can be cleverly used in some algorithms). Then the 201 following values codes for dosages, rounded to two decimal places, which is similar to the 8-bit encoding of the BGEN format used for the UK Biobank. I will also use the 209-th value to code a missing dosage. The 210-th to 256-th values are currently not used.</p>
<p>A similar data structure called FBM.code65636 (2 bytes) may be implemented to store dosages with a precision of at least <code>0.00005</code>. Yet, we think such a precision is not useful and a precision of <code>0.01</code> is sufficient for dosage data.</p>
</div>
<div id="first-solution-using-the-r-package-rbgen" class="section level2">
<h2 class="hasAnchor">
<a href="#first-solution-using-the-r-package-rbgen" class="anchor"></a>First solution: Using the R package rbgen</h2>
<p>Gavin Band developed an R package to load data in BGEN format. Let us use this R package in the first solution to read dosages from BGEN files in my format <em>bigSNP</em>. To install this R package, follow the instructions <a href="https://bitbucket.org/gavinband/bgen/wiki/rbgen">there</a>. In summary:</p>
<ol style="list-style-type: decimal">
<li>download <a href="http://bitbucket.org/gavinband/bgen/get/master.tar.gz" class="uri">http://bitbucket.org/gavinband/bgen/get/master.tar.gz</a>
</li>
<li>run <code>./waf-1.8.13 configure</code> and then <code>./waf-1.8.13</code> in a shell</li>
<li>verify that it is working by running <code>./build/test/test_bgen</code> and <code>./build/example/bgen_to_vcf example/example.8bits.bgen</code>
</li>
<li>run <code>sudo ./waf-1.8.13 install</code>
</li>
<li>run <code>R CMD INSTALL build/R/rbgen</code>
</li>
</ol>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">bgen2FBM &lt;-<span class="st"> </span><span class="cf">function</span>(bgen_list) {
  
  DIM &lt;-<span class="st"> </span><span class="kw">dim</span>(bgen_list<span class="op">$</span>data)
  
  <span class="co"># constructing a temporary genotype big.matrix</span>
  bigGeno &lt;-<span class="st"> </span><span class="kw">FBM.code256</span>(DIM[<span class="dv">2</span>], DIM[<span class="dv">1</span>], <span class="dt">code =</span> bigsnpr<span class="op">:::</span>CODE_DOSAGE)
  
  <span class="co"># compute dosages from probabilities</span>
  <span class="kw">dim</span>(bgen_list<span class="op">$</span>data) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="kw">prod</span>(DIM[<span class="dv">1</span><span class="op">:</span><span class="dv">2</span>]), DIM[<span class="dv">3</span>])
  dosages &lt;-<span class="st"> </span>bgen_list<span class="op">$</span>data <span class="op">%*%</span><span class="st"> </span><span class="dv">0</span><span class="op">:</span><span class="dv">2</span>
  <span class="co"># transform them in raw codes</span>
  nonas &lt;-<span class="st"> </span><span class="op">!</span><span class="kw">is.na</span>(dosages)
  dosages_raw &lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="kw">as.raw</span>(<span class="dv">3</span>), <span class="dt">nrow =</span> DIM[<span class="dv">1</span>], <span class="dt">ncol =</span> DIM[<span class="dv">2</span>])
  dosages_raw[nonas] &lt;-<span class="st"> </span><span class="kw">as.raw</span>(<span class="kw">round</span>(<span class="dv">100</span> <span class="op">*</span><span class="st"> </span>dosages[nonas]) <span class="op">+</span><span class="st"> </span><span class="dv">7</span>)
  <span class="co"># put these dosages in the FBM</span>
  bigGeno[] &lt;-<span class="st"> </span><span class="kw">t</span>(dosages_raw)
  
  <span class="kw">list</span>(
    <span class="dt">genotypes =</span> bigGeno,
    <span class="dt">variants =</span> bgen_list<span class="op">$</span>variants,
    <span class="dt">samples =</span> bgen_list<span class="op">$</span>samples
  )
}</code></pre></div>
<p>Use the <strong>rbgen</strong> package to read BGEN data into R:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ranges =<span class="st"> </span><span class="kw">data.frame</span>(
  <span class="dt">chromosome =</span> <span class="st">"01"</span>,
  <span class="dt">start =</span> <span class="dv">1</span>,
  <span class="dt">end =</span> <span class="fl">2e9</span>
)
bgen_file &lt;-<span class="st"> </span><span class="kw">path.expand</span>(<span class="st">"~/Téléchargements/gavinband-bgen-798eca81c0fa/example/example.8bits.bgen"</span>)
<span class="cf">if</span> (<span class="op">!</span><span class="kw">file.exists</span>(<span class="kw">paste0</span>(bgen_file, <span class="st">".bgi"</span>)))
  <span class="kw">system</span>(glue<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/glue/topics/glue">glue</a></span>(<span class="st">"bgenix -g {bgen_file} -index"</span>))
bgen_list &lt;-<span class="st"> </span>rbgen<span class="op">::</span><span class="kw">bgen.load</span>(bgen_file, ranges)</code></pre></div>
<p>Let’s see the dosages we get for some given probabilities:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(bigstatsr)
test &lt;-<span class="st"> </span><span class="kw">bgen2FBM</span>(bgen_list)
<span class="kw">t</span>(bgen_list<span class="op">$</span>data[<span class="dv">1</span><span class="op">:</span><span class="dv">12</span>, <span class="dv">1</span>, ])</code></pre></div>
<pre><code>##        RSID_101 RSID_2    RSID_102      RSID_3    RSID_103
## g=0 0.000000000     NA 0.917647059 0.003921569 0.992156863
## g=1 0.003921569     NA 0.007843137 0.003921569 0.003921569
## g=2 0.996078431     NA 0.074509804 0.992156863 0.003921569
##          RSID_4    RSID_104      RSID_5    RSID_105      RSID_6
## g=0 0.992156863 0.007843137 0.003921569 0.968627451 0.003921569
## g=1 0.000000000 0.000000000 0.027450980 0.027450980 0.996078431
## g=2 0.007843137 0.992156863 0.968627451 0.003921569 0.000000000
##        RSID_106 RSID_7
## g=0 0.000000000     NA
## g=1 0.996078431     NA
## g=2 0.003921569     NA</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">test<span class="op">$</span>genotypes[<span class="dv">1</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">12</span>]</code></pre></div>
<pre><code>##  [1] 2.00   NA 0.16 1.99 0.01 0.02 1.98 1.96 0.04 1.00 1.00   NA</code></pre>
<p>BGEN files store twice information as FBM files (probabilities vs dosages) but are compressed so that FBMs are just slightly smaller files.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">file.size</span>(test<span class="op">$</span>genotypes<span class="op">$</span>backingfile)</code></pre></div>
<pre><code>## [1] 99500</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">file.size</span>(bgen_file)</code></pre></div>
<pre><code>## [1] 128746</code></pre>
</div>
<div id="second-solution-using-plink-2-0" class="section level2">
<h2 class="hasAnchor">
<a href="#second-solution-using-plink-2-0" class="anchor"></a>Second solution: using PLINK 2.0</h2>
<p>The <a href="https://www.cog-genomics.org/plink/2.0/formats#pgen">PGEN</a> format is binary format of <a href="https://www.cog-genomics.org/plink/2.0/">PLINK 2.0</a> and can store dosages. PLINK 2.0 can also convert from BGEN to PGEN files. Yet, it’s not straightforward to read PGEN files either. Yet, PLINK has another function to write these dosages to a text file, which is easier to read from.</p>
<p>First, let us get variant and individual information:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">plink2 &lt;-<span class="st"> </span><span class="kw">path.expand</span>(<span class="st">"~/Téléchargements/gavinband-bgen-798eca81c0fa/plink2"</span>)
tmp &lt;-<span class="st"> </span><span class="kw">tempfile</span>()
<span class="co"># Write bim and fam files from the BGEN file</span>
<span class="kw">system</span>(glue<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/glue/topics/glue">glue</a></span>(<span class="st">"{plink2} --bgen {bgen_file}"</span>,
                  <span class="st">" --make-just-bim --make-just-fam"</span>,
                  <span class="st">" --sort-vars"</span>,
                  <span class="st">" --out {tmp}"</span>))
<span class="co"># Get the variants and individual information</span>
fam &lt;-<span class="st"> </span>data.table<span class="op">::</span><span class="kw">fread</span>(<span class="kw">paste0</span>(tmp, <span class="st">".fam"</span>), <span class="dt">data.table =</span> <span class="ot">FALSE</span>)
<span class="kw">names</span>(fam) &lt;-<span class="st"> </span>bigsnpr<span class="op">:::</span>NAMES.FAM
n &lt;-<span class="st"> </span><span class="kw">nrow</span>(fam)
bim &lt;-<span class="st"> </span>data.table<span class="op">::</span><span class="kw">fread</span>(<span class="kw">paste0</span>(tmp, <span class="st">".bim"</span>), <span class="dt">data.table =</span> <span class="ot">FALSE</span>)
<span class="kw">names</span>(bim) &lt;-<span class="st"> </span>bigsnpr<span class="op">:::</span>NAMES.MAP
m &lt;-<span class="st"> </span><span class="kw">nrow</span>(bim)</code></pre></div>
<p>Then, we write the text file of dosages:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">system</span>(glue<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/glue/topics/glue">glue</a></span>(<span class="st">"{plink2} --bgen {bgen_file}"</span>,
                  <span class="st">" --export A"</span>,
                  <span class="st">" --out {tmp}"</span>))</code></pre></div>
<p>Now, let us read these dosages into our FBM.code256 format:</p>
<div class="sourceCode"><pre class="sourceCode cpp"><code class="sourceCode cpp"><span class="co">/**** Simple Rcpp function to convert dosages in a data.frame to our format ****/</span>

<span class="co">// [[Rcpp::depends(BH, bigstatsr)]]</span>
<span class="pp">#include </span><span class="im">&lt;bigstatsr/BMCodeAcc.h&gt;</span>

<span class="co">// [[Rcpp::export]]</span>
<span class="dt">void</span> fill_FBM(Environment BM,
              <span class="at">const</span> IntegerVector&amp; colInd,
              <span class="at">const</span> DataFrame&amp; df) {
  
  Rcout &lt;&lt; colInd &lt;&lt; <span class="bu">std::</span>endl;
  
  XPtr&lt;FBM&gt; xpBM = BM[<span class="st">"address"</span>];
  SubBMAcc&lt;<span class="dt">unsigned</span> <span class="dt">char</span>&gt; macc(xpBM, seq_len(xpBM-&gt;nrow()) - <span class="dv">1</span>, colInd - <span class="dv">1</span>);
  
  <span class="dt">double</span> dosage;
  
  <span class="cf">for</span> (<span class="dt">size_t</span> j = <span class="dv">0</span>; j &lt; macc.ncol(); j++) {
    NumericVector variant = df[j];
    <span class="cf">for</span> (<span class="dt">size_t</span> i = <span class="dv">0</span>; i &lt; macc.nrow(); i++) {
      dosage = variant[i];
      macc(i, j) = R_IsNA(dosage) ? <span class="dv">208</span> : round(<span class="dv">100</span> * dosage + <span class="dv">7</span>);
    }
  }
}</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Get variants IDs to read</span>
colnames &lt;-<span class="st"> </span><span class="kw">names</span>(data.table<span class="op">::</span><span class="kw">fread</span>(<span class="kw">paste0</span>(tmp, <span class="st">".raw"</span>), <span class="dt">nrows =</span> <span class="dv">0</span>))
<span class="kw">all.equal</span>(bim<span class="op">$</span>marker.ID, <span class="kw">as.character</span>(test<span class="op">$</span>variants<span class="op">$</span>rsid))</code></pre></div>
<pre><code>## [1] TRUE</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">(cols &lt;-<span class="st"> </span><span class="kw">match</span>(<span class="kw">paste0</span>(bim<span class="op">$</span>marker.ID, <span class="st">"_G"</span>), colnames))</code></pre></div>
<pre><code>##   [1] 106   7 107   8 108   9 109  10 110  11 111  12 112  13 113
##  [16]  14 114  15 105 115  16 116  17 117  18 118  19 119  20 120
##  [31]  21 121  22 122  23 123  24 124  25 125  26 126  27 127  28
##  [46] 128  29 129  30 130  31 131  32 132  33 133  34 134  35 135
##  [61]  36 136  37 137  38 138  39 139  40 140  41 141  42 142  43
##  [76] 143  44 144  45 145  46 146  47 147  48 148  49 149  50 150
##  [91]  51 151  52 152  53 153  54 154  55 155  56 156  57 157  58
## [106] 158  59 159  60 160  61 161  62 162  63 163  64 164  65 165
## [121]  66 166  67 167  68 168  69 169  70 170  71 171  72 172  73
## [136] 173  74 174  75 175  76 176  77 177  78 178  79 179  80 180
## [151]  81 181  82 182  83 183  84 184  85 185  86 186  87 187  88
## [166] 188  89 189  90 190  91 191  92 192  93 193  94 194  95 195
## [181]  96 196  97 197  98 198  99 199 100 200 101 201 102 202 103
## [196] 203 104 204 205</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Construct a temporary genotype big.matrix</span>
bigGeno2 &lt;-<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/bigstatsr/topics/FBM.code256-class">FBM.code256</a></span>(n, m, <span class="dt">code =</span> bigsnpr<span class="op">:::</span>CODE_DOSAGE)
<span class="co"># Fill it by blocks</span>
<span class="kw"><a href="http://www.rdocumentation.org/packages/bigstatsr/topics/big_apply">big_apply</a></span>(bigGeno2, <span class="dt">a.FUN =</span> <span class="cf">function</span>(G, ind) {
  df &lt;-<span class="st"> </span>data.table<span class="op">::</span><span class="kw">fread</span>(<span class="kw">paste0</span>(tmp, <span class="st">".raw"</span>), <span class="dt">select =</span> colnames[cols[ind]])
  <span class="kw">fill_FBM</span>(G, <span class="dt">colInd =</span> ind, <span class="dt">df =</span> df)  <span class="co"># returns NULL</span>
}, <span class="dt">a.combine =</span> <span class="st">'c'</span>, <span class="dt">block.size =</span> <span class="dv">20</span>)</code></pre></div>
<pre><code>## 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20
## 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40
## 41 42 43 44 45 46 47 48 49 50 51 52 53 54 55 56 57 58 59 60
## 61 62 63 64 65 66 67 68 69 70 71 72 73 74 75 76 77 78 79 80
## 81 82 83 84 85 86 87 88 89 90 91 92 93 94 95 96 97 98 99 100
## 101 102 103 104 105 106 107 108 109 110 111 112 113 114 115 116 117 118 119
## 120 121 122 123 124 125 126 127 128 129 130 131 132 133 134 135 136 137 138 139
## 140 141 142 143 144 145 146 147 148 149 150 151 152 153 154 155 156 157 158 159
## 160 161 162 163 164 165 166 167 168 169 170 171 172 173 174 175 176 177 178 179
## 180 181 182 183 184 185 186 187 188 189 190 191 192 193 194 195 196 197 198 199</code></pre>
<pre><code>## NULL</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Construct the bigSNP object</span>
test2 &lt;-<span class="st"> </span><span class="kw">structure</span>(<span class="kw">list</span>(<span class="dt">genotypes =</span> bigGeno2, <span class="dt">fam =</span> fam, <span class="dt">map =</span> bim), <span class="dt">class =</span> <span class="st">"bigSNP"</span>)
<span class="kw">str</span>(test2, <span class="dt">max.level =</span> <span class="dv">2</span>)</code></pre></div>
<pre><code>## List of 3
##  $ genotypes:Reference class 'FBM.code256' [package "bigstatsr"] with 7 fields
##   ..and 21 methods, of which 7 are  possibly relevant:
##   ..  as.FBM, copy#envRefClass, initialize, initialize#FBM,
##   ..  save, show#envRefClass, show#FBM
##  $ fam      :'data.frame':   500 obs. of  6 variables:
##   ..$ family.ID  : chr [1:500] "sample" "sample" "sample" "sample" ...
##   ..$ sample.ID  : int [1:500] 1 2 3 4 5 6 7 8 9 10 ...
##   ..$ paternal.ID: int [1:500] 0 0 0 0 0 0 0 0 0 0 ...
##   ..$ maternal.ID: int [1:500] 0 0 0 0 0 0 0 0 0 0 ...
##   ..$ sex        : int [1:500] 0 0 0 0 0 0 0 0 0 0 ...
##   ..$ affection  : int [1:500] -9 -9 -9 -9 -9 -9 -9 -9 -9 -9 ...
##  $ map      :'data.frame':   199 obs. of  6 variables:
##   ..$ chromosome  : int [1:199] 1 1 1 1 1 1 1 1 1 1 ...
##   ..$ marker.ID   : chr [1:199] "RSID_101" "RSID_2" "RSID_102" "RSID_3" ...
##   ..$ genetic.dist: int [1:199] 0 0 0 0 0 0 0 0 0 0 ...
##   ..$ physical.pos: int [1:199] 1001 2000 2001 3000 3001 4000 4001 5000 5001 6000 ...
##   ..$ allele1     : chr [1:199] "A" "A" "A" "A" ...
##   ..$ allele2     : chr [1:199] "G" "G" "G" "G" ...
##  - attr(*, "class")= chr "bigSNP"</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Verification</span>
test2<span class="op">$</span>genotypes[<span class="dv">1</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">12</span>]</code></pre></div>
<pre><code>##  [1] 2.00   NA 0.16 1.99 0.01 0.02 1.98 1.96 0.04 1.00 1.00   NA</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">all.equal</span>(test2<span class="op">$</span>genotypes[], test<span class="op">$</span>genotypes[])</code></pre></div>
<pre><code>## [1] TRUE</code></pre>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#the-fbm-code256-format">The FBM.code256 format</a></li>
      <li><a href="#first-solution-using-the-r-package-rbgen">First solution: Using the R package rbgen</a></li>
      <li><a href="#second-solution-using-plink-2-0">Second solution: using PLINK 2.0</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Florian Privé.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  </body>
</html>
