<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Exercises using packages bigsnpr and bigstatsr • bigsnpr</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Exercises using packages bigsnpr and bigstatsr">
<meta property="og:description" content="bigsnpr">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">bigsnpr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.4.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Manual</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/demo.html">Quick demo</a>
    </li>
    <li>
      <a href="../articles/exo.html">Exercise</a>
    </li>
    <li>
      <a href="../articles/pruning-vs-clumping.html">Why clumping should be preferred over pruning</a>
    </li>
    <li>
      <a href="../articles/how-to-PCA.html">How to capture Population Structure with PCA (LD problem explained)</a>
    </li>
    <li>
      <a href="../articles/bedpca.html">How to capture Population Structure with PCA (directly on PLINK bed files)</a>
    </li>
    <li>
      <a href="../articles/SCT.html">Computing polygenic scores using Stacked Clumping and Thresholding (SCT)</a>
    </li>
    <li>
      <a href="../articles/LDpred2.html">Computing polygenic scores using LDpred2</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://privefl.github.io/about.html">
    <span class="fa fa-user"></span>
     
    About
  </a>
</li>
<li>
  <a href="https://github.com/privefl/bigsnpr">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Exercises using packages bigsnpr and bigstatsr</h1>
                        <h4 class="author">Florian Privé</h4>
            
            <h4 class="date">April 18, 2019</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/privefl/bigsnpr/blob/master/vignettes/exo.Rmd"><code>vignettes/exo.Rmd</code></a></small>
      <div class="hidden name"><code>exo.Rmd</code></div>

    </div>

    
    
<p>In this document, I show how to use some of the features of packages {bigsnpr} and {bigstatsr}. Note that many functions used here come from package {bigstatsr} and could therefore be used on other data encoded as matrix-like (outside of the field of genotype data).</p>
<div id="get-data" class="section level2">
<h2 class="hasAnchor">
<a href="#get-data" class="anchor"></a>Get data</h2>
<p>Download <a href="https://github.com/privefl/bigsnpr/raw/master/data-raw/public-data.zip">data</a> and unzip files. I store those files in a directory called <code>"tmp-data"</code> here.</p>
<p>You can see <a href="https://github.com/privefl/bigsnpr/blob/master/data-raw/public-data.R">there</a> how I generated these data from <a href="https://www.nature.com/articles/nature15393">the 1000 Genomes project</a>.</p>
</div>
<div id="read-from-the-plink-files" class="section level2">
<h2 class="hasAnchor">
<a href="#read-from-the-plink-files" class="anchor"></a>Read from the PLINK files</h2>
<div class="sourceCode" id="cb1"><html><body><pre class="r"><span class="co"># Load packages bigsnpr and bigstatsr</span>
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">bigsnpr</span>)</pre></body></html></div>
<pre><code>## Loading required package: bigstatsr</code></pre>
<div class="sourceCode" id="cb3"><html><body><pre class="r"><span class="co"># Read from bed/bim/fam, it will create new files.</span>
<span class="fu"><a href="../reference/snp_readBed.html">snp_readBed</a></span>(<span class="st">"tmp-data/public-data.bed"</span>)</pre></body></html></div>
<pre><code>## [1] "C:\\Users\\au639593\\Desktop\\bigsnpr\\tmp-data\\public-data.rds"</code></pre>
<div class="sourceCode" id="cb5"><html><body><pre class="r"><span class="co"># Attach the "bigSNP" object in R session</span>
<span class="no">obj.bigSNP</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/snp_attach.html">snp_attach</a></span>(<span class="st">"tmp-data/public-data.rds"</span>)
<span class="co"># See how it looks like</span>
<span class="fu"><a href="https://rdrr.io/r/utils/str.html">str</a></span>(<span class="no">obj.bigSNP</span>, <span class="kw">max.level</span> <span class="kw">=</span> <span class="fl">2</span>, <span class="kw">strict.width</span> <span class="kw">=</span> <span class="st">"cut"</span>)</pre></body></html></div>
<pre><code>## List of 3
##  $ genotypes:Reference class 'FBM.code256' [package "bigstatsr"] with 15 ..
##   ..and 26 methods, of which 12 are  possibly relevant:
##   ..  add_columns, as.FBM, bm, bm.desc, check_dimensions,
##   ..  check_write_permissions, copy#envRefClass, initialize,
##   ..  initialize#FBM, save, show#envRefClass, show#FBM
##  $ fam      :'data.frame':   559 obs. of  6 variables:
##   ..$ family.ID  : chr [1:559] "EUR_GBR" "EUR_GBR" "EUR_GBR" "EUR_GBR" ...
##   ..$ sample.ID  : chr [1:559] "HG00096" "HG00097" "HG00099" "HG00100" ...
##   ..$ paternal.ID: int [1:559] 0 0 0 0 0 0 0 0 0 0 ...
##   ..$ maternal.ID: int [1:559] 0 0 0 0 0 0 0 0 0 0 ...
##   ..$ sex        : int [1:559] 1 2 2 2 1 2 1 2 2 1 ...
##   ..$ affection  : int [1:559] 1 2 1 1 1 1 2 1 2 1 ...
##  $ map      :'data.frame':   130816 obs. of  6 variables:
##   ..$ chromosome  : int [1:130816] 2 2 2 2 2 2 2 2 2 2 ...
##   ..$ marker.ID   : chr [1:130816] "rs13400442" "rs7594567" "rs7597758" "..
##   ..$ genetic.dist: int [1:130816] 0 0 0 0 0 0 0 0 0 0 ...
##   ..$ physical.pos: int [1:130816] 18506 21833 22398 28228 32003 32005 36..
##   ..$ allele1     : chr [1:130816] "C" "G" "T" "A" ...
##   ..$ allele2     : chr [1:130816] "T" "C" "C" "G" ...
##  - attr(*, "class")= chr "bigSNP"</code></pre>
<div class="sourceCode" id="cb7"><html><body><pre class="r"><span class="co"># Get aliases for useful slots</span>
<span class="no">G</span>   <span class="kw">&lt;-</span> <span class="no">obj.bigSNP</span>$<span class="no">genotypes</span>
<span class="no">CHR</span> <span class="kw">&lt;-</span> <span class="no">obj.bigSNP</span>$<span class="no">map</span>$<span class="no">chromosome</span>
<span class="no">POS</span> <span class="kw">&lt;-</span> <span class="no">obj.bigSNP</span>$<span class="no">map</span>$<span class="no">physical.pos</span>
<span class="no">y</span>   <span class="kw">&lt;-</span> <span class="no">obj.bigSNP</span>$<span class="no">fam</span>$<span class="no">affection</span> - <span class="fl">1</span>
<span class="no">sex</span> <span class="kw">&lt;-</span> <span class="no">obj.bigSNP</span>$<span class="no">fam</span>$<span class="no">sex</span>
<span class="no">pop</span> <span class="kw">&lt;-</span> <span class="no">obj.bigSNP</span>$<span class="no">fam</span>$<span class="no">family.ID</span>
<span class="no">NCORES</span> <span class="kw">&lt;-</span> <span class="fu">nb_cores</span>()
<span class="co"># Check some counts for the 10 first SNPs</span>
<span class="fu">big_counts</span>(<span class="no">G</span>, <span class="kw">ind.col</span> <span class="kw">=</span> <span class="fl">1</span>:<span class="fl">10</span>)</pre></body></html></div>
<pre><code>##      [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8] [,9] [,10]
## 0     504  468  421  483  476  420  380  383  478   389
## 1      48   69  107   66   73  124  136  135   70   133
## 2       7   22   31   10   10   15   43   41   11    37
## &lt;NA&gt;    0    0    0    0    0    0    0    0    0     0</code></pre>
</div>
<div id="what-you-must-do" class="section level2">
<h2 class="hasAnchor">
<a href="#what-you-must-do" class="anchor"></a>What you must do</h2>
<p>You need to</p>
<ul>
<li><p>Explore the data</p></li>
<li><p>Assess population structure of the data</p></li>
<li><p>Find the variables associated with the disease status (<code>y</code>)</p></li>
<li>
<p>Predict the disease status using the following indices as training/test sets:</p>
<div class="sourceCode" id="cb9"><html><body><pre class="r"><span class="co"># Divide the indices in training/test sets</span>
<span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span>(<span class="fl">1</span>)
<span class="no">ind.train</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span>(<span class="no">G</span>), <span class="fl">400</span>)
<span class="no">ind.test</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sets.html">setdiff</a></span>(<span class="fu">rows_along</span>(<span class="no">G</span>), <span class="no">ind.train</span>)</pre></body></html></div>
</li>
</ul>
<p>For this, you can use whatever tools you want because the data is quite small. In the following section, I give some (scalable) solutions using using packages {bigstatsr} and {bigsnpr}.</p>
</div>
<div id="solution-using-bigstatsr-and-bigsnpr" class="section level2">
<h2 class="hasAnchor">
<a href="#solution-using-bigstatsr-and-bigsnpr" class="anchor"></a>Solution using {bigstatsr} and {bigsnpr}</h2>
<div id="population-structure-principal-component-analysis" class="section level3">
<h3 class="hasAnchor">
<a href="#population-structure-principal-component-analysis" class="anchor"></a>Population structure: Principal Component Analysis</h3>
<p>Let us compute first principal components of the scaled genotype matrix:</p>
<div class="sourceCode" id="cb10"><html><body><pre class="r"><span class="co"># Compute partial SVD (10 PCs by default) using random projections</span>
<span class="co"># big_scale() computes means and standard deviations for scaling</span>
<span class="no">svd</span> <span class="kw">&lt;-</span> <span class="fu">big_randomSVD</span>(<span class="no">G</span>, <span class="fu">big_scale</span>(), <span class="kw">ncores</span> <span class="kw">=</span> <span class="no">NCORES</span>)</pre></body></html></div>
<div class="sourceCode" id="cb11"><html><body><pre class="r"><span class="co"># Scree plot</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.html">plot</a></span>(<span class="no">svd</span>)</pre></body></html></div>
<p><img src="exo_files/figure-html/unnamed-chunk-6-1.png" width="692.708333333333" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb12"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">ggplot2</span>)</pre></body></html></div>
<pre><code>## Warning: package 'ggplot2' was built under R version 3.6.3</code></pre>
<div class="sourceCode" id="cb14"><html><body><pre class="r"><span class="co"># Scores plot + color for population</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.html">plot</a></span>(<span class="no">svd</span>, <span class="kw">type</span> <span class="kw">=</span> <span class="st">"scores"</span>) +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="kw">color</span> <span class="kw">=</span> <span class="no">pop</span>)</pre></body></html></div>
<p><img src="exo_files/figure-html/unnamed-chunk-6-2.png" width="692.708333333333" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb15"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/graphics/plot.html">plot</a></span>(<span class="no">svd</span>, <span class="kw">type</span> <span class="kw">=</span> <span class="st">"scores"</span>, <span class="kw">scores</span> <span class="kw">=</span> <span class="fl">3</span>:<span class="fl">4</span>) +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="kw">color</span> <span class="kw">=</span> <span class="no">pop</span>)</pre></body></html></div>
<p><img src="exo_files/figure-html/unnamed-chunk-6-3.png" width="692.708333333333" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb16"><html><body><pre class="r"><span class="co"># Loadings (effects of each variable for each PC)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.html">plot</a></span>(<span class="no">svd</span>, <span class="kw">type</span> <span class="kw">=</span> <span class="st">"loadings"</span>, <span class="kw">loadings</span> <span class="kw">=</span> <span class="fl">1</span>:<span class="fl">10</span>, <span class="kw">coeff</span> <span class="kw">=</span> <span class="fl">0.4</span>)</pre></body></html></div>
<p><img src="exo_files/figure-html/unnamed-chunk-7-1.png" width="692.708333333333" style="display: block; margin: auto;"></p>
<p>Looking at the loadings, we can see that the PCA captures some variation due to large correlation between variables. To learn more about this possible pitfall, please look at <a href="https://privefl.github.io/bigsnpr/articles/how-to-PCA.html">this vignette</a>.</p>
</div>
<div id="association-genome-wide-association-study-gwas" class="section level3">
<h3 class="hasAnchor">
<a href="#association-genome-wide-association-study-gwas" class="anchor"></a>Association: Genome-Wide Association Study (GWAS)</h3>
<div class="sourceCode" id="cb17"><html><body><pre class="r"><span class="co"># Association of each variable of `G` with `y` (adjusting for 10 PCs)</span>
<span class="no">gwas</span> <span class="kw">&lt;-</span> <span class="fu">big_univLogReg</span>(<span class="no">G</span>, <span class="no">y</span>, <span class="kw">covar.train</span> <span class="kw">=</span> <span class="no">svd</span>$<span class="no">u</span>, <span class="kw">ncores</span> <span class="kw">=</span> <span class="no">NCORES</span>)
<span class="co"># Histogram of p-values</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.html">plot</a></span>(<span class="no">gwas</span>)</pre></body></html></div>
<p><img src="exo_files/figure-html/unnamed-chunk-8-1.png" width="692.708333333333" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb18"><html><body><pre class="r"><span class="co"># Q-Q plot</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.html">plot</a></span>(<span class="no">gwas</span>, <span class="kw">type</span> <span class="kw">=</span> <span class="st">"Q-Q"</span>) + <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html">xlim</a></span>(<span class="fl">1</span>, <span class="fl">NA</span>)  <span class="co"># snp_qq(gwas) + xlim(1, NA)</span></pre></body></html></div>
<pre><code>## Warning: Removed 117734 rows containing missing values (geom_point).</code></pre>
<p><img src="exo_files/figure-html/unnamed-chunk-8-2.png" width="692.708333333333" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb20"><html><body><pre class="r"><span class="co"># Manhattan plot</span>
<span class="fu"><a href="../reference/snp_manhattan.html">snp_manhattan</a></span>(<span class="no">gwas</span>, <span class="no">CHR</span>, <span class="no">POS</span>, <span class="kw">npoints</span> <span class="kw">=</span> <span class="fl">20e3</span>) +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_hline</a></span>(<span class="kw">yintercept</span> <span class="kw">=</span> -<span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span>(<span class="fl">5e-8</span>), <span class="kw">color</span> <span class="kw">=</span> <span class="st">"red"</span>)</pre></body></html></div>
<p><img src="exo_files/figure-html/unnamed-chunk-8-3.png" width="692.708333333333" style="display: block; margin: auto;"></p>
</div>
<div id="polygenic-risk-score-prs" class="section level3">
<h3 class="hasAnchor">
<a href="#polygenic-risk-score-prs" class="anchor"></a>Polygenic Risk Score (PRS)</h3>
<div id="with-clumping-and-thresholding-ct" class="section level4">
<h4 class="hasAnchor">
<a href="#with-clumping-and-thresholding-ct" class="anchor"></a>with Clumping and Thresholding (C+T)</h4>
<p><span class="math display">\[\rm{PRS}_i = \sum_{\substack{j \in S_\text{clumping} \\ p_j~&lt;~p_T}} \hat\beta_j \cdot G_{i,j}~,\]</span></p>
<p>where <span class="math inline">\(\hat\beta_j\)</span> (<span class="math inline">\(p_j\)</span>) are the effect sizes (p-values) estimated from the GWAS and <span class="math inline">\(G_{i,j}\)</span> is the allele count (genotype) for individual <span class="math inline">\(i\)</span> and SNP <span class="math inline">\(j\)</span>.</p>
<div class="sourceCode" id="cb21"><html><body><pre class="r"><span class="no">sumstats</span> <span class="kw">&lt;-</span> <span class="kw pkg">bigreadr</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/bigreadr/man/fread2.html">fread2</a></span>(<span class="st">"tmp-data/public-data-sumstats.txt"</span>)
<span class="no">lpval</span> <span class="kw">&lt;-</span> -<span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span>(<span class="no">sumstats</span>$<span class="no">p</span>)
<span class="no">ind.keep</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/snp_clumping.html">snp_clumping</a></span>(<span class="no">G</span>, <span class="no">CHR</span>, <span class="kw">ind.row</span> <span class="kw">=</span> <span class="no">ind.train</span>, <span class="kw">S</span> <span class="kw">=</span> <span class="no">lpval</span>, <span class="kw">infos.pos</span> <span class="kw">=</span> <span class="no">POS</span>, <span class="kw">ncores</span> <span class="kw">=</span> <span class="fl">3</span>)
<span class="no">THR</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/seq_log.html">seq_log</a></span>(<span class="fl">1</span>, <span class="fl">8</span>, <span class="kw">length.out</span> <span class="kw">=</span> <span class="fl">20</span>)
<span class="no">prs</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/snp_PRS.html">snp_PRS</a></span>(<span class="no">G</span>, <span class="no">sumstats</span>$<span class="no">beta</span>[<span class="no">ind.keep</span>], <span class="kw">ind.keep</span> <span class="kw">=</span> <span class="no">ind.keep</span>,
               <span class="kw">lpS.keep</span> <span class="kw">=</span> <span class="no">lpval</span>[<span class="no">ind.keep</span>], <span class="kw">thr.list</span> <span class="kw">=</span> <span class="no">THR</span>)
<span class="co"># Learn the optimal threshold on the training set</span>
<span class="no">aucs</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html">apply</a></span>(<span class="no">prs</span>[<span class="no">ind.train</span>, ], <span class="fl">2</span>, <span class="no">AUC</span>, <span class="kw">target</span> <span class="kw">=</span> <span class="no">y</span>[<span class="no">ind.train</span>])
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.html">plot</a></span>(<span class="no">THR</span>, <span class="no">aucs</span>, <span class="kw">xlab</span> <span class="kw">=</span> <span class="st">"-log10(p-value)"</span>, <span class="kw">ylab</span> <span class="kw">=</span> <span class="st">"AUC"</span>, <span class="kw">pch</span> <span class="kw">=</span> <span class="fl">20</span>)</pre></body></html></div>
<p><img src="exo_files/figure-html/unnamed-chunk-9-1.png" width="692.708333333333" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb22"><html><body><pre class="r"><span class="co"># Evaluate on the test set</span>
<span class="fu">AUC</span>(<span class="no">prs</span>[<span class="no">ind.test</span>, <span class="fu"><a href="https://rdrr.io/r/base/which.min.html">which.max</a></span>(<span class="no">aucs</span>)], <span class="no">y</span>[<span class="no">ind.test</span>])</pre></body></html></div>
<pre><code>## [1] 0.6689111</code></pre>
<p>(TODO: ADD SCT)</p>
</div>
<div id="with-penalized-logistic-regression-plr" class="section level4">
<h4 class="hasAnchor">
<a href="#with-penalized-logistic-regression-plr" class="anchor"></a>with Penalized Logistic Regression (PLR)</h4>
<p><span class="math display">\[\arg\!\min_{\beta_0,~\beta}(\lambda, \alpha)\left\{ \underbrace{ -\sum_{i=1}^n \left( y_i \log\left(p_i\right) + (1 - y_i) \log\left(1 - p_i\right) \right) }_\text{Loss function} + \underbrace{ \lambda \left((1-\alpha)\frac{1}{2}\|\beta\|_2^2 + \alpha \|\beta\|_1\right) }_\text{Penalization} \right\}\]</span></p>
<p>where</p>
<ul>
<li><p><span class="math inline">\(p_i=1/\left(1+\exp\left(-(\beta_0 + x_i^T\beta)\right)\right)\)</span></p></li>
<li><p><span class="math inline">\(x\)</span> is denoting the genotypes and covariables (e.g. principal components),</p></li>
<li><p><span class="math inline">\(y\)</span> is the disease status we want to predict,</p></li>
<li><p><span class="math inline">\(\lambda\)</span> is a regularization parameter that needs to be determined and</p></li>
<li><p><span class="math inline">\(\alpha\)</span> determines relative parts of the regularization <span class="math inline">\(0 \le \alpha \le 1\)</span>.</p></li>
</ul>
<p><br></p>
<p>If you want to learn more about our implementation of PLR, please look at <a href="https://doi.org/10.1101/403337">this paper</a>.</p>
<p><br></p>
<div class="sourceCode" id="cb24"><html><body><pre class="r"><span class="co"># Penalized logistic regression for many alphas and lambdas</span>
<span class="no">mod</span> <span class="kw">&lt;-</span> <span class="fu">big_spLogReg</span>(<span class="no">G</span>, <span class="no">y</span>[<span class="no">ind.train</span>], <span class="no">ind.train</span>, <span class="kw">covar.train</span> <span class="kw">=</span> <span class="no">svd</span>$<span class="no">u</span>[<span class="no">ind.train</span>, ],
                    <span class="kw">K</span> <span class="kw">=</span> <span class="fl">5</span>, <span class="kw">alphas</span> <span class="kw">=</span> <span class="fl">10</span>^(-(<span class="fl">0</span>:<span class="fl">4</span>)), <span class="kw">ncores</span> <span class="kw">=</span> <span class="no">NCORES</span>)</pre></body></html></div>
<div class="sourceCode" id="cb25"><html><body><pre class="r"><span class="co"># Plot regularization paths (from high lambda to low lambda) </span>
<span class="co"># for each validation set (color) and each alpha (facet)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.html">plot</a></span>(<span class="no">mod</span>)</pre></body></html></div>
<p><img src="exo_files/figure-html/unnamed-chunk-11-1.png" width="692.708333333333" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb26"><html><body><pre class="r"><span class="co"># Get summaries of models</span>
<span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span>(<span class="no">mod</span>)</pre></body></html></div>
<pre><code>## # A tibble: 5 x 7
##    alpha validation_loss intercept beta            nb_var message  all_conv
##    &lt;dbl&gt;           &lt;dbl&gt;     &lt;dbl&gt; &lt;list&gt;           &lt;int&gt; &lt;list&gt;   &lt;lgl&gt;   
## 1 0.0001           0.557     -1.41 &lt;dbl [130,826]&gt;  67876 &lt;chr [5~ TRUE    
## 2 0.001            0.550     -1.94 &lt;dbl [130,826]&gt;  16008 &lt;chr [5~ TRUE    
## 3 0.01             0.547     -2.16 &lt;dbl [130,826]&gt;   2928 &lt;chr [5~ TRUE    
## 4 0.1              0.545     -2.30 &lt;dbl [130,826]&gt;    722 &lt;chr [5~ TRUE    
## 5 1                0.543     -2.40 &lt;dbl [130,826]&gt;    398 &lt;chr [5~ TRUE</code></pre>
<div class="sourceCode" id="cb28"><html><body><pre class="r"><span class="co"># Get the predictions for the test set</span>
<span class="no">pred</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/predict.html">predict</a></span>(<span class="no">mod</span>, <span class="no">G</span>, <span class="no">ind.test</span>, <span class="kw">covar.row</span> <span class="kw">=</span> <span class="no">svd</span>$<span class="no">u</span>[<span class="no">ind.test</span>, ])
<span class="co"># Assess the Area Under the ROC Curve</span>
<span class="fu">AUC</span>(<span class="no">pred</span>, <span class="no">y</span>[<span class="no">ind.test</span>])</pre></body></html></div>
<pre><code>## [1] 0.7148903</code></pre>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Florian Privé.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
