<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Exercises using packages bigsnpr and bigstatsr • bigsnpr</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Exercises using packages bigsnpr and bigstatsr">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">bigsnpr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.9.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Manual</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/demo.html">Quick demo</a>
    </li>
    <li>
      <a href="../articles/exo.html">Exercise</a>
    </li>
    <li>
      <a href="../articles/how-to-PCA.html">How to capture Population Structure with PCA</a>
    </li>
    <li>
      <a href="../articles/pruning-vs-clumping.html">Why clumping should be preferred over pruning</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://privefl.github.io/about.html">
    <span class="fa fa-user"></span>
     
    About
  </a>
</li>
<li>
  <a href="https://github.com/privefl/bigsnpr">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Exercises using packages bigsnpr and bigstatsr</h1>
                        <h4 class="author">Florian Privé</h4>
            
            <h4 class="date">January 22, 2019</h4>
      
      
      <div class="hidden name"><code>exo.Rmd</code></div>

    </div>

    
    
<p>In this document, I show how to use some of the features of packages {bigsnpr} and {bigstatsr}. Note that many functions used here come from package {bigstatsr} and could therefore be used on other data encoded as matrix-like (outside of the field of genotype data).</p>
<div id="get-data" class="section level2">
<h2 class="hasAnchor">
<a href="#get-data" class="anchor"></a>Get data</h2>
<p>Download <a href="https://github.com/privefl/bigsnpr/raw/master/data-raw/public-data.zip">data</a> and unzip files. I store those files in a directory called <code>"tmp-data"</code> here.</p>
<p>You can see <a href="https://github.com/privefl/bigsnpr/blob/master/data-raw/public-data.R">there</a> how I generated these data from <a href="https://www.nature.com/articles/nature15393">the 1000 Genomes project</a>.</p>
</div>
<div id="read-from-the-plink-files" class="section level2">
<h2 class="hasAnchor">
<a href="#read-from-the-plink-files" class="anchor"></a>Read from the PLINK files</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Load packages bigsnpr and bigstatsr</span>
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(bigsnpr)</code></pre></div>
<pre><code>## Le chargement a nécessité le package : bigstatsr</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Read from bed/bim/fam, it will create new files.</span>
<span class="kw"><a href="../reference/snp_readBed.html">snp_readBed</a></span>(<span class="st">"tmp-data/public-data.bed"</span>)</code></pre></div>
<pre><code>## [1] "/home/privef/Bureau/bigsnpr/vignettes/tmp-data/public-data.rds"</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Attach the "bigSNP" object in R session</span>
obj.bigSNP &lt;-<span class="st"> </span><span class="kw"><a href="../reference/snp_attach.html">snp_attach</a></span>(<span class="st">"tmp-data/public-data.rds"</span>)
<span class="co"># See how it looks like</span>
<span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/str">str</a></span>(obj.bigSNP, <span class="dt">max.level =</span> <span class="dv">2</span>, <span class="dt">strict.width =</span> <span class="st">"cut"</span>)</code></pre></div>
<pre><code>## List of 3
##  $ genotypes:Reference class 'FBM.code256' [package "bigstatsr"] with 10 ..
##   ..and 24 methods, of which 10 are  possibly relevant:
##   ..  add_columns, as.FBM, bm, bm.desc, copy#envRefClass, initialize,
##   ..  initialize#FBM, save, show#envRefClass, show#FBM
##  $ fam      :'data.frame':   559 obs. of  6 variables:
##   ..$ family.ID  : chr [1:559] "EUR_GBR" "EUR_GBR" "EUR_GBR" "EUR_GBR" ...
##   ..$ sample.ID  : chr [1:559] "HG00096" "HG00097" "HG00099" "HG00100" ...
##   ..$ paternal.ID: int [1:559] 0 0 0 0 0 0 0 0 0 0 ...
##   ..$ maternal.ID: int [1:559] 0 0 0 0 0 0 0 0 0 0 ...
##   ..$ sex        : int [1:559] 1 2 2 2 1 2 1 2 2 1 ...
##   ..$ affection  : int [1:559] 0 0 1 0 1 1 1 1 0 1 ...
##  $ map      :'data.frame':   130619 obs. of  6 variables:
##   ..$ chromosome  : int [1:130619] 2 2 2 2 2 2 2 2 2 2 ...
##   ..$ marker.ID   : chr [1:130619] "rs28469191" "rs6548219" "rs4522651" "..
##   ..$ genetic.dist: int [1:130619] 0 0 0 0 0 0 0 0 0 0 ...
##   ..$ physical.pos: int [1:130619] 29970 30762 34049 38733 41054 47648 52..
##   ..$ allele1     : chr [1:130619] "C" "A" "A" "G" ...
##   ..$ allele2     : chr [1:130619] "T" "G" "G" "GA" ...
##  - attr(*, "class")= chr "bigSNP"</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Get aliases for useful slots</span>
G   &lt;-<span class="st"> </span>obj.bigSNP<span class="op">$</span>genotypes
CHR &lt;-<span class="st"> </span>obj.bigSNP<span class="op">$</span>map<span class="op">$</span>chromosome
POS &lt;-<span class="st"> </span>obj.bigSNP<span class="op">$</span>map<span class="op">$</span>physical.pos
y   &lt;-<span class="st"> </span>obj.bigSNP<span class="op">$</span>fam<span class="op">$</span>affection
sex &lt;-<span class="st"> </span>obj.bigSNP<span class="op">$</span>fam<span class="op">$</span>sex
pop &lt;-<span class="st"> </span>obj.bigSNP<span class="op">$</span>fam<span class="op">$</span>family.ID
NCORES &lt;-<span class="st"> </span><span class="kw">nb_cores</span>()
<span class="co"># Check some counts for the 10 first SNPs</span>
<span class="kw">big_counts</span>(G, <span class="dt">ind.col =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">10</span>)</code></pre></div>
<pre><code>##      [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8] [,9] [,10]
## 0     374  335  453  378  492  483  479  483  479   469
## 1     142  178   85  138   62   69   70   69   71    78
## 2      43   46   21   43    5    7   10    7    9    12
## &lt;NA&gt;    0    0    0    0    0    0    0    0    0     0</code></pre>
</div>
<div id="what-you-must-do" class="section level2">
<h2 class="hasAnchor">
<a href="#what-you-must-do" class="anchor"></a>What you must do</h2>
<p>You need to</p>
<ul>
<li><p>Explore the data</p></li>
<li><p>Assess population structure of the data</p></li>
<li><p>Find the variables associated with the disease status (<code>y</code>)</p></li>
<li>
<p>Predict the disease status using the following indices as training/test sets:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Divide the indices in training/test sets</span>
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/Random">set.seed</a></span>(<span class="dv">1</span>)
ind.train &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/sample">sample</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/nrow">nrow</a></span>(G), <span class="dv">400</span>)
ind.test &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/sets">setdiff</a></span>(<span class="kw">rows_along</span>(G), ind.train)</code></pre></div>
</li>
</ul>
<p>For this, you can use whatever tools you want because the data is quite small. In the following section, I give some (scalable) solutions using using packages {bigstatsr} and {bigsnpr}.</p>
</div>
<div id="solution-using-bigstatsr-and-bigsnpr" class="section level2">
<h2 class="hasAnchor">
<a href="#solution-using-bigstatsr-and-bigsnpr" class="anchor"></a>Solution using {bigstatsr} and {bigsnpr}</h2>
<div id="population-structure-principal-component-analysis" class="section level3">
<h3 class="hasAnchor">
<a href="#population-structure-principal-component-analysis" class="anchor"></a>Population structure: Principal Component Analysis</h3>
<p>Let us compute first principal components of the scaled genotype matrix:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Compute partial SVD (10 PCs by default) using random projections</span>
<span class="co"># big_scale() computes means and standard deviations for scaling</span>
svd &lt;-<span class="st"> </span><span class="kw">big_randomSVD</span>(G, <span class="kw">big_scale</span>(), <span class="dt">ncores =</span> NCORES)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Scree plot</span>
<span class="kw"><a href="https://www.rdocumentation.org/packages/graphics/topics/plot">plot</a></span>(svd)</code></pre></div>
<p><img src="exo_files/figure-html/unnamed-chunk-4-1.png" width="692.708333333333" style="display: block; margin: auto;"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(ggplot2)
<span class="co"># Scores plot + color for population</span>
<span class="kw"><a href="https://www.rdocumentation.org/packages/graphics/topics/plot">plot</a></span>(svd, <span class="dt">type =</span> <span class="st">"scores"</span>) <span class="op">+</span>
<span class="st">  </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="dt">color =</span> pop)</code></pre></div>
<p><img src="exo_files/figure-html/unnamed-chunk-4-2.png" width="692.708333333333" style="display: block; margin: auto;"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/graphics/topics/plot">plot</a></span>(svd, <span class="dt">type =</span> <span class="st">"scores"</span>, <span class="dt">scores =</span> <span class="dv">3</span><span class="op">:</span><span class="dv">4</span>) <span class="op">+</span>
<span class="st">  </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="dt">color =</span> pop)</code></pre></div>
<p><img src="exo_files/figure-html/unnamed-chunk-4-3.png" width="692.708333333333" style="display: block; margin: auto;"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Loadings (effects of each variable for each PC)</span>
<span class="kw"><a href="https://www.rdocumentation.org/packages/graphics/topics/plot">plot</a></span>(svd, <span class="dt">type =</span> <span class="st">"loadings"</span>, <span class="dt">loadings =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">10</span>, <span class="dt">coeff =</span> <span class="fl">0.4</span>)</code></pre></div>
<p><img src="exo_files/figure-html/unnamed-chunk-5-1.png" width="692.708333333333" style="display: block; margin: auto;"></p>
<p>Looking at the loadings, we can see that the PCA captures some variation due to large correlation between variables. To learn more about this possible pitfall, please look at <a href="https://privefl.github.io/bigsnpr/articles/how-to-PCA.html">this vignette</a>.</p>
</div>
<div id="association-genome-wide-association-study-gwas" class="section level3">
<h3 class="hasAnchor">
<a href="#association-genome-wide-association-study-gwas" class="anchor"></a>Association: Genome-Wide Association Study (GWAS)</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Association of each variable of `G` with `y` (adjusting for 10 PCs)</span>
gwas &lt;-<span class="st"> </span><span class="kw">big_univLogReg</span>(G, y, <span class="dt">covar.train =</span> svd<span class="op">$</span>u, <span class="dt">ncores =</span> NCORES)</code></pre></div>
<pre><code>## For 1 columns, IRLS didn't converge; `glm` was used instead.</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Histogram of p-values</span>
<span class="kw"><a href="https://www.rdocumentation.org/packages/graphics/topics/plot">plot</a></span>(gwas)</code></pre></div>
<p><img src="exo_files/figure-html/unnamed-chunk-6-1.png" width="692.708333333333" style="display: block; margin: auto;"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Q-Q plot</span>
<span class="kw"><a href="https://www.rdocumentation.org/packages/graphics/topics/plot">plot</a></span>(gwas, <span class="dt">type =</span> <span class="st">"Q-Q"</span>) <span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/lims.html">xlim</a></span>(<span class="dv">1</span>, <span class="ot">NA</span>)  <span class="co"># snp_qq(gwas) + xlim(1, NA)</span></code></pre></div>
<pre><code>## Warning: Removed 117557 rows containing missing values (geom_point).</code></pre>
<p><img src="exo_files/figure-html/unnamed-chunk-6-2.png" width="692.708333333333" style="display: block; margin: auto;"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Manhattan plot</span>
<span class="kw"><a href="../reference/snp_manhattan.html">snp_manhattan</a></span>(gwas, CHR, POS, <span class="dt">npoints =</span> <span class="fl">20e3</span>) <span class="op">+</span>
<span class="st">  </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_hline</a></span>(<span class="dt">yintercept =</span> <span class="op">-</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/Log">log10</a></span>(<span class="fl">5e-8</span>), <span class="dt">color =</span> <span class="st">"red"</span>)</code></pre></div>
<p><img src="exo_files/figure-html/unnamed-chunk-6-3.png" width="692.708333333333" style="display: block; margin: auto;"></p>
</div>
<div id="polygenic-risk-score-prs-with-penalized-logistic-regression-plr" class="section level3">
<h3 class="hasAnchor">
<a href="#polygenic-risk-score-prs-with-penalized-logistic-regression-plr" class="anchor"></a>Polygenic Risk Score (PRS) with Penalized Logistic Regression (PLR)</h3>
<p><span class="math display">\[\arg\!\min_{\beta_0,~\beta}(\lambda, \alpha)\left\{ \underbrace{ -\sum_{i=1}^n \left( y_i \log\left(p_i\right) + (1 - y_i) \log\left(1 - p_i\right) \right) }_\text{Loss function} + \underbrace{ \lambda \left((1-\alpha)\frac{1}{2}\|\beta\|_2^2 + \alpha \|\beta\|_1\right) }_\text{Penalization} \right\}\]</span></p>
<p>where</p>
<ul>
<li><p><span class="math inline">\(p_i=1/\left(1+\exp\left(-(\beta_0 + x_i^T\beta)\right)\right)\)</span></p></li>
<li><p><span class="math inline">\(x\)</span> is denoting the genotypes and covariables (e.g. principal components),</p></li>
<li><p><span class="math inline">\(y\)</span> is the disease status we want to predict,</p></li>
<li><p><span class="math inline">\(\lambda\)</span> is a regularization parameter that needs to be determined and</p></li>
<li><p><span class="math inline">\(\alpha\)</span> determines relative parts of the regularization <span class="math inline">\(0 \le \alpha \le 1\)</span>.</p></li>
</ul>
<p><br></p>
<p>If you want to learn more about our implementation of PLR, please look at <a href="https://doi.org/10.1101/403337">this paper</a>.</p>
<p><br></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Penalized logistic regression for many alphas and lambdas</span>
mod &lt;-<span class="st"> </span><span class="kw">big_spLogReg</span>(G, y[ind.train], ind.train, <span class="dt">covar.train =</span> svd<span class="op">$</span>u[ind.train, ],
                    <span class="dt">K =</span> <span class="dv">5</span>, <span class="dt">alphas =</span> <span class="dv">10</span><span class="op">^</span>(<span class="op">-</span>(<span class="dv">0</span><span class="op">:</span><span class="dv">4</span>)), <span class="dt">ncores =</span> NCORES)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Plot regularization paths (from high lambda to low lambda) </span>
<span class="co"># for each validation set (color) and each alpha (facet)</span>
<span class="kw"><a href="https://www.rdocumentation.org/packages/graphics/topics/plot">plot</a></span>(mod)</code></pre></div>
<p><img src="exo_files/figure-html/unnamed-chunk-8-1.png" width="692.708333333333" style="display: block; margin: auto;"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Get summaries of models</span>
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/summary">summary</a></span>(mod)</code></pre></div>
<pre><code>## # A tibble: 5 x 6
##    alpha validation_loss intercept beta            nb_var message  
##    &lt;dbl&gt;           &lt;dbl&gt;     &lt;dbl&gt; &lt;list&gt;           &lt;int&gt; &lt;list&gt;   
## 1 0.0001           0.552     -1.56 &lt;dbl [130,629]&gt;  57869 &lt;chr [5]&gt;
## 2 0.001            0.533     -3.09 &lt;dbl [130,629]&gt;  12255 &lt;chr [5]&gt;
## 3 0.01             0.486     -4.61 &lt;dbl [130,629]&gt;   2156 &lt;chr [5]&gt;
## 4 0.1              0.452     -3.52 &lt;dbl [130,629]&gt;    775 &lt;chr [5]&gt;
## 5 1                0.434     -3.24 &lt;dbl [130,629]&gt;    367 &lt;chr [5]&gt;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Get the predictions for the test set</span>
pred &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/predict">predict</a></span>(mod, G, ind.test, <span class="dt">covar.row =</span> svd<span class="op">$</span>u[ind.test, ])
<span class="co"># Assess the Area Under the ROC Curve</span>
<span class="kw">AUC</span>(pred, y[ind.test])</code></pre></div>
<pre><code>## [1] 0.8294991</code></pre>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#get-data">Get data</a></li>
      <li><a href="#read-from-the-plink-files">Read from the PLINK files</a></li>
      <li><a href="#what-you-must-do">What you must do</a></li>
      <li><a href="#solution-using-bigstatsr-and-bigsnpr">Solution using {bigstatsr} and {bigsnpr}</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Florian Privé.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
