<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Computing polygenic scores using LDpred2 • bigsnpr</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Computing polygenic scores using LDpred2">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">bigsnpr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.3.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Manual</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/demo.html">Quick demo</a>
    </li>
    <li>
      <a href="../articles/exo.html">Exercise</a>
    </li>
    <li>
      <a href="../articles/pruning-vs-clumping.html">Why clumping should be preferred over pruning</a>
    </li>
    <li>
      <a href="../articles/how-to-PCA.html">How to capture Population Structure with PCA (LD problem explained)</a>
    </li>
    <li>
      <a href="../articles/bedpca.html">How to capture Population Structure with PCA (directly on PLINK bed files)</a>
    </li>
    <li>
      <a href="../articles/SCT.html">Computing polygenic scores using Stacked Clumping and Thresholding (SCT)</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://privefl.github.io/about.html">
    <span class="fa fa-user"></span>
     
    About
  </a>
</li>
<li>
  <a href="https://github.com/privefl/bigsnpr">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Computing polygenic scores using LDpred2</h1>
                        <h4 class="author">Florian Privé</h4>
            
            <h4 class="date">April 2, 2020</h4>
      
      
      <div class="hidden name"><code>LDpred2.Rmd</code></div>

    </div>

    
    
<p>In this document, we show how to compute polygenic risk scores using LDpred2.</p>
<div id="downloading-genotype-data-and-summary-statistics" class="section level2">
<h2 class="hasAnchor">
<a href="#downloading-genotype-data-and-summary-statistics" class="anchor"></a>Downloading genotype data and summary statistics</h2>
<p>You can download <a href="https://github.com/privefl/bigsnpr/raw/master/data-raw/public-data.zip">data</a> and unzip files in R. We store those files in a directory called <code>"tmp-data"</code> here.</p>
<p>You can see <a href="https://github.com/privefl/bigsnpr/blob/master/data-raw/public-data.R">there</a> how we generated these data from <a href="https://www.nature.com/articles/nature15393">the 1000 Genomes project</a>.</p>
<p>First, you need to read genotype data from the PLINK files (or BGEN files) as well as the text file containing summary statistics.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" title="1"><span class="co"># Load packages bigsnpr and bigstatsr</span></a>
<a class="sourceLine" id="cb1-2" title="2"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(bigsnpr)</a></code></pre></div>
<pre><code>## Loading required package: bigstatsr</code></pre>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" title="1"><span class="co"># Read from bed/bim/fam, it generates .bk and .rds files.</span></a>
<a class="sourceLine" id="cb3-2" title="2"><span class="kw"><a href="../reference/snp_readBed.html">snp_readBed</a></span>(<span class="st">"tmp-data/public-data.bed"</span>)</a></code></pre></div>
<pre><code>## [1] "C:\\Users\\au639593\\Desktop\\bigsnpr\\tmp-data\\public-data.rds"</code></pre>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" title="1"><span class="co"># Attach the "bigSNP" object in R session</span></a>
<a class="sourceLine" id="cb5-2" title="2">obj.bigSNP &lt;-<span class="st"> </span><span class="kw"><a href="../reference/snp_attach.html">snp_attach</a></span>(<span class="st">"tmp-data/public-data.rds"</span>)</a>
<a class="sourceLine" id="cb5-3" title="3"><span class="co"># See how the file looks like</span></a>
<a class="sourceLine" id="cb5-4" title="4"><span class="kw"><a href="https://rdrr.io/r/utils/str.html">str</a></span>(obj.bigSNP, <span class="dt">max.level =</span> <span class="dv">2</span>, <span class="dt">strict.width =</span> <span class="st">"cut"</span>)</a></code></pre></div>
<pre><code>## List of 3
##  $ genotypes:Reference class 'FBM.code256' [package "bigstatsr"] with 15 ..
##   ..and 26 methods, of which 12 are  possibly relevant:
##   ..  add_columns, as.FBM, bm, bm.desc, check_dimensions,
##   ..  check_write_permissions, copy#envRefClass, initialize,
##   ..  initialize#FBM, save, show#envRefClass, show#FBM
##  $ fam      :'data.frame':   559 obs. of  6 variables:
##   ..$ family.ID  : chr [1:559] "EUR_GBR" "EUR_GBR" "EUR_GBR" "EUR_GBR" ...
##   ..$ sample.ID  : chr [1:559] "HG00096" "HG00097" "HG00099" "HG00100" ...
##   ..$ paternal.ID: int [1:559] 0 0 0 0 0 0 0 0 0 0 ...
##   ..$ maternal.ID: int [1:559] 0 0 0 0 0 0 0 0 0 0 ...
##   ..$ sex        : int [1:559] 1 2 2 2 1 2 1 2 2 1 ...
##   ..$ affection  : int [1:559] 1 2 1 1 1 1 2 1 2 1 ...
##  $ map      :'data.frame':   130816 obs. of  6 variables:
##   ..$ chromosome  : int [1:130816] 2 2 2 2 2 2 2 2 2 2 ...
##   ..$ marker.ID   : chr [1:130816] "rs13400442" "rs7594567" "rs7597758" "..
##   ..$ genetic.dist: int [1:130816] 0 0 0 0 0 0 0 0 0 0 ...
##   ..$ physical.pos: int [1:130816] 18506 21833 22398 28228 32003 32005 36..
##   ..$ allele1     : chr [1:130816] "C" "G" "T" "A" ...
##   ..$ allele2     : chr [1:130816] "T" "C" "C" "G" ...
##  - attr(*, "class")= chr "bigSNP"</code></pre>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" title="1"><span class="co"># Get aliases for useful slots</span></a>
<a class="sourceLine" id="cb7-2" title="2">G   &lt;-<span class="st"> </span>obj.bigSNP<span class="op">$</span>genotypes</a>
<a class="sourceLine" id="cb7-3" title="3">CHR &lt;-<span class="st"> </span>obj.bigSNP<span class="op">$</span>map<span class="op">$</span>chromosome</a>
<a class="sourceLine" id="cb7-4" title="4">POS &lt;-<span class="st"> </span>obj.bigSNP<span class="op">$</span>map<span class="op">$</span>physical.pos</a>
<a class="sourceLine" id="cb7-5" title="5">y   &lt;-<span class="st"> </span>obj.bigSNP<span class="op">$</span>fam<span class="op">$</span>affection <span class="op">-</span><span class="st"> </span><span class="dv">1</span></a>
<a class="sourceLine" id="cb7-6" title="6">NCORES &lt;-<span class="st"> </span><span class="kw">nb_cores</span>()</a>
<a class="sourceLine" id="cb7-7" title="7"><span class="co"># Read external summary statistics</span></a>
<a class="sourceLine" id="cb7-8" title="8">sumstats &lt;-<span class="st"> </span>bigreadr<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/bigreadr/man/fread2.html">fread2</a></span>(<span class="st">"tmp-data/public-data-sumstats.txt"</span>)</a>
<a class="sourceLine" id="cb7-9" title="9"><span class="kw"><a href="https://rdrr.io/r/utils/str.html">str</a></span>(sumstats)</a></code></pre></div>
<pre><code>## 'data.frame':    130816 obs. of  10 variables:
##  $ chromosome  : int  2 2 2 2 2 2 2 2 2 2 ...
##  $ marker.ID   : chr  "rs13400442" "rs7594567" "rs7597758" "rs13383216" ...
##  $ physical.pos: int  18506 21833 22398 28228 32003 32005 36787 55237 56916 61687 ...
##  $ allele1     : chr  "C" "G" "T" "A" ...
##  $ allele2     : chr  "T" "C" "C" "G" ...
##  $ beta        : num  -0.073 0.0439 -0.3325 -0.5445 -0.4881 ...
##  $ beta_se     : num  0.277 0.248 0.192 0.247 0.242 ...
##  $ n_case      : int  157 157 157 157 157 157 157 157 157 157 ...
##  $ n_control   : int  402 402 402 402 402 402 402 402 402 402 ...
##  $ p           : num  0.7925 0.8593 0.0846 0.028 0.0439 ...</code></pre>
<p>We split genotype data using part of the data to learn parameters of stacking and another part of the data to evaluate statistical properties of polygenic risk score such as AUC. Here we consider that there are 400 individuals in the training set.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" title="1"><span class="kw"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span>(<span class="dv">1</span>)</a>
<a class="sourceLine" id="cb9-2" title="2">ind.val &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/sample.html">sample</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span>(G), <span class="dv">400</span>)</a>
<a class="sourceLine" id="cb9-3" title="3">ind.test &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/sets.html">setdiff</a></span>(<span class="kw">rows_along</span>(G), ind.val)</a></code></pre></div>
</div>
<div id="matching-variants-between-genotype-data-and-summary-statistics" class="section level2">
<h2 class="hasAnchor">
<a href="#matching-variants-between-genotype-data-and-summary-statistics" class="anchor"></a>Matching variants between genotype data and summary statistics</h2>
<p>To match variants contained in genotype data and summary statistics, the variables <code>"chr"</code> (chromosome number), <code>"pos"</code> (genetic position), <code>"a0"</code> (reference allele) and <code>"a1"</code> (derived allele) should be available in the summary statistics and in the genotype data. These 4 variables are used to match variants between the two data frames.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb10-1" title="1">sumstats<span class="op">$</span>n_eff &lt;-<span class="st"> </span><span class="dv">4</span> <span class="op">/</span><span class="st"> </span>(<span class="dv">1</span> <span class="op">/</span><span class="st"> </span>sumstats<span class="op">$</span>n_case <span class="op">+</span><span class="st"> </span><span class="dv">1</span> <span class="op">/</span><span class="st"> </span>sumstats<span class="op">$</span>n_control)</a>
<a class="sourceLine" id="cb10-2" title="2">sumstats<span class="op">$</span>n_case &lt;-<span class="st"> </span>sumstats<span class="op">$</span>n_control &lt;-<span class="st"> </span><span class="ot">NULL</span></a>
<a class="sourceLine" id="cb10-3" title="3"><span class="kw"><a href="https://rdrr.io/r/base/names.html">names</a></span>(sumstats) &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"chr"</span>, <span class="st">"rsid"</span>, <span class="st">"pos"</span>, <span class="st">"a0"</span>, <span class="st">"a1"</span>, <span class="st">"beta"</span>, <span class="st">"beta_se"</span>, <span class="st">"p"</span>, <span class="st">"n_eff"</span>)</a>
<a class="sourceLine" id="cb10-4" title="4">map &lt;-<span class="st"> </span>obj.bigSNP<span class="op">$</span>map[<span class="op">-</span>(<span class="dv">2</span><span class="op">:</span><span class="dv">3</span>)]</a>
<a class="sourceLine" id="cb10-5" title="5"><span class="kw"><a href="https://rdrr.io/r/base/names.html">names</a></span>(map) &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"chr"</span>, <span class="st">"pos"</span>, <span class="st">"a0"</span>, <span class="st">"a1"</span>)</a>
<a class="sourceLine" id="cb10-6" title="6">info_snp &lt;-<span class="st"> </span><span class="kw"><a href="../reference/snp_match.html">snp_match</a></span>(sumstats, map)</a></code></pre></div>
<pre><code>## 130,816 variants to be matched.</code></pre>
<pre><code>## 18,932 ambiguous SNPs have been removed.</code></pre>
<pre><code>## 111,884 variants have been matched; 0 were flipped and 0 were reversed.</code></pre>
<p>If no or few variants are actually flipped, you might want to disable the strand flipping option. Here, these are simulated data so all variants use the same strand and the same reference.</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb14-1" title="1">info_snp &lt;-<span class="st"> </span><span class="kw"><a href="../reference/snp_match.html">snp_match</a></span>(sumstats, map, <span class="dt">strand_flip =</span> <span class="ot">FALSE</span>)</a></code></pre></div>
<pre><code>## 130,816 variants to be matched.</code></pre>
<pre><code>## 130,816 variants have been matched; 0 were flipped and 0 were reversed.</code></pre>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb17-1" title="1">df_beta &lt;-<span class="st"> </span>sumstats[<span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"beta"</span>, <span class="st">"beta_se"</span>, <span class="st">"n_eff"</span>)]</a></code></pre></div>
</div>
<div id="computing-ldpred2-scores-for-one-chromosome" class="section level2">
<h2 class="hasAnchor">
<a href="#computing-ldpred2-scores-for-one-chromosome" class="anchor"></a>Computing LDpred2 scores for one chromosome</h2>
<div id="correlation" class="section level3">
<h3 class="hasAnchor">
<a href="#correlation" class="anchor"></a>Correlation</h3>
<p>First, you need to compute correlations between variants. We recommend to use a window size of 3 cM (see ref).</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb18-1" title="1">POS2 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/snp_asGeneticPos.html">snp_asGeneticPos</a></span>(CHR, POS, <span class="dt">dir =</span> <span class="st">"tmp-data"</span>, <span class="dt">ncores =</span> <span class="dv">3</span>)</a></code></pre></div>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb19-1" title="1">ind.chr &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/which.html">which</a></span>(info_snp<span class="op">$</span>chr <span class="op">==</span><span class="st"> </span><span class="dv">2</span>)         <span class="co">## indices in info_snp</span></a>
<a class="sourceLine" id="cb19-2" title="2">ind.chr2 &lt;-<span class="st"> </span>info_snp<span class="op">$</span><span class="st">`</span><span class="dt">_NUM_ID_</span><span class="st">`</span>[ind.chr]    <span class="co">## indices in G</span></a>
<a class="sourceLine" id="cb19-3" title="3">corr &lt;-<span class="st"> </span><span class="kw"><a href="../reference/snp_cor.html">snp_cor</a></span>(G, <span class="dt">ind.col =</span> ind.chr2, <span class="dt">ncores =</span> NCORES, </a>
<a class="sourceLine" id="cb19-4" title="4">                <span class="dt">infos.pos =</span> POS2[ind.chr2], <span class="dt">size =</span> <span class="dv">3</span> <span class="op">/</span><span class="st"> </span><span class="dv">1000</span>)</a></code></pre></div>
</div>
<div id="infinitesimal-model" class="section level3">
<h3 class="hasAnchor">
<a href="#infinitesimal-model" class="anchor"></a>Infinitesimal model</h3>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb20-1" title="1">(ldsc &lt;-<span class="st"> </span><span class="kw"><a href="../reference/snp_ldsc.html">snp_ldsc2</a></span>(corr, df_beta[ind.chr, ], <span class="dt">blocks =</span> <span class="dv">20</span>, <span class="dt">intercept =</span> <span class="dv">1</span>))</a></code></pre></div>
<pre><code>##      int   int_se       h2    h2_se 
## 1.000000 0.000000 0.114605 0.120726</code></pre>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb22-1" title="1">beta_inf &lt;-<span class="st"> </span><span class="kw"><a href="../reference/LDpred2.html">snp_ldpred2_inf</a></span>(corr, df_beta[ind.chr, ], <span class="dt">h2 =</span> ldsc[[<span class="st">"h2"</span>]])</a></code></pre></div>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb23-1" title="1">pred_inf &lt;-<span class="st"> </span><span class="kw">big_prodVec</span>(G, beta_inf, <span class="dt">ind.row =</span> ind.test, <span class="dt">ind.col =</span> ind.chr2)</a>
<a class="sourceLine" id="cb23-2" title="2"><span class="kw">AUC</span>(pred_inf, y[ind.test])</a></code></pre></div>
<pre><code>## [1] 0.6637168</code></pre>
</div>
<div id="grid-of-models" class="section level3">
<h3 class="hasAnchor">
<a href="#grid-of-models" class="anchor"></a>Grid of models</h3>
<p>In practice, we recommend to test multiple values for h2 and p. </p>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb25-1" title="1">(h2_seq &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/Round.html">round</a></span>(ldsc[[<span class="st">"h2"</span>]] <span class="op">+</span><span class="st"> </span><span class="dv">-2</span><span class="op">:</span><span class="dv">2</span> <span class="op">*</span><span class="st"> </span>ldsc[[<span class="st">"h2_se"</span>]], <span class="dv">3</span>))</a></code></pre></div>
<pre><code>## [1] -0.127 -0.006  0.115  0.235  0.356</code></pre>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb27-1" title="1">(p_seq &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/Round.html">signif</a></span>(<span class="kw"><a href="../reference/seq_log.html">seq_log</a></span>(<span class="fl">1e-5</span>, <span class="dv">1</span>, <span class="dt">length.out =</span> <span class="dv">21</span>), <span class="dv">2</span>))</a></code></pre></div>
<pre><code>##  [1] 1.0e-05 1.8e-05 3.2e-05 5.6e-05 1.0e-04 1.8e-04 3.2e-04 5.6e-04
##  [9] 1.0e-03 1.8e-03 3.2e-03 5.6e-03 1.0e-02 1.8e-02 3.2e-02 5.6e-02
## [17] 1.0e-01 1.8e-01 3.2e-01 5.6e-01 1.0e+00</code></pre>
<p>Because the data is so small here, heritability values are off; let us just use one (otherwise, you can use <code>h2_seq</code>).</p>
<div class="sourceCode" id="cb29"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb29-1" title="1">(params &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/expand.grid.html">expand.grid</a></span>(<span class="dt">p =</span> p_seq, <span class="dt">h2 =</span> ldsc[[<span class="st">"h2"</span>]], <span class="dt">sparse =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="ot">FALSE</span>, <span class="ot">TRUE</span>)))</a></code></pre></div>
<pre><code>##          p       h2 sparse
## 1  1.0e-05 0.114605  FALSE
## 2  1.8e-05 0.114605  FALSE
## 3  3.2e-05 0.114605  FALSE
## 4  5.6e-05 0.114605  FALSE
## 5  1.0e-04 0.114605  FALSE
## 6  1.8e-04 0.114605  FALSE
## 7  3.2e-04 0.114605  FALSE
## 8  5.6e-04 0.114605  FALSE
## 9  1.0e-03 0.114605  FALSE
## 10 1.8e-03 0.114605  FALSE
## 11 3.2e-03 0.114605  FALSE
## 12 5.6e-03 0.114605  FALSE
## 13 1.0e-02 0.114605  FALSE
## 14 1.8e-02 0.114605  FALSE
## 15 3.2e-02 0.114605  FALSE
## 16 5.6e-02 0.114605  FALSE
## 17 1.0e-01 0.114605  FALSE
## 18 1.8e-01 0.114605  FALSE
## 19 3.2e-01 0.114605  FALSE
## 20 5.6e-01 0.114605  FALSE
## 21 1.0e+00 0.114605  FALSE
## 22 1.0e-05 0.114605   TRUE
## 23 1.8e-05 0.114605   TRUE
## 24 3.2e-05 0.114605   TRUE
## 25 5.6e-05 0.114605   TRUE
## 26 1.0e-04 0.114605   TRUE
## 27 1.8e-04 0.114605   TRUE
## 28 3.2e-04 0.114605   TRUE
## 29 5.6e-04 0.114605   TRUE
## 30 1.0e-03 0.114605   TRUE
## 31 1.8e-03 0.114605   TRUE
## 32 3.2e-03 0.114605   TRUE
## 33 5.6e-03 0.114605   TRUE
## 34 1.0e-02 0.114605   TRUE
## 35 1.8e-02 0.114605   TRUE
## 36 3.2e-02 0.114605   TRUE
## 37 5.6e-02 0.114605   TRUE
## 38 1.0e-01 0.114605   TRUE
## 39 1.8e-01 0.114605   TRUE
## 40 3.2e-01 0.114605   TRUE
## 41 5.6e-01 0.114605   TRUE
## 42 1.0e+00 0.114605   TRUE</code></pre>
<div class="sourceCode" id="cb31"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb31-1" title="1">beta_grid &lt;-<span class="st"> </span><span class="kw"><a href="../reference/LDpred2.html">snp_ldpred2_grid</a></span>(corr, df_beta[ind.chr, ], params, <span class="dt">ncores =</span> NCORES)</a></code></pre></div>
<div class="sourceCode" id="cb32"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb32-1" title="1">pred_grid &lt;-<span class="st"> </span><span class="kw">big_prodMat</span>(G, beta_grid, <span class="dt">ind.col =</span> ind.chr2)</a>
<a class="sourceLine" id="cb32-2" title="2">params<span class="op">$</span>auc &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/apply.html">apply</a></span>(pred_grid[ind.val, ], <span class="dv">2</span>, AUC, <span class="dt">target =</span> y[ind.val])</a></code></pre></div>
<div class="sourceCode" id="cb33"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb33-1" title="1"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(ggplot2)</a>
<a class="sourceLine" id="cb33-2" title="2"><span class="kw">ggplot</span>(params, <span class="kw">aes</span>(<span class="dt">x =</span> p, <span class="dt">y =</span> auc, <span class="dt">color =</span> <span class="kw"><a href="https://rdrr.io/r/base/factor.html">as.factor</a></span>(h2))) <span class="op">+</span></a>
<a class="sourceLine" id="cb33-3" title="3"><span class="st">  </span><span class="kw">theme_bigstatsr</span>() <span class="op">+</span></a>
<a class="sourceLine" id="cb33-4" title="4"><span class="st">  </span><span class="kw">geom_point</span>() <span class="op">+</span></a>
<a class="sourceLine" id="cb33-5" title="5"><span class="st">  </span><span class="kw">geom_line</span>() <span class="op">+</span></a>
<a class="sourceLine" id="cb33-6" title="6"><span class="st">  </span><span class="kw">scale_x_log10</span>(<span class="dt">breaks =</span> <span class="dv">10</span><span class="op">^</span>(<span class="op">-</span><span class="dv">5</span><span class="op">:</span><span class="dv">0</span>), <span class="dt">minor_breaks =</span> params<span class="op">$</span>p) <span class="op">+</span></a>
<a class="sourceLine" id="cb33-7" title="7"><span class="st">  </span><span class="kw">facet_wrap</span>(<span class="op">~</span><span class="st"> </span>sparse, <span class="dt">labeller =</span> label_both) <span class="op">+</span></a>
<a class="sourceLine" id="cb33-8" title="8"><span class="st">  </span><span class="kw">labs</span>(<span class="dt">y =</span> <span class="st">"AUC"</span>, <span class="dt">color =</span> <span class="st">"h2"</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb33-9" title="9"><span class="st">  </span><span class="kw">theme</span>(<span class="dt">legend.position =</span> <span class="st">"top"</span>)</a></code></pre></div>
<p><img src="LDpred2_files/figure-html/unnamed-chunk-16-1.png" width="90%" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb34"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb34-1" title="1"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(dplyr)</a></code></pre></div>
<pre><code>## 
## Attaching package: 'dplyr'</code></pre>
<pre><code>## The following objects are masked from 'package:stats':
## 
##     filter, lag</code></pre>
<pre><code>## The following objects are masked from 'package:base':
## 
##     intersect, setdiff, setequal, union</code></pre>
<div class="sourceCode" id="cb38"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb38-1" title="1">params <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb38-2" title="2"><span class="st">  </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span>(<span class="dt">sparsity =</span> <span class="kw"><a href="https://rdrr.io/r/base/colSums.html">colMeans</a></span>(beta_grid <span class="op">==</span><span class="st"> </span><span class="dv">0</span>)) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb38-3" title="3"><span class="st">  </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/arrange.html">arrange</a></span>(<span class="kw"><a href="https://dplyr.tidyverse.org/reference/desc.html">desc</a></span>(auc)) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb38-4" title="4"><span class="st">  </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/mutate_all.html">mutate_at</a></span>(<span class="dv">4</span><span class="op">:</span><span class="dv">5</span>, round, <span class="dt">digits =</span> <span class="dv">3</span>) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb38-5" title="5"><span class="st">  </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/slice.html">slice</a></span>(<span class="dv">1</span><span class="op">:</span><span class="dv">10</span>)</a></code></pre></div>
<pre><code>##          p       h2 sparse   auc sparsity
## 1  0.00100 0.114605   TRUE 0.654    0.587
## 2  0.00180 0.114605  FALSE 0.654    0.000
## 3  0.00100 0.114605  FALSE 0.653    0.000
## 4  0.00056 0.114605  FALSE 0.653    0.000
## 5  0.00180 0.114605   TRUE 0.653    0.570
## 6  0.00320 0.114605   TRUE 0.652    0.561
## 7  0.00560 0.114605   TRUE 0.651    0.565
## 8  0.00320 0.114605  FALSE 0.650    0.000
## 9  0.01000 0.114605   TRUE 0.650    0.563
## 10 0.01800 0.114605   TRUE 0.650    0.565</code></pre>
<p>You can then choose the best one according to your preferred criterion (e.g. max AUC) and evaluate it in the test set.</p>
<div class="sourceCode" id="cb40"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb40-1" title="1">ind_max &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/which.min.html">which.max</a></span>(params<span class="op">$</span>auc)</a>
<a class="sourceLine" id="cb40-2" title="2">best_beta &lt;-<span class="st"> </span>beta_grid[, ind_max]</a>
<a class="sourceLine" id="cb40-3" title="3">best_pred &lt;-<span class="st"> </span>pred_grid[ind.test, ind_max]</a>
<a class="sourceLine" id="cb40-4" title="4"><span class="kw">AUC</span>(best_pred, y[ind.test])</a></code></pre></div>
<pre><code>## [1] 0.6696806</code></pre>
</div>
<div id="automatic-model" class="section level3">
<h3 class="hasAnchor">
<a href="#automatic-model" class="anchor"></a>Automatic model</h3>
<div class="sourceCode" id="cb42"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb42-1" title="1">auto &lt;-<span class="st"> </span><span class="kw"><a href="../reference/LDpred2.html">snp_ldpred2_auto</a></span>(corr, df_beta[ind.chr, ], <span class="dt">h2_init =</span> ldsc[[<span class="st">"h2"</span>]])</a>
<a class="sourceLine" id="cb42-2" title="2"><span class="kw"><a href="https://rdrr.io/r/utils/str.html">str</a></span>(auto)</a></code></pre></div>
<pre><code>## List of 1
##  $ :List of 5
##   ..$ beta_est   : num [1:52556] -4.48e-07 2.70e-07 -2.04e-06 -3.35e-06 -3.00e-06 ...
##   ..$ p_est      : num 0.287
##   ..$ h2_est     : num 0.000716
##   ..$ path_p_est : num [1:2500] 0.0995 0.0986 0.0998 0.1006 0.1005 ...
##   ..$ path_h2_est: num [1:2500] 0.00146 0.00147 0.00145 0.00147 0.00148 ...</code></pre>
<p>You should verify if the algorithm “converged”. This is not the case here, which is probably because the data is so small.</p>
<div class="sourceCode" id="cb44"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb44-1" title="1"><span class="kw">plot_grid</span>(</a>
<a class="sourceLine" id="cb44-2" title="2">  <span class="kw">qplot</span>(<span class="dt">y =</span> auto[[<span class="dv">1</span>]]<span class="op">$</span>path_p_est) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb44-3" title="3"><span class="st">    </span><span class="kw">theme_bigstatsr</span>() <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb44-4" title="4"><span class="st">    </span><span class="kw">geom_hline</span>(<span class="dt">yintercept =</span> auto[[<span class="dv">1</span>]]<span class="op">$</span>p_est, <span class="dt">col =</span> <span class="st">"blue"</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb44-5" title="5"><span class="st">    </span><span class="kw">scale_y_log10</span>() <span class="op">+</span></a>
<a class="sourceLine" id="cb44-6" title="6"><span class="st">    </span><span class="kw">labs</span>(<span class="dt">y =</span> <span class="st">"p"</span>),</a>
<a class="sourceLine" id="cb44-7" title="7">  <span class="kw">qplot</span>(<span class="dt">y =</span> auto[[<span class="dv">1</span>]]<span class="op">$</span>path_h2_est) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb44-8" title="8"><span class="st">    </span><span class="kw">theme_bigstatsr</span>() <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb44-9" title="9"><span class="st">    </span><span class="kw">geom_hline</span>(<span class="dt">yintercept =</span> auto[[<span class="dv">1</span>]]<span class="op">$</span>h2_est, <span class="dt">col =</span> <span class="st">"blue"</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb44-10" title="10"><span class="st">    </span><span class="kw">labs</span>(<span class="dt">y =</span> <span class="st">"h2"</span>),</a>
<a class="sourceLine" id="cb44-11" title="11">  <span class="dt">ncol =</span> <span class="dv">1</span>, <span class="dt">align =</span> <span class="st">"hv"</span></a>
<a class="sourceLine" id="cb44-12" title="12">)</a></code></pre></div>
<p><img src="LDpred2_files/figure-html/unnamed-chunk-20-1.png" width="700" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb45"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb45-1" title="1">beta_auto &lt;-<span class="st"> </span>auto[[<span class="dv">1</span>]]<span class="op">$</span>beta_est</a>
<a class="sourceLine" id="cb45-2" title="2">pred_auto &lt;-<span class="st"> </span><span class="kw">big_prodVec</span>(G, beta_auto, <span class="dt">ind.row =</span> ind.test, <span class="dt">ind.col =</span> ind.chr2)</a>
<a class="sourceLine" id="cb45-3" title="3"><span class="kw">AUC</span>(pred_auto, y[ind.test])</a></code></pre></div>
<pre><code>## [1] 0.6631397</code></pre>
</div>
</div>
<div id="conclusion" class="section level2">
<h2 class="hasAnchor">
<a href="#conclusion" class="anchor"></a>Conclusion</h2>
<p>We have seen how to run 3 versions of LDpred2 (“-inf”, “-grid” and “-auto”) for one chromosome. You should do this for each chromosome and combine results.</p>
<div id="reference" class="section level3">
<h3 class="hasAnchor">
<a href="#reference" class="anchor"></a>Reference</h3>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">

        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#downloading-genotype-data-and-summary-statistics">Downloading genotype data and summary statistics</a></li>
      <li><a href="#matching-variants-between-genotype-data-and-summary-statistics">Matching variants between genotype data and summary statistics</a></li>
      <li><a href="#computing-ldpred2-scores-for-one-chromosome">Computing LDpred2 scores for one chromosome</a></li>
      <li><a href="#conclusion">Conclusion</a></li>
      </ul>
</div>
      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by Florian Privé.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
