<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Comparison with PLINK pruning • bigsnpr</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootswatch/3.3.7/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script><script src="../pkgdown.js"></script><!-- mathjax --><script src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-vignette">
      <header><div class="navbar navbar-inverse navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">bigsnpr</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">Home</a>
</li>
<li>
  <a href="../reference/index.html">Manual</a>
</li>
<li>
  <a href="../articles/index.html">Vignettes</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://privefl.github.io/about.html">
    <span class="fa fa-user"></span>
     
    About
  </a>
</li>
<li>
  <a href="https://github.com/privefl/bigsnpr">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9">
    <div class="page-header toc-ignore">
      <h1>Comparison with PLINK pruning</h1>
            
          </div>

    
    
<div class="contents">
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">require</span>(bigsnpr)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">test &lt;-<span class="st"> </span><span class="kw"><a href="../reference/readplink.html">snp_readExample</a></span>()
<span class="kw">str</span>(test)</code></pre></div>
<pre><code>## List of 5
##  $ genotypes  :Formal class 'big.matrix' [package "bigmemory"] with 1 slot
##   .. ..@ address:&lt;externalptr&gt; 
##  $ fam        :'data.frame': 517 obs. of  6 variables:
##   ..$ family.ID  : chr [1:517] "POP1" "POP1" "POP1" "POP1" ...
##   ..$ sample.ID  : chr [1:517] "IND0" "IND1" "IND2" "IND3" ...
##   ..$ paternal.ID: int [1:517] 0 0 0 0 0 0 0 0 0 0 ...
##   ..$ maternal.ID: int [1:517] 0 0 0 0 0 0 0 0 0 0 ...
##   ..$ sex        : int [1:517] 0 0 0 0 0 0 0 0 0 0 ...
##   ..$ affection  : int [1:517] -9 -9 -9 -9 -9 -9 -9 -9 -9 -9 ...
##  $ map        :'data.frame': 4542 obs. of  6 variables:
##   ..$ chromosome  : int [1:4542] 1 1 1 1 1 1 1 1 1 1 ...
##   ..$ marker.ID   : chr [1:4542] "SNP0" "SNP1" "SNP2" "SNP3" ...
##   ..$ genetic.dist: int [1:4542] 0 0 0 0 0 0 0 0 0 0 ...
##   ..$ physical.pos: int [1:4542] 0 1000 2000 3000 4000 5000 6000 7000 8000 9000 ...
##   ..$ allele1     : chr [1:4542] "A" "T" "T" "T" ...
##   ..$ allele2     : chr [1:4542] "T" "A" "A" "A" ...
##  $ backingfile: chr "test_doc"
##  $ backingpath: chr "/home/privef/Bureau/thesis/bigsnpr/docs/articles/backingfiles"
##  - attr(*, "class")= chr "bigSNP"</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">bedfile &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">"extdata"</span>, <span class="st">"example.bed"</span>, <span class="dt">package =</span> <span class="st">"bigsnpr"</span>)
path &lt;-<span class="st"> </span>tools::<span class="kw">file_path_sans_ext</span>(bedfile)
size &lt;-<span class="st"> </span><span class="dv">50</span>
step &lt;-<span class="st"> </span><span class="dv">1</span>
r2.thr &lt;-<span class="st"> </span><span class="fl">0.05</span></code></pre></div>
<p>If using</p>
<pre><code>system(sprintf(
  "../../plink-1.07-x86_64/plink --bfile %s --indep-pairwise %s %s %s --noweb",
  path, size, step, r2.thr
))</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">logfile &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">"extdata"</span>, <span class="st">"plink.log"</span>, <span class="dt">package =</span> <span class="st">"bigsnpr"</span>) 
<span class="kw">writeLines</span>(<span class="kw">readLines</span>(logfile))</code></pre></div>
<pre><code>## 
## @----------------------------------------------------------@
## |        PLINK!       |     v1.07      |   10/Aug/2009     |
## |----------------------------------------------------------|
## |  (C) 2009 Shaun Purcell, GNU General Public License, v2  |
## |----------------------------------------------------------|
## |  For documentation, citation &amp; bug-report instructions:  |
## |        http://pngu.mgh.harvard.edu/purcell/plink/        |
## @----------------------------------------------------------@
## 
## Skipping web check... [ --noweb ] 
## Writing this text to log file [ plink.log ]
## Analysis started: Sat Dec 10 21:56:34 2016
## 
## Options in effect:
##  --bfile /home/privef/Bureau/bigsnpr/inst/extdata/example
##  --indep-pairwise 50 1 0.05
##  --noweb
## 
## Reading map (extended format) from [ /home/privef/Bureau/bigsnpr/inst/extdata/example.bim ] 
## 4542 markers to be included from [ /home/privef/Bureau/bigsnpr/inst/extdata/example.bim ]
## Reading pedigree information from [ /home/privef/Bureau/bigsnpr/inst/extdata/example.fam ] 
## 517 individuals read from [ /home/privef/Bureau/bigsnpr/inst/extdata/example.fam ] 
## 0 individuals with nonmissing phenotypes
## Assuming a disease phenotype (1=unaff, 2=aff, 0=miss)
## Missing phenotype value is also -9
## 0 cases, 0 controls and 517 missing
## 0 males, 0 females, and 517 of unspecified sex
## Warning, found 517 individuals with ambiguous sex codes
## These individuals will be set to missing ( or use --allow-no-sex )
## Writing list of these individuals to [ plink.nosex ]
## Reading genotype bitfile from [ /home/privef/Bureau/bigsnpr/inst/extdata/example.bed ] 
## Detected that binary PED file is v1.00 SNP-major mode
## Before frequency and genotyping pruning, there are 4542 SNPs
## 517 founders and 0 non-founders found
## Total genotyping rate in remaining individuals is 1
## 0 SNPs failed missingness test ( GENO &gt; 1 )
## 0 SNPs failed frequency test ( MAF &lt; 0 )
## After frequency and genotyping pruning, there are 4542 SNPs
## After filtering, 0 cases, 0 controls and 517 missing
## After filtering, 0 males, 0 females, and 517 of unspecified sex
## Performing LD-based pruning...
## Writing pruned-in SNPs to [ plink.prune.in ]
## Writing pruned-out SNPs to [ plink.prune.out ]
## Scanning from chromosome 1 to 1
## 
## Scan region on chromosome 1 from [ SNP0 ] to [ SNP4541 ]
## For chromosome 1, 159 SNPs pruned out, 4383 remaining
## 
## Analysis finished: Sat Dec 10 21:57:37 2016</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">prune.infile &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">"extdata"</span>, <span class="st">"plink.prune.in"</span>, <span class="dt">package =</span> <span class="st">"bigsnpr"</span>)
snps &lt;-<span class="st"> </span><span class="kw">scan</span>(prune.infile, <span class="dt">what =</span> <span class="st">"character"</span>)
snps.ind &lt;-<span class="st"> </span><span class="kw">match</span>(snps, test$map$marker.ID)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">print</span>(<span class="kw">system.time</span>(
  snps.ind2 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/pruning-clumping.html">snp_pruning</a></span>(test, <span class="dt">size =</span> size, <span class="dt">thr.corr =</span> r2.thr)
))</code></pre></div>
<pre><code>##    user  system elapsed 
##   0.158   0.025   0.179</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">print</span>(<span class="kw">all.equal</span>(snps.ind, snps.ind2))</code></pre></div>
<pre><code>## [1] TRUE</code></pre>
<p>So, I get the same results while being way faster.</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Florian Privé.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  </body>
</html>
