<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Why clumping should be preferred over pruning • bigsnpr</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script><script src="../pkgdown.js"></script><!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-vignette">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">bigsnpr</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Manual</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/pruning-vs-clumping.html">Why clumping should be preferred over pruning</a>
    </li>
    <li>
      <a href="../articles/demo.html">Small demo</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://privefl.github.io/about.html">
    <span class="fa fa-user"></span>
     
    About
  </a>
</li>
<li>
  <a href="https://github.com/privefl/bigsnpr">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9">
    <div class="page-header toc-ignore">
      <h1>Why clumping should be preferred over pruning</h1>
                        <h4 class="author">Florian Privé</h4>
            
            <h4 class="date">November 24, 2016</h4>
          </div>

    
    
<div class="contents">
<p>In this short document, I will show why clumping should be preferred over pruning for single-nucleotide polymorphism (SNP) arrays.</p>
<p>Pruning and clumping are used to keep a subset of SNPs that are nearly uncorrelated with each other. For instance, pruning is used before Principal Component Analysis to avoid capturing too much variance of linkage disequilibrium (LD) regions. Clumping is used to keep only one representative SNP per region of LD. <!-- , for example when reporting p-values of genome-wide association studies (GWAS). --></p>
<div id="simulation" class="section level2">
<h2 class="hasAnchor">
<a href="#simulation" class="anchor"></a>Simulation</h2>
<p>I generate a (toy) SNP array with 500 individuals and 10 SNPs, where each SNP has a squared correlation &gt; 0.2 with their direct neighbors and only with them. Moreover, the SNPs have an increasing MAF.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">gen &lt;-<span class="st"> </span>function(n, m) {
  I &lt;-<span class="st"> </span><span class="dv">1</span>:m
  p &lt;-<span class="st"> </span>I /<span class="st"> </span>(<span class="dv">2</span> *<span class="st"> </span>m +<span class="st"> </span><span class="dv">1</span>)

  mat &lt;-<span class="st"> </span><span class="kw">outer</span>(I, I, <span class="dt">FUN =</span> function(i, j) {
    <span class="dv">1</span> /<span class="st"> </span>(<span class="kw">abs</span>(i -<span class="st"> </span>j) +<span class="st"> </span><span class="dv">1</span>)
  })

  bindata::<span class="kw">rmvbin</span>(n, p, <span class="dt">bincorr =</span> mat) +<span class="st"> </span>
<span class="st">    </span>bindata::<span class="kw">rmvbin</span>(n, p, <span class="dt">bincorr =</span> mat)
}

<span class="kw">set.seed</span>(<span class="dv">1</span>)
X &lt;-<span class="st"> </span><span class="kw">gen</span>(<span class="dv">500</span>, <span class="dv">10</span>)
<span class="kw">print</span>(<span class="kw">head</span>(X, <span class="dv">20</span>))</code></pre></div>
<pre><code>##       [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8] [,9] [,10]
##  [1,]    0    0    0    1    0    0    1    1    1     1
##  [2,]    0    0    0    0    0    0    1    1    1     1
##  [3,]    0    1    1    0    0    0    0    0    0     1
##  [4,]    0    0    0    0    0    0    0    0    1     1
##  [5,]    0    0    0    0    0    0    0    1    1     1
##  [6,]    0    0    0    0    1    1    0    0    1     1
##  [7,]    1    0    1    0    1    1    1    2    2     1
##  [8,]    1    1    1    1    1    1    0    1    1     1
##  [9,]    0    0    0    0    0    1    1    0    1     1
## [10,]    0    0    1    1    1    1    1    1    1     1
## [11,]    0    1    1    0    0    2    2    2    2     1
## [12,]    0    0    0    0    0    0    0    0    0     0
## [13,]    0    0    0    0    0    0    1    0    1     1
## [14,]    0    0    0    0    0    0    0    0    1     1
## [15,]    0    0    0    0    0    0    1    0    0     0
## [16,]    0    0    0    0    0    0    0    1    1     1
## [17,]    0    0    1    1    0    1    0    0    1     1
## [18,]    1    1    2    1    1    2    2    2    2     2
## [19,]    0    1    0    0    0    1    2    1    1     0
## [20,]    0    0    0    0    0    0    1    1    0     0</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">print</span>(<span class="kw">round</span>(<span class="kw">cor</span>(X)^<span class="dv">2</span>, <span class="dv">2</span>)) <span class="co"># squared correlation between SNPs</span></code></pre></div>
<pre><code>##       [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8] [,9] [,10]
##  [1,] 1.00 0.23 0.13 0.08 0.05 0.03 0.02 0.01 0.01  0.00
##  [2,] 0.23 1.00 0.21 0.09 0.05 0.05 0.01 0.01 0.01  0.01
##  [3,] 0.13 0.21 1.00 0.25 0.12 0.10 0.06 0.06 0.02  0.01
##  [4,] 0.08 0.09 0.25 1.00 0.31 0.16 0.07 0.07 0.03  0.04
##  [5,] 0.05 0.05 0.12 0.31 1.00 0.29 0.11 0.10 0.07  0.04
##  [6,] 0.03 0.05 0.10 0.16 0.29 1.00 0.25 0.17 0.08  0.06
##  [7,] 0.02 0.01 0.06 0.07 0.11 0.25 1.00 0.23 0.13  0.07
##  [8,] 0.01 0.01 0.06 0.07 0.10 0.17 0.23 1.00 0.28  0.14
##  [9,] 0.01 0.01 0.02 0.03 0.07 0.08 0.13 0.28 1.00  0.31
## [10,] 0.00 0.01 0.01 0.04 0.04 0.06 0.07 0.14 0.31  1.00</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">print</span>(<span class="kw">colMeans</span>(X) /<span class="st"> </span><span class="dv">2</span>) <span class="co"># MAF of SNPs</span></code></pre></div>
<pre><code>##  [1] 0.043 0.091 0.154 0.177 0.234 0.283 0.338 0.372 0.436 0.450</code></pre>
</div>
<div id="pruning" class="section level2">
<h2 class="hasAnchor">
<a href="#pruning" class="anchor"></a>Pruning</h2>
<p>I then pseudo-convert the SNP matrix in my format <code>BigSNP</code> and use my re-implementation of PLINK pruning on this dataset (similar to <code>--indep-pairwise 10 1 0.2</code>)</p>
</div>
<div id="results-and-conclusion" class="section level2">
<h2 class="hasAnchor">
<a href="#results-and-conclusion" class="anchor"></a>Results and conclusion</h2>
<p>The first SNP is pruned because of its correlation with the second. The second SNP is pruned because of its correlation with the third and so on. In the end, only the last SNP (10th) is kept with the LD pruning procedure of PLINK, which corresponds to less than 18% of the total variance.</p>
<p>On the contrary, a clumping approach would consider the SNPs sorted (in a decreasing order) by a statistic (here the MAF). So the last SNP (10th) would be considered first and the 9th SNP would be pruned. Then the 8th SNP would be considered and the 7th SNP would be pruned and so on. So, the even SNPs would be kept and the odd SNPs would be pruned, which corresponds to 56.7% of the total variance.</p>
<p>So, pruning remove SNPs in a way that can leave regions of the genome with no representative SNP at all.</p>
<!-- ## Acknowledgements  -->
<!-- I thank Michael G.B. Blum and Hugues Aschard, my two thesis supervisors, for thoughtful comments. -->
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#simulation">Simulation</a></li>
      <li><a href="#pruning">Pruning</a></li>
      <li><a href="#results-and-conclusion">Results and conclusion</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Florian Privé.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  </body>
</html>
