<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Computing polygenic scores using Stacked Clumping and Thresholding (SCT) • bigsnpr</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Computing polygenic scores using Stacked Clumping and Thresholding (SCT)">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">bigsnpr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.11.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Manual</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/demo.html">Quick demo</a>
    </li>
    <li>
      <a href="../articles/exo.html">Exercise</a>
    </li>
    <li>
      <a href="../articles/how-to-PCA.html">How to capture Population Structure with PCA</a>
    </li>
    <li>
      <a href="../articles/pruning-vs-clumping.html">Why clumping should be preferred over pruning</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://privefl.github.io/about.html">
    <span class="fa fa-user"></span>
     
    About
  </a>
</li>
<li>
  <a href="https://github.com/privefl/bigsnpr">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Computing polygenic scores using Stacked Clumping and Thresholding (SCT)</h1>
                        <h4 class="author">Florian Privé and Michael Blum</h4>
            
            <h4 class="date">May 23, 2019</h4>
      
      
      <div class="hidden name"><code>SCT.Rmd</code></div>

    </div>

    
    
<p>In this document, we show how to compute polygenic risk scores using Stacked Clumping and Thresholding (SCT).</p>
<div id="downloading-genotype-data-and-summary-statistics" class="section level2">
<h2 class="hasAnchor">
<a href="#downloading-genotype-data-and-summary-statistics" class="anchor"></a>Downloading genotype data and summary statistics</h2>
<p>You can download <a href="https://github.com/privefl/bigsnpr/raw/master/data-raw/public-data.zip">data</a> and unzip files in R. We store those files in a directory called <code>"tmp-data"</code> here.</p>
<p>You can see <a href="https://github.com/privefl/bigsnpr/blob/master/data-raw/public-data.R">there</a> how we generated these data from <a href="https://www.nature.com/articles/nature15393">the 1000 Genomes project</a>.</p>
<p>First, you need to read genotype data from the PLINK files (or BGEN files) as well as the text file containing summary statistics.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Load packages bigsnpr and bigstatsr</span>
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(bigsnpr)</code></pre></div>
<pre><code>## Loading required package: bigstatsr</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Read from bed/bim/fam, it generates .bk and .rds files.</span>
<span class="kw"><a href="../reference/snp_readBed.html">snp_readBed</a></span>(<span class="st">"tmp-data/public-data.bed"</span>)</code></pre></div>
<pre><code>## [1] "/home/privef/Bureau/bigsnpr/tmp-data/public-data.rds"</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Attach the "bigSNP" object in R session</span>
obj.bigSNP &lt;-<span class="st"> </span><span class="kw"><a href="../reference/snp_attach.html">snp_attach</a></span>(<span class="st">"tmp-data/public-data.rds"</span>)
<span class="co"># See how the file looks like</span>
<span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/str">str</a></span>(obj.bigSNP, <span class="dt">max.level =</span> <span class="dv">2</span>, <span class="dt">strict.width =</span> <span class="st">"cut"</span>)</code></pre></div>
<pre><code>## List of 3
##  $ genotypes:Reference class 'FBM.code256' [package "bigstatsr"] with 10 ..
##   ..and 24 methods, of which 10 are  possibly relevant:
##   ..  add_columns, as.FBM, bm, bm.desc, copy#envRefClass, initialize,
##   ..  initialize#FBM, save, show#envRefClass, show#FBM
##  $ fam      :'data.frame':   559 obs. of  6 variables:
##   ..$ family.ID  : chr [1:559] "EUR_GBR" "EUR_GBR" "EUR_GBR" "EUR_GBR" ...
##   ..$ sample.ID  : chr [1:559] "HG00096" "HG00097" "HG00099" "HG00100" ...
##   ..$ paternal.ID: int [1:559] 0 0 0 0 0 0 0 0 0 0 ...
##   ..$ maternal.ID: int [1:559] 0 0 0 0 0 0 0 0 0 0 ...
##   ..$ sex        : int [1:559] 1 2 2 2 1 2 1 2 2 1 ...
##   ..$ affection  : int [1:559] 2 1 1 1 1 1 2 1 1 1 ...
##  $ map      :'data.frame':   131276 obs. of  6 variables:
##   ..$ chromosome  : int [1:131276] 2 2 2 2 2 2 2 2 2 2 ...
##   ..$ marker.ID   : chr [1:131276] "rs13386112" "rs4263140" "rs28446791""..
##   ..$ genetic.dist: int [1:131276] 0 0 0 0 0 0 0 0 0 0 ...
##   ..$ physical.pos: int [1:131276] 21243 29422 30091 36787 51141 53020 61..
##   ..$ allele1     : chr [1:131276] "T" "G" "G" "T" ...
##   ..$ allele2     : chr [1:131276] "C" "A" "C" "C" ...
##  - attr(*, "class")= chr "bigSNP"</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Get aliases for useful slots</span>
G   &lt;-<span class="st"> </span>obj.bigSNP<span class="op">$</span>genotypes
CHR &lt;-<span class="st"> </span>obj.bigSNP<span class="op">$</span>map<span class="op">$</span>chromosome
POS &lt;-<span class="st"> </span>obj.bigSNP<span class="op">$</span>map<span class="op">$</span>physical.pos
y   &lt;-<span class="st"> </span>obj.bigSNP<span class="op">$</span>fam<span class="op">$</span>affection <span class="op">-</span><span class="st"> </span><span class="dv">1</span>
NCORES &lt;-<span class="st"> </span><span class="kw">nb_cores</span>()
<span class="co"># Check some counts for the 10 first variants</span>
<span class="kw">big_counts</span>(G, <span class="dt">ind.col =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">10</span>)</code></pre></div>
<pre><code>##      [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8] [,9] [,10]
## 0     476  376  378  380  481  463  389  476  483   478
## 1      64  143  139  136   71   84  133   71   70    70
## 2      19   40   42   43    7   12   37   12    6    11
## &lt;NA&gt;    0    0    0    0    0    0    0    0    0     0</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Read external summary statistics</span>
sumstats &lt;-<span class="st"> </span>bigreadr<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/bigreadr/topics/fread2">fread2</a></span>(<span class="st">"tmp-data/public-data-sumstats.txt"</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/str">str</a></span>(sumstats)</code></pre></div>
<pre><code>## 'data.frame':    131276 obs. of  7 variables:
##  $ chromosome  : int  2 2 2 2 2 2 2 2 2 2 ...
##  $ marker.ID   : chr  "rs13386112" "rs4263140" "rs28446791" "rs11900053" ...
##  $ physical.pos: int  21243 29422 30091 36787 51141 53020 61687 67535 69022 70325 ...
##  $ allele1     : chr  "T" "G" "G" "T" ...
##  $ allele2     : chr  "C" "A" "C" "C" ...
##  $ beta        : num  0.0121 0.0778 0.083 0.1203 -0.2444 ...
##  $ p           : num  0.961 0.658 0.635 0.491 0.377 ...</code></pre>
<p>We split genotype data using part of the data to learn parameters of stacking and another part of the data to evaluate statistical properties of polygenic risk score such as AUC. Here we consider that there are 400 individuals in the training set.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/Random">set.seed</a></span>(<span class="dv">1</span>)
ind.train &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/sample">sample</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/nrow">nrow</a></span>(G), <span class="dv">400</span>)
ind.test &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/sets">setdiff</a></span>(<span class="kw">rows_along</span>(G), ind.train)</code></pre></div>
</div>
<div id="matching-variants-between-genotype-data-and-summary-statistics" class="section level2">
<h2 class="hasAnchor">
<a href="#matching-variants-between-genotype-data-and-summary-statistics" class="anchor"></a>Matching variants between genotype data and summary statistics</h2>
<p>To match variants contained in genotype data and summary statistics, the variables <code>"chr"</code> (chromosome number), <code>"pos"</code> (genetic position), <code>"a0"</code> (reference allele) and <code>"a1"</code> (derived allele) should be available in the summary statistics and in the genotype data. These 4 variables are used to match variants between the two data frames.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/names">names</a></span>(sumstats) &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"chr"</span>, <span class="st">"rsid"</span>, <span class="st">"pos"</span>, <span class="st">"a0"</span>, <span class="st">"a1"</span>, <span class="st">"beta"</span>, <span class="st">"p"</span>)
map &lt;-<span class="st"> </span>obj.bigSNP<span class="op">$</span>map[,<span class="op">-</span>(<span class="dv">2</span><span class="op">:</span><span class="dv">3</span>)]
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/names">names</a></span>(map) &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"chr"</span>, <span class="st">"pos"</span>, <span class="st">"a0"</span>, <span class="st">"a1"</span>)
info_snp &lt;-<span class="st"> </span><span class="kw"><a href="../reference/snp_match.html">snp_match</a></span>(sumstats, map)</code></pre></div>
<pre><code>## 131,276 variants in summary statistics.</code></pre>
<pre><code>## 18,798 ambiguous SNPs have been removed.</code></pre>
<pre><code>## 112,478 variants have been matched; 0 were flipped and 0 were reversed.</code></pre>
<p>If no or few variants are actually flipped, you might want to disable the strand flipping option. Here, these are simulated data so all variants use the same strand and the same reference.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">info_snp &lt;-<span class="st"> </span><span class="kw"><a href="../reference/snp_match.html">snp_match</a></span>(sumstats, map, <span class="dt">strand_flip =</span> <span class="ot">FALSE</span>)</code></pre></div>
<pre><code>## 131,276 variants in summary statistics.</code></pre>
<pre><code>## 131,276 variants have been matched; 0 were flipped and 0 were reversed.</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">beta &lt;-<span class="st"> </span>info_snp<span class="op">$</span>beta
lpval &lt;-<span class="st"> </span><span class="op">-</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/Log">log10</a></span>(info_snp<span class="op">$</span>p)</code></pre></div>
</div>
<div id="computing-ct-scores-for-a-grid-of-parameters-and-chromosomes" class="section level2">
<h2 class="hasAnchor">
<a href="#computing-ct-scores-for-a-grid-of-parameters-and-chromosomes" class="anchor"></a>Computing C+T scores for a grid of parameters and chromosomes</h2>
<div id="clumping" class="section level3">
<h3 class="hasAnchor">
<a href="#clumping" class="anchor"></a>Clumping</h3>
<p>First, the function <code><a href="../reference/SCT.html">snp_grid_clumping()</a></code> computes sets of variants resulting from the clumping procedure that is applied repeatedly with different values of hyper-parameters (threshold of correlation for clumping, window size, and possibly imputation accuracy threshold). By default, the function uses 28 (7 thresholds of correlation x 4 window sizes) different sets of hyper-parameters for generating sets of variants resulting from clumping.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># The clumping step might take some time to complete</span>
all_keep &lt;-<span class="st"> </span><span class="kw"><a href="../reference/SCT.html">snp_grid_clumping</a></span>(G, CHR, POS, <span class="dt">ind.row =</span> ind.train,
                              <span class="dt">lpS =</span> lpval, <span class="dt">ncores =</span> NCORES)
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/attr">attr</a></span>(all_keep, <span class="st">"grid"</span>)</code></pre></div>
<pre><code>##     size thr.r2 grp.num thr.imp
## 1   5000   0.01       1       1
## 2  10000   0.01       1       1
## 3  20000   0.01       1       1
## 4  50000   0.01       1       1
## 5   1000   0.05       1       1
## 6   2000   0.05       1       1
## 7   4000   0.05       1       1
## 8  10000   0.05       1       1
## 9    500   0.10       1       1
## 10  1000   0.10       1       1
## 11  2000   0.10       1       1
## 12  5000   0.10       1       1
## 13   250   0.20       1       1
## 14   500   0.20       1       1
## 15  1000   0.20       1       1
## 16  2500   0.20       1       1
## 17   100   0.50       1       1
## 18   200   0.50       1       1
## 19   400   0.50       1       1
## 20  1000   0.50       1       1
## 21    62   0.80       1       1
## 22   125   0.80       1       1
## 23   250   0.80       1       1
## 24   625   0.80       1       1
## 25    52   0.95       1       1
## 26   105   0.95       1       1
## 27   210   0.95       1       1
## 28   526   0.95       1       1</code></pre>
</div>
<div id="thresholding" class="section level3">
<h3 class="hasAnchor">
<a href="#thresholding" class="anchor"></a>Thresholding</h3>
<p>Then, for each chromosome, for each set of variants resulting from clumping and for each p-value threshold, the function <code><a href="../reference/SCT.html">snp_grid_PRS()</a></code> computes C+T scores.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">multi_PRS &lt;-<span class="st"> </span><span class="kw"><a href="../reference/SCT.html">snp_grid_PRS</a></span>(G, all_keep, beta, lpval, <span class="dt">ind.row =</span> ind.train,
                          <span class="dt">backingfile =</span> <span class="st">"tmp-data/public-data-scores"</span>, 
                          <span class="dt">n_thr_lpS =</span> <span class="dv">50</span>, <span class="dt">ncores =</span> NCORES)
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/dim">dim</a></span>(multi_PRS)  ## 4200 C+T scores for 400 individuals</code></pre></div>
<pre><code>## [1]  400 4200</code></pre>
</div>
</div>
<div id="stacking-ct-predictions" class="section level2">
<h2 class="hasAnchor">
<a href="#stacking-ct-predictions" class="anchor"></a>Stacking C+T predictions</h2>
<p>A penalized regression is finally used to learn an optimal linear combination of C+T scores.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">final_mod &lt;-<span class="st"> </span><span class="kw"><a href="../reference/SCT.html">snp_grid_stacking</a></span>(multi_PRS, y[ind.train], <span class="dt">ncores =</span> NCORES, <span class="dt">K =</span> <span class="dv">4</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/summary">summary</a></span>(final_mod<span class="op">$</span>mod)</code></pre></div>
<pre><code>## # A tibble: 3 x 6
##    alpha validation_loss intercept beta          nb_var message  
##    &lt;dbl&gt;           &lt;dbl&gt;     &lt;dbl&gt; &lt;list&gt;         &lt;int&gt; &lt;list&gt;   
## 1 0.0001           0.517     -1.40 &lt;dbl [4,060]&gt;   3884 &lt;chr [4]&gt;
## 2 0.01             0.519     -1.58 &lt;dbl [4,060]&gt;    994 &lt;chr [4]&gt;
## 3 1                0.519     -1.50 &lt;dbl [4,060]&gt;    198 &lt;chr [4]&gt;</code></pre>
<p>For options for fitting penalized regressions, see <a href="https://privefl.github.io/bigstatsr/articles/penalized-regressions.html">this vignette</a>.</p>
<p>From stacking C+T scores, we can derive a unique vector of weights and compare effects resulting from stacking to the initial regression coefficients provided as summary statistics.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">new_beta &lt;-<span class="st"> </span>final_mod<span class="op">$</span>beta.G
ind &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/which">which</a></span>(new_beta <span class="op">!=</span><span class="st"> </span><span class="dv">0</span>)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(ggplot2)
<span class="kw"><a href="https://www.rdocumentation.org/packages/ggplot2/topics/ggplot">ggplot</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/data.frame">data.frame</a></span>(<span class="dt">y =</span> new_beta, <span class="dt">x =</span> beta)[ind, ]) <span class="op">+</span>
<span class="st">  </span><span class="kw"><a href="https://www.rdocumentation.org/packages/ggplot2/topics/geom_abline">geom_abline</a></span>(<span class="dt">slope =</span> <span class="dv">1</span>, <span class="dt">intercept =</span> <span class="dv">0</span>, <span class="dt">color =</span> <span class="st">"red"</span>) <span class="op">+</span>
<span class="st">  </span><span class="kw"><a href="https://www.rdocumentation.org/packages/ggplot2/topics/geom_abline">geom_abline</a></span>(<span class="dt">slope =</span> <span class="dv">0</span>, <span class="dt">intercept =</span> <span class="dv">0</span>, <span class="dt">color =</span> <span class="st">"blue"</span>) <span class="op">+</span>
<span class="st">  </span><span class="kw"><a href="https://www.rdocumentation.org/packages/ggplot2/topics/geom_point">geom_point</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/ggplot2/topics/aes">aes</a></span>(x, y), <span class="dt">size =</span> <span class="fl">0.6</span>) <span class="op">+</span>
<span class="st">  </span><span class="kw">theme_bigstatsr</span>() <span class="op">+</span>
<span class="st">  </span><span class="kw"><a href="https://www.rdocumentation.org/packages/ggplot2/topics/labs">labs</a></span>(<span class="dt">x =</span> <span class="st">"Effect sizes from GWAS"</span>, <span class="dt">y =</span> <span class="st">"Non-zero effect sizes from SCT"</span>)</code></pre></div>
<p><img src="SCT_files/figure-html/unnamed-chunk-11-1.png" width="700" style="display: block; margin: auto;"></p>
<p>We can use this vector of variant weights to compute polygenic risk scores on the test set and evaluate the Area Under the Curve (AUC).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pred &lt;-<span class="st"> </span>final_mod<span class="op">$</span>intercept <span class="op">+</span><span class="st"> </span>
<span class="st">  </span><span class="kw">big_prodVec</span>(G, new_beta[ind], <span class="dt">ind.row =</span> ind.test, <span class="dt">ind.col =</span> ind)
<span class="kw">AUCBoot</span>(pred, y[ind.test])</code></pre></div>
<pre><code>##       Mean       2.5%      97.5%         Sd 
## 0.67970973 0.58814437 0.76682221 0.04584201</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/ggplot2/topics/ggplot">ggplot</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/data.frame">data.frame</a></span>(
  <span class="dt">Phenotype =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/factor">factor</a></span>(y[ind.test], <span class="dt">levels =</span> <span class="dv">0</span><span class="op">:</span><span class="dv">1</span>, <span class="dt">labels =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"Control"</span>, <span class="st">"Case"</span>)),
  <span class="dt">Probability =</span> <span class="dv">1</span> <span class="op">/</span><span class="st"> </span>(<span class="dv">1</span> <span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/Log">exp</a></span>(<span class="op">-</span>pred)))) <span class="op">+</span><span class="st"> </span>
<span class="st">  </span><span class="kw">theme_bigstatsr</span>() <span class="op">+</span>
<span class="st">  </span><span class="kw"><a href="https://www.rdocumentation.org/packages/ggplot2/topics/geom_density">geom_density</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/ggplot2/topics/aes">aes</a></span>(Probability, <span class="dt">fill =</span> Phenotype), <span class="dt">alpha =</span> <span class="fl">0.3</span>)</code></pre></div>
<p><img src="SCT_files/figure-html/unnamed-chunk-13-1.png" width="700" style="display: block; margin: auto;"></p>
</div>
<div id="best-ct-predictions" class="section level2">
<h2 class="hasAnchor">
<a href="#best-ct-predictions" class="anchor"></a>Best C+T predictions</h2>
<p>Instead of stacking, an alternative is to choose the best C+T score based on the computed grid. This procedure is appealing when there are not enough individuals to learn the stacking weights.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(tidyverse)
grid2 &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/attr">attr</a></span>(all_keep, <span class="st">"grid"</span>) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">mutate</span>(<span class="dt">thr.lp =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/attr">attr</a></span>(multi_PRS, <span class="st">"grid.lpS.thr"</span>)), <span class="dt">num =</span> <span class="kw">row_number</span>()) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">unnest</span>()
s &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/nrow">nrow</a></span>(grid2)
grid2<span class="op">$</span>auc &lt;-<span class="st"> </span><span class="kw">big_apply</span>(multi_PRS, <span class="dt">a.FUN =</span> <span class="cf">function</span>(X, ind, s, y.train) {
  <span class="co"># Sum over all chromosomes, for the same C+T parameters</span>
  single_PRS &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colSums">rowSums</a></span>(X[, ind <span class="op">+</span><span class="st"> </span>s <span class="op">*</span><span class="st"> </span>(<span class="dv">0</span><span class="op">:</span><span class="dv">2</span>)])  ## replace by 0:21 in real data
  bigstatsr<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/bigstatsr/topics/AUC">AUC</a></span>(single_PRS, y.train)
}, <span class="dt">ind =</span> <span class="dv">1</span><span class="op">:</span>s, <span class="dt">s =</span> s, <span class="dt">y.train =</span> y[ind.train],
<span class="dt">a.combine =</span> <span class="st">'c'</span>, <span class="dt">block.size =</span> <span class="dv">1</span>, <span class="dt">ncores =</span> NCORES)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">max_prs &lt;-<span class="st"> </span>grid2 <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">arrange</span>(<span class="kw">desc</span>(auc)) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">slice</span>(<span class="dv">1</span><span class="op">:</span><span class="dv">10</span>) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/print">print</a></span>() <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">slice</span>(<span class="dv">1</span>)</code></pre></div>
<pre><code>##     size thr.r2 grp.num thr.imp num   thr.lp       auc
## 1     52   0.95       1       1  25 4.187722 0.6871956
## 2    105   0.95       1       1  26 4.187722 0.6871956
## 3    210   0.95       1       1  27 4.187722 0.6871956
## 4    526   0.95       1       1  28 4.187722 0.6871956
## 5     62   0.80       1       1  21 4.187722 0.6716028
## 6    125   0.80       1       1  22 4.187722 0.6716028
## 7    250   0.80       1       1  23 4.187722 0.6716028
## 8    625   0.80       1       1  24 4.187722 0.6716028
## 9   5000   0.01       1       1   1 4.187722 0.6610441
## 10 10000   0.01       1       1   2 4.187722 0.6610441</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ind.keep &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/unlist">unlist</a></span>(<span class="kw">map</span>(all_keep, max_prs<span class="op">$</span>num))
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/sum">sum</a></span>(lpval[ind.keep] <span class="op">&gt;</span><span class="st"> </span>max_prs<span class="op">$</span>thr.lp)</code></pre></div>
<pre><code>## [1] 7</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">AUCBoot</span>(
  <span class="kw"><a href="../reference/snp_PRS.html">snp_PRS</a></span>(G, beta[ind.keep], <span class="dt">ind.test =</span> ind.test, <span class="dt">ind.keep =</span> ind.keep,
          <span class="dt">lpS.keep =</span> lpval[ind.keep], <span class="dt">thr.list =</span> max_prs<span class="op">$</span>thr.lp),
  y[ind.test]
)</code></pre></div>
<pre><code>##       Mean       2.5%      97.5%         Sd 
## 0.62958138 0.53717881 0.72081822 0.04692842</code></pre>
<p>For this example, the best C+T predictions provides an AUC of 63% whereas stacking, which should be preferred, provides an AUC of 68%.</p>
<div id="reference" class="section level3">
<h3 class="hasAnchor">
<a href="#reference" class="anchor"></a>Reference</h3>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#downloading-genotype-data-and-summary-statistics">Downloading genotype data and summary statistics</a></li>
      <li><a href="#matching-variants-between-genotype-data-and-summary-statistics">Matching variants between genotype data and summary statistics</a></li>
      <li><a href="#computing-ct-scores-for-a-grid-of-parameters-and-chromosomes">Computing C+T scores for a grid of parameters and chromosomes</a></li>
      <li><a href="#stacking-ct-predictions">Stacking C+T predictions</a></li>
      <li><a href="#best-ct-predictions">Best C+T predictions</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Florian Privé.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
