<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Imputation, quality control and subset for a &quot;bigSNP&quot;. — impute-qc-sub • bigsnpr</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">


<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script>
<script src="../pkgdown.js"></script>

<!-- mathjax -->
<script src='https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">bigsnpr</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/index.html">Articles</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/privefl/bigsnpr">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

      <div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Imputation, quality control and subset for a &quot;bigSNP&quot;.</h1>
    </div>

    
    <p><code>sub.bigSNP</code>: a function
to get a subset of an object of class <code>bigSNP</code>.</p>

    <p><code>QC</code>: Quality control (filters)
for a <code>bigSNP</code> resulting in a <code>bigSNP</code> of lower dimension.</p>

    <p><code>Impute</code>: Imputation function
for a <code>bigSNP</code>.</p>
    

    <pre><span class='fu'>sub.bigSNP</span>(<span class='no'>x</span>, <span class='kw'>ind.row</span> <span class='kw'>=</span> <span class='fu'>seq</span>(<span class='fu'>nrow</span>(<span class='no'>x</span>$<span class='no'>genotypes</span>)),
  <span class='kw'>ind.col</span> <span class='kw'>=</span> <span class='fu'>seq</span>(<span class='fu'>ncol</span>(<span class='no'>x</span>$<span class='no'>genotypes</span>)), <span class='kw'>backed</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>)

<span class='fu'>QC</span>(<span class='no'>x</span>, <span class='kw'>row.cr.min</span> <span class='kw'>=</span> <span class='fl'>0.95</span>, <span class='kw'>col.cr.min</span> <span class='kw'>=</span> <span class='fl'>0.95</span>, <span class='kw'>hwe.pval</span> <span class='kw'>=</span> <span class='fl'>1e-06</span>,
  <span class='kw'>maf.min</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>rm.sex</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>)

<span class='fu'>Impute</span>(<span class='no'>x</span>, <span class='kw'>ncores</span> <span class='kw'>=</span> <span class='fl'>1</span>, <span class='kw'>verbose</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a> Arguments</h2>
    <dl class="dl-horizontal">
      <dt>x</dt>
      <dd>A <a href='bigSNP-class.html'>bigSNP</a>.</dd>
      <dt>ind.row</dt>
      <dd>Indices of the rows (individuals) to keep.
Negative indices can be used to exclude row indices.
Default: keep them all.</dd>
      <dt>ind.col</dt>
      <dd>Indices of the columns (SNPs) to keep.
Negative indices can be used to exclude column indices.
Default: keep them all.</dd>
      <dt>backed</dt>
      <dd>Should the new <code>bigSNP</code> be filebacked? Default is <code>TRUE</code>.</dd>
      <dt>row.cr.min</dt>
      <dd>Minimun individuals&#39; call rate that is allowed.
Default is 95%.</dd>
      <dt>col.cr.min</dt>
      <dd>Minimum SNPs&#39; call rate that is allowed. Default is 95%.</dd>
      <dt>hwe.pval</dt>
      <dd>Level threshold (allowed type-I error) to test deviations
from Hardy–Weinberg equilibrium (HWE) from controls only. Default is <code>1e-6</code>.</dd>
      <dt>maf.min</dt>
      <dd>Minimum Minor Allele Frequency that is allowed.
Usually, <code>0.01</code> is used. Default only removes SNPs that have a zero MAF.</dd>
      <dt>rm.sex</dt>
      <dd>Keep only SNPs on the first 22 chrosmosomes? Default is <code>FALSE</code>.</dd>
      <dt>ncores</dt>
      <dd>Number of cores used. Default doesn&#39;t use parallelism.</dd>
      <dt>verbose</dt>
      <dd>Print progress? Default is <code>FALSE</code>.</dd>
    </dl>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A new <code>bigSNP</code>.</p>
    
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <p><a href='bigSNP-class.html'>bigSNP</a></p>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'>## Not run: ------------------------------------</span>
<span class='co'># </span>
<span class='co'>#   bedfile &lt;- system.file("extdata", "example.bed", package = "bigsnpr")</span>
<span class='co'># </span>
<span class='co'>#   # Creating directory for backing files</span>
<span class='co'>#   if (!dir.exists("backingfiles")) dir.create("backingfiles")</span>
<span class='co'># </span>
<span class='co'>#   if (!file.exists("backingfiles/test_doc.bk")) {</span>
<span class='co'>#     # Reading the bedfile and storing the data</span>
<span class='co'>#     test &lt;- BedToBig(bedfile, 50, "test_doc")</span>
<span class='co'>#   } else {</span>
<span class='co'>#     # Loading it from backing files</span>
<span class='co'>#     test &lt;- AttachBigSNP("test_doc")</span>
<span class='co'>#   }</span>
<span class='co'># </span>
<span class='co'>#   # dimensions</span>
<span class='co'>#   print(rbind(dim(test$genotypes), dim(test$fam), dim(test$map)))</span>
<span class='co'># </span>
<span class='co'>#   # taking only the first 50 individuals and 500 SNPs at random</span>
<span class='co'>#   ind.row &lt;- 1:50</span>
<span class='co'>#   ind.col &lt;- sort(sample(ncol(test$genotypes), 500))</span>
<span class='co'>#   test2 &lt;- sub.bigSNP(test, ind.row, ind.col)</span>
<span class='co'># </span>
<span class='co'>#   # dimensions</span>
<span class='co'>#   print(rbind(dim(test2$genotypes), dim(test2$fam), dim(test2$map)))</span>
<span class='co'>#   # new backing files</span>
<span class='co'>#   print(test2$backingfile)</span>
<span class='co'>#   print(test2$backingpath)</span>
<span class='co'># </span>
<span class='co'># </span>
<span class='co'>#   # removing the 100th and 120th individuals</span>
<span class='co'>#   ind.row.del &lt;- c(100, 120)</span>
<span class='co'>#   test3 &lt;- sub.bigSNP(test, -ind.row.del)</span>
<span class='co'># </span>
<span class='co'>#   # dimensions</span>
<span class='co'>#   print(rbind(dim(test3$genotypes), dim(test3$fam), dim(test3$map)))</span>
<span class='co'>#   print(test3$fam$sample.ID[98:120])</span>
<span class='co'>#   # new backing files</span>
<span class='co'>#   print(test3$backingfile) # incrementation of the file's number</span>
<span class='co'>## ---------------------------------------------</span></div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#see-also">See also</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Florian Privé.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  </body>
</html>
