<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Truncated SVD with pruning — snp_autoSVD • bigsnpr</title>

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



<meta property="og:title" content="Truncated SVD with pruning — snp_autoSVD" />

<meta property="og:description" content="Fast truncated SVD which iteratively try to remove long-range LD regions
which appear in loadings of SVD." />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">bigsnpr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.11.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../reference/index.html">Manual</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/demo.html">Quick demo</a>
    </li>
    <li>
      <a href="../articles/exo.html">Exercise</a>
    </li>
    <li>
      <a href="../articles/how-to-PCA.html">How to capture Population Structure with PCA</a>
    </li>
    <li>
      <a href="../articles/pruning-vs-clumping.html">Why clumping should be preferred over pruning</a>
    </li>
    <li>
      <a href="../articles/SCT.html">Computing polygenic scores using Stacked Clumping and Thresholding (SCT)</a>
    </li>
  </ul>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://privefl.github.io/about.html">
    <span class="fa fa-user"></span>
     
    About
  </a>
</li>
<li>
  <a href="https://github.com/privefl/bigsnpr">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Truncated SVD with pruning</h1>
    
    <div class="hidden name"><code>snp_autoSVD.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>Fast truncated SVD which iteratively try to remove long-range LD regions
which appear in loadings of SVD.</p>
    
    </div>

    <pre class="usage"><span class='fu'>snp_autoSVD</span>(<span class='no'>G</span>, <span class='no'>infos.chr</span>, <span class='kw'>infos.pos</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>ind.row</span> <span class='kw'>=</span> <span class='fu'>rows_along</span>(<span class='no'>G</span>),
  <span class='kw'>ind.col</span> <span class='kw'>=</span> <span class='fu'>cols_along</span>(<span class='no'>G</span>), <span class='kw'>fun.scaling</span> <span class='kw'>=</span> <span class='fu'><a href='snp_scaleBinom.html'>snp_scaleBinom</a></span>(),
  <span class='kw'>thr.r2</span> <span class='kw'>=</span> <span class='fl'>0.2</span>, <span class='kw'>size</span> <span class='kw'>=</span> <span class='fl'>100</span>/<span class='no'>thr.r2</span>, <span class='kw'>k</span> <span class='kw'>=</span> <span class='fl'>10</span>, <span class='kw'>roll.size</span> <span class='kw'>=</span> <span class='fl'>50</span>,
  <span class='kw'>int.min.size</span> <span class='kw'>=</span> <span class='fl'>20</span>, <span class='kw'>is.size.in.bp</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>ncores</span> <span class='kw'>=</span> <span class='fl'>1</span>,
  <span class='kw'>verbose</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>G</th>
      <td><p>A FBM.code256
(typically <code>&lt;bigSNP&gt;$genotypes</code>).<br />
<strong>You shouldn't have missing values in your data or SNPs with 0 MAF.</strong></p></td>
    </tr>
    <tr>
      <th>infos.chr</th>
      <td><p>Vector of integers specifying each SNP's chromosome.<br />
Typically <code>&lt;bigSNP&gt;$map$chromosome</code>.</p></td>
    </tr>
    <tr>
      <th>infos.pos</th>
      <td><p>Vector of integers specifying the physical position
on a chromosome (in base pairs) of each SNP.<br />
Typically <code>&lt;bigSNP&gt;$map$physical.pos</code>.</p></td>
    </tr>
    <tr>
      <th>ind.row</th>
      <td><p>An optional vector of the row indices (individuals) that
are used. If not specified, all rows are used.<br />
<strong>Don't use negative indices.</strong></p></td>
    </tr>
    <tr>
      <th>ind.col</th>
      <td><p>An optional vector of the column indices (SNPs) that are used.
If not specified, all columns are used.<br />
<strong>Don't use negative indices.</strong></p></td>
    </tr>
    <tr>
      <th>fun.scaling</th>
      <td><p>A function that returns a named list of <code>mean</code> and <code>sd</code>
for every column, to scale each of their elements such as followed:
$$\frac{X_{i,j} - mean_j}{sd_j}.$$ Default is <code><a href='snp_scaleBinom.html'>snp_scaleBinom()</a></code>.</p></td>
    </tr>
    <tr>
      <th>thr.r2</th>
      <td><p>Threshold over the squared correlation between two SNPs.
Default is <code>0.2</code>. Use <code>NA</code> if you want to skip the clumping step.</p></td>
    </tr>
    <tr>
      <th>size</th>
      <td><p>For one SNP, window size around this SNP to compute correlations.
Default is <code>100 / thr.r2</code> for clumping (0.2 -&gt; 500; 0.1 -&gt; 1000; 0.5 -&gt; 200).
If not providing <code>infos.pos</code> (<code>NULL</code>, the default), this is a window in
number of SNPs, otherwise it is a window in kb (genetic distance).
I recommend that you provide the positions if available.</p></td>
    </tr>
    <tr>
      <th>k</th>
      <td><p>Number of singular vectors/values to compute. Default is <code>10</code>.
<strong>This algorithm should be used to compute a few singular vectors/values.</strong></p></td>
    </tr>
    <tr>
      <th>roll.size</th>
      <td><p>Radius of rolling windows to smooth log-p-values.
Default is <code>50</code>.</p></td>
    </tr>
    <tr>
      <th>int.min.size</th>
      <td><p>Minimum number of consecutive outlier SNPs
in order to be reported as long-range LD region. Default is <code>20</code>.</p></td>
    </tr>
    <tr>
      <th>is.size.in.bp</th>
      <td><p>Deprecated.</p></td>
    </tr>
    <tr>
      <th>ncores</th>
      <td><p>Number of cores used. Default doesn't use parallelism.
You may use nb_cores.</p></td>
    </tr>
    <tr>
      <th>verbose</th>
      <td><p>Output some information on the iterations? Default is <code>TRUE</code>.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A named list (an S3 class "big_SVD") of</p><ul>
<li><p><code>d</code>, the singular values,</p></li>
<li><p><code>u</code>, the left singular vectors,</p></li>
<li><p><code>v</code>, the right singular vectors,</p></li>
<li><p><code>niter</code>, the number of the iteration of the algorithm,</p></li>
<li><p><code>nops</code>, number of Matrix-Vector multiplications used,</p></li>
<li><p><code>center</code>, the centering vector,</p></li>
<li><p><code>scale</code>, the scaling vector.</p></li>
</ul>

<p>Note that to obtain the Principal Components, you must use
predict on the result. See examples.</p>
    
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>If you don't have any information about SNPs, you can try using</p><ul>
<li><p><code>infos.chr = rep(1, ncol(G))</code>,</p></li>
<li><p><code>size = ncol(G)</code> (if SNPs are not sorted),</p></li>
<li><p><code>roll.size = 0</code> (if SNPs are not sorted).</p></li>
</ul>
    <p>Improvements will come in the future, as for example, warm starts in order
to make the SVD computations faster.</p>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='no'>ex</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='snp_attachExtdata.html'>snp_attachExtdata</a></span>()

<span class='no'>obj.svd</span> <span class='kw'>&lt;-</span> <span class='fu'>snp_autoSVD</span>(<span class='kw'>G</span> <span class='kw'>=</span> <span class='no'>ex</span>$<span class='no'>genotypes</span>,
                       <span class='kw'>infos.chr</span> <span class='kw'>=</span> <span class='no'>ex</span>$<span class='no'>map</span>$<span class='no'>chromosome</span>,
                       <span class='kw'>infos.pos</span> <span class='kw'>=</span> <span class='no'>ex</span>$<span class='no'>map</span>$<span class='no'>physical.position</span>)</div><div class='output co'>#&gt; 
#&gt; Phase of clumping (on MAF) at r^2 &gt; 0.2.. keep 4270 SNPs.
#&gt; 
#&gt; Iteration 1:
#&gt; Computing SVD..
#&gt; 0 outlier detected..
#&gt; 
#&gt; Converged!</div><div class='input'>
<span class='fu'><a href='https://www.rdocumentation.org/packages/utils/topics/str'>str</a></span>(<span class='no'>obj.svd</span>)</div><div class='output co'>#&gt; List of 7
#&gt;  $ d     : num [1:10] 235.4 148 105.5 96.4 94.9 ...
#&gt;  $ u     : num [1:517, 1:10] 0.0801 0.0798 0.0646 0.0781 0.0818 ...
#&gt;  $ v     : num [1:4270, 1:10] -0.00174 0.03142 -0.01527 0.0132 0.0154 ...
#&gt;  $ niter : int 10
#&gt;  $ nops  : int 170
#&gt;  $ center: num [1:4270] 0.412 0.474 0.369 0.913 0.712 ...
#&gt;  $ scale : num [1:4270] 0.572 0.601 0.549 0.704 0.677 ...
#&gt;  - attr(*, "class")= chr "big_SVD"
#&gt;  - attr(*, "subset")= int [1:4270] 2 3 4 5 6 7 8 9 10 11 ...
#&gt;  - attr(*, "lrldr")='data.frame':	0 obs. of  3 variables:
#&gt;   ..$ Chr  : int(0) 
#&gt;   ..$ Start: int(0) 
#&gt;   ..$ Stop : int(0) </div><div class='input'>
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#details">Details</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Florian Privé.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
   </div>

  

  </body>
</html>

