<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>MAX3 statistic — snp_MAX3 • bigsnpr</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">


<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script>
<script src="../pkgdown.js"></script>

<!-- mathjax -->
<script src='https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">bigsnpr</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/index.html">Articles</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/privefl/bigsnpr">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

      <div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>MAX3 statistic</h1>
    </div>

    
    <p>Counts the number of 0, 1, 2 by SNP and phenotype (cases/controls)
and then compute the MAX3 statistic.</p>
    

    <pre><span class='fu'>snp_MAX3</span>(<span class='no'>x</span>, <span class='kw'>ind.train</span> <span class='kw'>=</span> <span class='fu'>seq</span>(<span class='fu'>nrow</span>(<span class='no'>X</span>)), <span class='kw'>val</span> <span class='kw'>=</span> <span class='fu'>c</span>(<span class='fl'>0</span>, <span class='fl'>0.5</span>, <span class='fl'>1</span>))</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a> Arguments</h2>
    <dl class="dl-horizontal">
      <dt>x</dt>
      <dd>A <a href='bigSNP-class.html'>bigSNP</a>.</dd>
      <dt>ind.train</dt>
      <dd>An optional vector of the row indices that are used,
for the training part. If not specified, all data are used.</dd>
      <dt>val</dt>
      <dd>Computing <code>ZCATT(x)</code> for <code>x</code> in <code>val</code>. Default is <code>c(0, 0.5, 1)</code>
and corresponds to the <em>MAX3</em> statistic. Only <code>c(0, 1)</code> corresponds to
<em>MAX2</em>. And only <code>0.5</code> corresponds to the Armatige trend test.
<code>seq(0, 1, length.out = L)</code> corresponds to <em>MAXL</em>.</dd>
    </dl>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A named list of <strong><code>S</code></strong> and <strong><code>pS</code></strong> for every column,
which are MAX3 statistics and associated p-values. <strong>P-values are in
fact the minimum of the 3 p-values of each test separately.</strong> One can use
genomic control to rescale these p-values.</p>
    
    <h2 class="hasAnchor" id="references"><a class="anchor" href="#references"></a>References</h2>

    <p>Zheng, G., Yang, Y., Zhu, X., &amp; Elston, R. (2012).
Robust Procedures. Analysis Of Genetic Association Studies, 151-206.
<a href = 'http://dx.doi.org/10.1007/978-1-4614-2245-7_6'>http://dx.doi.org/10.1007/978-1-4614-2245-7_6</a>.</p>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='fu'>set.seed</span>(<span class='fl'>1</span>)

<span class='co'># constructing a fake genotype big.matrix</span>
<span class='no'>a</span> <span class='kw'>&lt;-</span> <span class='fu'>big.matrix</span>(<span class='fl'>10</span>, <span class='fl'>15</span>, <span class='kw'>type</span> <span class='kw'>=</span> <span class='st'>"char"</span>, <span class='kw'>shared</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>)
<span class='no'>a</span>[] <span class='kw'>&lt;-</span> <span class='fu'>sample</span>(<span class='fu'>c</span>(<span class='fl'>0</span>, <span class='fl'>1</span>, <span class='fl'>2</span>), <span class='fl'>150</span>, <span class='fl'>TRUE</span>)</div><div class='output co'>#&gt; Assignment will down cast from double to char
#&gt; Hint: To remove this warning type:  options(bigmemory.typecast.warning=FALSE)</div><div class='output co'>#&gt; <span class='warning'>Warning: </span></div><div class='input'><span class='fu'>print</span>(<span class='no'>a</span>[,])</div><div class='output co'>#&gt;       [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8] [,9] [,10] [,11] [,12] [,13]
#&gt;  [1,]    0    0    2    1    2    1    2    1    1     0     1     2     2
#&gt;  [2,]    1    0    0    1    1    2    0    2    2     0     1     2     1
#&gt;  [3,]    1    2    1    1    2    1    1    1    1     1     0     1     1
#&gt;  [4,]    2    1    0    0    1    0    0    1    0     2     2     1     0
#&gt;  [5,]    0    2    0    2    1    0    1    1    2     2     1     0     2
#&gt;  [6,]    2    1    1    2    2    0    0    2    0     2     0     0     1
#&gt;  [7,]    2    2    0    2    0    0    1    2    2     1     0     2     1
#&gt;  [8,]    1    2    1    0    1    1    2    1    0     1     1     0     0
#&gt;  [9,]    1    1    2    2    2    1    0    2    0     2     2     1     0
#&gt; [10,]    0    2    1    1    2    1    2    2    0     1     1     1     1
#&gt;       [,14] [,15]
#&gt;  [1,]     1     2
#&gt;  [2,]     0     1
#&gt;  [3,]     0     0
#&gt;  [4,]     1     0
#&gt;  [5,]     2     2
#&gt;  [6,]     1     1
#&gt;  [7,]     1     0
#&gt;  [8,]     1     2
#&gt;  [9,]     2     0
#&gt; [10,]     1     2</div><div class='input'>
<span class='co'># constructing a fake incomplete bigSNP with 10 individuals and 15 SNPs</span>
<span class='co'># where the 5 first individuals are cases and the 5 last are controls.</span>
<span class='no'>fake</span> <span class='kw'>&lt;-</span> <span class='fu'>list</span>()
<span class='fu'>class</span>(<span class='no'>fake</span>) <span class='kw'>&lt;-</span> <span class='st'>"bigSNP"</span>
<span class='no'>fake</span>$<span class='no'>genotypes</span> <span class='kw'>&lt;-</span> <span class='no'>a</span>
<span class='no'>fake</span>$<span class='no'>fam</span>$<span class='no'>affection</span> <span class='kw'>&lt;-</span> <span class='fu'>c</span>(<span class='fu'>rep</span>(<span class='fl'>1</span>, <span class='fl'>5</span>), <span class='fu'>rep</span>(-<span class='fl'>1</span>, <span class='fl'>5</span>))

<span class='co'># Get MAX3 statistics</span>
<span class='fu'>print</span>(<span class='fu'>snp_MAX3</span>(<span class='no'>fake</span>))</div><div class='output co'>#&gt; $S
#&gt;  [1] 0.6666667 2.5000000 1.6666667 1.6666667 1.1111111 1.1111111 0.4761905
#&gt;  [8] 3.6000000 3.6000000 2.5000000 0.4761905 0.6666667 2.5000000 2.5000000
#&gt; [15] 0.0000000
#&gt; 
#&gt; $pS
#&gt;  [1] 0.41421618 0.11384630 0.19670560 0.19670560 0.29184055 0.29184055
#&gt;  [7] 0.49015296 0.05777957 0.05777957 0.11384630 0.49015296 0.41421618
#&gt; [13] 0.11384630 0.11384630 1.00000000
#&gt; </div><div class='input'>
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#references">References</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Florian Privé.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  </body>
</html>
