<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Fast imputation — snp_fastImpute • bigsnpr</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script>
<script src="../pkgdown.js"></script>



<meta property="og:title" content="Fast imputation — snp_fastImpute" />

<meta property="og:description" content="Fast imputation algorithm based on local XGBoost models." />
<meta name="twitter:card" content="summary" />


<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">bigsnpr</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">0.3.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../reference/index.html">Manual</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/demo.html">Quick demo</a>
    </li>
    <li>
      <a href="../articles/how-to-PCA.html">How to capture Population Structure with PCA</a>
    </li>
    <li>
      <a href="../articles/pruning-vs-clumping.html">Why clumping should be preferred over pruning</a>
    </li>
    <li>
      <a href="../articles/dosage.html">Dosage data: proof of concept</a>
    </li>
  </ul>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://privefl.github.io/about.html">
    <span class="fa fa-user"></span>
     
    About
  </a>
</li>
<li>
  <a href="https://github.com/privefl/bigsnpr">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Fast imputation</h1>
    
    <div class="hidden name"><code>snp_fastImpute.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>Fast imputation algorithm based on local XGBoost models.</p>
    
    </div>

    <pre class="usage"><span class='fu'>snp_fastImpute</span>(<span class='no'>Gna</span>, <span class='no'>infos.chr</span>, <span class='kw'>alpha</span> <span class='kw'>=</span> <span class='fl'>1e-04</span>, <span class='kw'>size</span> <span class='kw'>=</span> <span class='fl'>200</span>, <span class='kw'>p.train</span> <span class='kw'>=</span> <span class='fl'>0.8</span>,
  <span class='kw'>n.cor</span> <span class='kw'>=</span> <span class='fu'>nrow</span>(<span class='no'>Gna</span>), <span class='kw'>seed</span> <span class='kw'>=</span> <span class='fl'>NA</span>, <span class='kw'>ncores</span> <span class='kw'>=</span> <span class='fl'>1</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>Gna</th>
      <td><p>A FBM.code256
(typically <code>&lt;bigSNP&gt;$genotypes</code>).<br />
You can have missing values in your data.</p></td>
    </tr>
    <tr>
      <th>infos.chr</th>
      <td><p>Vector of integers specifying each SNP's chromosome.<br />
Typically <code>&lt;bigSNP&gt;$map$chromosome</code>.</p></td>
    </tr>
    <tr>
      <th>alpha</th>
      <td><p>Type-I error for testing correlations. Default is <code>1e-4</code>.</p></td>
    </tr>
    <tr>
      <th>size</th>
      <td><p>Number of neighbor SNPs to be possibly included in the model
imputing this particular SNP. Default is <code>200</code>.</p></td>
    </tr>
    <tr>
      <th>p.train</th>
      <td><p>Proportion of non missing genotypes that are used for training
the imputation model while the rest is used to assess the accuracy of
this imputation model. Default is <code>0.8</code>.</p></td>
    </tr>
    <tr>
      <th>n.cor</th>
      <td><p>Number of rows that are used to estimate correlations.
Default uses them all.</p></td>
    </tr>
    <tr>
      <th>seed</th>
      <td><p>An integer, for reproducibility. Default doesn't use seeds.</p></td>
    </tr>
    <tr>
      <th>ncores</th>
      <td><p>Number of cores used. Default doesn't use parallelism.
You may use nb_cores.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>An FBM with</p><ul>
<li><p>the proportion of missing values by SNP (first row),</p></li>
<li><p>the estimated proportion of imputation errors by SNP (second row).</p></li>
</ul>

    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='no'>fake</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='snp_attachExtdata.html'>snp_attachExtdata</a></span>(<span class='st'>"example-missing.bed"</span>)
<span class='no'>G</span> <span class='kw'>&lt;-</span> <span class='no'>fake</span>$<span class='no'>genotypes</span>
<span class='no'>CHR</span> <span class='kw'>&lt;-</span> <span class='no'>fake</span>$<span class='no'>map</span>$<span class='no'>chromosome</span>
<span class='no'>G2</span> <span class='kw'>&lt;-</span> <span class='fu'>snp_fastImpute</span>(<span class='no'>G</span>, <span class='no'>CHR</span>)
<span class='no'>G2</span>[, <span class='fl'>1</span>:<span class='fl'>5</span>]</div><div class='output co'>#&gt;      [,1]  [,2]  [,3]       [,4] [,5]
#&gt; [1,]    0 0.020 0.005 0.03500000 0.01
#&gt; [2,]   NA 0.025 0.000 0.02564103 0.00</div><div class='input'>
<span class='co'># Still missing values</span>
<span class='fu'>big_counts</span>(<span class='no'>G</span>, <span class='kw'>ind.col</span> <span class='kw'>=</span> <span class='fl'>1</span>:<span class='fl'>10</span>)</div><div class='output co'>#&gt;      [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8] [,9] [,10]
#&gt; 0     193  181  177  161  162  156  142  142  126   131
#&gt; 1       7   15   21   29   33   32   44   49   49    55
#&gt; 2       0    0    1    3    3    4    3    8    5     8
#&gt; &lt;NA&gt;    0    4    1    7    2    8   11    1   20     6</div><div class='input'><span class='co'># You need to change the code of G</span>
<span class='co'># To make this permanent, you need to save (modify) the file on disk</span>
<span class='no'>fake</span>$<span class='no'>genotypes</span>$<span class='no'>code256</span> <span class='kw'>&lt;-</span> <span class='kw pkg'>bigsnpr</span><span class='kw ns'>:::</span><span class='no'>CODE_IMPUTE_PRED</span>
<span class='no'>fake</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='snp_save.html'>snp_save</a></span>(<span class='no'>fake</span>)
<span class='fu'>big_counts</span>(<span class='no'>fake</span>$<span class='no'>genotypes</span>)[<span class='fl'>4</span>, ]</div><div class='output co'>#&gt;   [1] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
#&gt;  [38] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
#&gt;  [75] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
#&gt; [112] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
#&gt; [149] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
#&gt; [186] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
#&gt; [223] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
#&gt; [260] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
#&gt; [297] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
#&gt; [334] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
#&gt; [371] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
#&gt; [408] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
#&gt; [445] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
#&gt; [482] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0</div><div class='input'>
<span class='co'># Plot for post-checking</span>
<span class='co'>## Here there is no SNP with more than 1% error (estimated)</span>
<span class='no'>pvals</span> <span class='kw'>&lt;-</span> <span class='fu'>c</span>(<span class='fl'>0.01</span>, <span class='fl'>0.005</span>, <span class='fl'>0.002</span>, <span class='fl'>0.001</span>); <span class='no'>colvals</span> <span class='kw'>&lt;-</span> <span class='fl'>2</span>:<span class='fl'>5</span>
<span class='no'>df</span> <span class='kw'>&lt;-</span> <span class='fu'>data.frame</span>(<span class='kw'>pNA</span> <span class='kw'>=</span> <span class='no'>G2</span>[<span class='fl'>1</span>, ], <span class='kw'>pError</span> <span class='kw'>=</span> <span class='no'>G2</span>[<span class='fl'>2</span>, ])

<span class='co'># base R</span>
<span class='fu'>plot</span>(<span class='fu'>subset</span>(<span class='no'>df</span>, <span class='no'>pNA</span> <span class='kw'>&gt;</span> <span class='fl'>0.001</span>), <span class='kw'>pch</span> <span class='kw'>=</span> <span class='fl'>20</span>)</div><div class='input'><span class='no'>idc</span> <span class='kw'>&lt;-</span> <span class='fu'>lapply</span>(<span class='fu'>seq_along</span>(<span class='no'>pvals</span>), <span class='kw'>function</span>(<span class='no'>i</span>) {
  <span class='fu'>curve</span>(<span class='no'>pvals</span>[<span class='no'>i</span>] / <span class='no'>x</span>, <span class='kw'>from</span> <span class='kw'>=</span> <span class='fl'>0</span>, <span class='kw'>lwd</span> <span class='kw'>=</span> <span class='fl'>2</span>,
        <span class='kw'>col</span> <span class='kw'>=</span> <span class='no'>colvals</span>[<span class='no'>i</span>], <span class='kw'>add</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>)
})</div><div class='input'><span class='fu'>legend</span>(<span class='st'>"topright"</span>, <span class='kw'>legend</span> <span class='kw'>=</span> <span class='no'>pvals</span>, <span class='kw'>title</span> <span class='kw'>=</span> <span class='st'>"p(NA &amp; Error)"</span>,
       <span class='kw'>col</span> <span class='kw'>=</span> <span class='no'>colvals</span>, <span class='kw'>lty</span> <span class='kw'>=</span> <span class='fl'>1</span>, <span class='kw'>lwd</span> <span class='kw'>=</span> <span class='fl'>2</span>)</div><div class='img'><img src='snp_fastImpute-1.png' alt='' width='700' height='433' /></div><div class='input'>
<span class='co'># ggplot2</span>
<span class='fu'>library</span>(<span class='no'>ggplot2</span>)
<span class='fu'>Reduce</span>(<span class='kw'>function</span>(<span class='no'>p</span>, <span class='no'>i</span>) {
  <span class='no'>p</span> + <span class='fu'><a href='http://www.rdocumentation.org/packages/ggplot2/topics/stat_function'>stat_function</a></span>(<span class='kw'>fun</span> <span class='kw'>=</span> <span class='kw'>function</span>(<span class='no'>x</span>) <span class='no'>pvals</span>[<span class='no'>i</span>] / <span class='no'>x</span>, <span class='kw'>color</span> <span class='kw'>=</span> <span class='no'>colvals</span>[<span class='no'>i</span>])
}, <span class='kw'>x</span> <span class='kw'>=</span> <span class='fu'>seq_along</span>(<span class='no'>pvals</span>), <span class='kw'>init</span> <span class='kw'>=</span> <span class='fu'><a href='http://www.rdocumentation.org/packages/ggplot2/topics/ggplot'>ggplot</a></span>(<span class='no'>df</span>, <span class='fu'><a href='http://www.rdocumentation.org/packages/ggplot2/topics/aes'>aes</a></span>(<span class='no'>pNA</span>, <span class='no'>pError</span>))) +
  <span class='fu'><a href='http://www.rdocumentation.org/packages/ggplot2/topics/geom_point'>geom_point</a></span>() +
  <span class='fu'><a href='http://www.rdocumentation.org/packages/ggplot2/topics/coord_cartesian'>coord_cartesian</a></span>(<span class='kw'>ylim</span> <span class='kw'>=</span> <span class='fu'>range</span>(<span class='no'>df</span>$<span class='no'>pError</span>, <span class='kw'>na.rm</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>)) +
  <span class='fu'><a href='http://www.rdocumentation.org/packages/ggplot2/topics/ggtheme'>theme_bw</a></span>(<span class='fl'>15</span>)</div><div class='output co'>#&gt; <span class='warning'>Warning: Removed 140 rows containing missing values (geom_point).</span></div><div class='img'><img src='snp_fastImpute-2.png' alt='' width='700' height='433' /></div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Florian Privé.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  

  </body>
</html>

