<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Manhattan plot — snp_manhattan • bigsnpr</title>

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



<meta property="og:title" content="Manhattan plot — snp_manhattan" />

<meta property="og:description" content="Creates a manhattan plot." />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">bigsnpr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.9.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../reference/index.html">Manual</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/demo.html">Quick demo</a>
    </li>
    <li>
      <a href="../articles/exo.html">Exercise</a>
    </li>
    <li>
      <a href="../articles/how-to-PCA.html">How to capture Population Structure with PCA</a>
    </li>
    <li>
      <a href="../articles/pruning-vs-clumping.html">Why clumping should be preferred over pruning</a>
    </li>
  </ul>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://privefl.github.io/about.html">
    <span class="fa fa-user"></span>
     
    About
  </a>
</li>
<li>
  <a href="https://github.com/privefl/bigsnpr">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Manhattan plot</h1>
    
    <div class="hidden name"><code>snp_manhattan.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>Creates a manhattan plot.</p>
    
    </div>

    <pre class="usage"><span class='fu'>snp_manhattan</span>(<span class='no'>gwas</span>, <span class='no'>infos.chr</span>, <span class='no'>infos.pos</span>, <span class='kw'>colors</span> <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/c'>c</a></span>(<span class='st'>"black"</span>, <span class='st'>"grey60"</span>),
  <span class='kw'>dist.sep.chrs</span> <span class='kw'>=</span> <span class='fl'>1e+07</span>, <span class='kw'>ind.highlight</span> <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/integer'>integer</a></span>(<span class='fl'>0</span>),
  <span class='kw'>col.highlight</span> <span class='kw'>=</span> <span class='st'>"red"</span>, <span class='kw'>labels</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>npoints</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>coeff</span> <span class='kw'>=</span> <span class='fl'>1</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>gwas</th>
      <td><p>A <code>mhtest</code> object with the p-values associated with each SNP.
Typically, the output of big_univLinReg, big_univLogReg or <a href='snp_pcadapt.html'>snp_pcadapt</a>.</p></td>
    </tr>
    <tr>
      <th>infos.chr</th>
      <td><p>Vector of integers specifying each SNP's chromosome.<br />
Typically <code>&lt;bigSNP&gt;$map$chromosome</code>.</p></td>
    </tr>
    <tr>
      <th>infos.pos</th>
      <td><p>Vector of integers specifying the physical position
on a chromosome (in base pairs) of each SNP.<br />
Typically <code>&lt;bigSNP&gt;$map$physical.pos</code>.</p></td>
    </tr>
    <tr>
      <th>colors</th>
      <td><p>Colors used for each chromosome (they are recycled).
Default is an alternation of black and gray.</p></td>
    </tr>
    <tr>
      <th>dist.sep.chrs</th>
      <td><p>"Physical" distance that separates two chromosomes.
Default is 10 Mbp.</p></td>
    </tr>
    <tr>
      <th>ind.highlight</th>
      <td><p>Indices of SNPs you want to highlight (of interest).
Default doesn't highlight any SNPs.</p></td>
    </tr>
    <tr>
      <th>col.highlight</th>
      <td><p>Color used for highlighting SNPs. Default uses red.</p></td>
    </tr>
    <tr>
      <th>labels</th>
      <td><p>Labels of the x axis. Default uses the number of the
chromosome there are in <code>infos.chr</code>(<code><a href='https://www.rdocumentation.org/packages/base/topics/sort'>sort(unique(infos.chr))</a></code>). This may be
useful to restrict the number of labels so that they are not overlapping.</p></td>
    </tr>
    <tr>
      <th>npoints</th>
      <td><p>Number of points to keep (ranked by p-value) in order to get
a lighter object (and plot). Default doesn't cut anything.
If used, the resulting object will have an attribute called <code>subset</code>
giving the indices of the kept points.</p></td>
    </tr>
    <tr>
      <th>coeff</th>
      <td><p>Relative size of text. Default is <code>1</code>.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A <code>ggplot2</code> object. You can plot it using the <code>print</code> method.
You can modify it as you wish by adding layers. You might want to read
<a href='http://r4ds.had.co.nz/data-visualisation.html'>this chapter</a>
to get more familiar with the package <strong>ggplot2</strong>.</p>
    
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>If you don't have information of chromosome and position, you should simply
use <code>plot</code> instead.</p>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/Random'>set.seed</a></span>(<span class='fl'>9</span>)

<span class='no'>test</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='snp_attachExtdata.html'>snp_attachExtdata</a></span>()
<span class='no'>G</span> <span class='kw'>&lt;-</span> <span class='no'>test</span>$<span class='no'>genotypes</span>
<span class='no'>y</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/stats/topics/Normal'>rnorm</a></span>(<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/nrow'>nrow</a></span>(<span class='no'>G</span>))

<span class='no'>gwas</span> <span class='kw'>&lt;-</span> <span class='fu'>big_univLinReg</span>(<span class='no'>G</span>, <span class='no'>y</span>)
<span class='fu'><a href='snp_qq.html'>snp_qq</a></span>(<span class='no'>gwas</span>)</div><div class='img'><img src='snp_manhattan-1.png' alt='' width='700' height='433' /></div><div class='input'><span class='no'>gwas_gc</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='snp_gc.html'>snp_gc</a></span>(<span class='no'>gwas</span>) <span class='co'># change attr(gwas_gc, "transfo")</span>

<span class='fu'><a href='snp_qq.html'>snp_qq</a></span>(<span class='no'>gwas_gc</span>)</div><div class='img'><img src='snp_manhattan-2.png' alt='' width='700' height='433' /></div><div class='input'><span class='co'># The next plot should be prettier with a real dataset</span>
<span class='fu'>snp_manhattan</span>(<span class='no'>gwas_gc</span>,
              <span class='kw'>infos.chr</span> <span class='kw'>=</span> <span class='no'>test</span>$<span class='no'>map</span>$<span class='no'>chromosome</span>,
              <span class='kw'>infos.pos</span> <span class='kw'>=</span> <span class='no'>test</span>$<span class='no'>map</span>$<span class='no'>physical.pos</span>)</div><div class='img'><img src='snp_manhattan-3.png' alt='' width='700' height='433' /></div><div class='input'>
<span class='no'>p</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='snp_qq.html'>snp_qq</a></span>(<span class='no'>gwas_gc</span>) + <span class='kw pkg'>ggplot2</span><span class='kw ns'>::</span><span class='fu'><a href='https://ggplot2.tidyverse.org/reference/aes.html'>aes</a></span>(<span class='kw'>text</span> <span class='kw'>=</span> <span class='fu'>asPlotlyText</span>(<span class='no'>test</span>$<span class='no'>map</span>))</div><span class='co'># NOT RUN {</span>
<span class='kw pkg'>plotly</span><span class='kw ns'>::</span><span class='fu'><a href='https://www.rdocumentation.org/packages/plotly/topics/ggplotly'>ggplotly</a></span>(<span class='no'>p</span>, <span class='kw'>tooltip</span> <span class='kw'>=</span> <span class='st'>"text"</span>)
<span class='co'># }</span></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#details">Details</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Florian Privé.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
   </div>

  

  </body>
</html>

