<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Truncated SVD with pruning — snp_clumpedSVD • bigsnpr</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://maxcdn.bootstrapcdn.com/bootswatch/3.3.7/readable/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">

<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">


<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script>
<script src="../pkgdown.js"></script>

<!-- mathjax -->
<script src='https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">bigsnpr</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../reference/index.html">Manual</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/pruning-vs-clumping.html">Why clumping should be preferred over pruning</a>
    </li>
    <li>
      <a href="../articles/demo.html">Small demo</a>
    </li>
    <li>
      <a href="../articles/bcm-seminar-packages.html">Two example analyses</a>
    </li>
  </ul>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://privefl.github.io/about.html">
    <span class="fa fa-user"></span>
     
    About
  </a>
</li>
<li>
  <a href="https://github.com/privefl/bigsnpr">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

      <div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Truncated SVD with pruning</h1>
    </div>

    
    <p>Fast truncated SVD which iteratively try to remove long-range LD regions
which appear in loadings of SVD.</p>
    

    <pre><span class='fu'>snp_clumpedSVD</span>(<span class='no'>G</span>, <span class='no'>infos.chr</span>, <span class='kw'>ind.row</span> <span class='kw'>=</span> <span class='fu'>rows_along</span>(<span class='no'>G</span>),
  <span class='kw'>ind.col</span> <span class='kw'>=</span> <span class='fu'>cols_along</span>(<span class='no'>G</span>), <span class='kw'>fun.scaling</span> <span class='kw'>=</span> <span class='fu'><a href='snp_scaleBinom.html'>snp_scaleBinom</a></span>(),
  <span class='kw'>thr.r2.init</span> <span class='kw'>=</span> <span class='fl'>0.2</span>, <span class='kw'>thr.r2.step</span> <span class='kw'>=</span> <span class='fl'>2</span>, <span class='kw'>size</span> <span class='kw'>=</span> <span class='fl'>500</span>, <span class='kw'>k</span> <span class='kw'>=</span> <span class='fl'>10</span>,
  <span class='kw'>roll.size</span> <span class='kw'>=</span> <span class='fl'>50</span>, <span class='kw'>int.min.size</span> <span class='kw'>=</span> <span class='fl'>20</span>, <span class='kw'>ncores</span> <span class='kw'>=</span> <span class='fl'>1</span>, <span class='kw'>verbose</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a> Arguments</h2>
    <dl class="dl-horizontal">
      <dt>G</dt>
      <dd>A BM.code.descriptor
(typically <code>&lt;bigSNP&gt;$genotypes</code>).
<strong>You shouldn&#39;t have missing values in your data or SNPs with 0 MAF.</strong></dd>
      <dt>infos.chr</dt>
      <dd>Vector of integers specifying each SNP&#39;s chromosome.
Typically <code>&lt;bigSNP&gt;$map$chromosome</code>.</dd>
      <dt>ind.row</dt>
      <dd>An optional vector of the row indices (individuals) that
are used. If not specified, all rows are used.
<strong>Don&#39;t use negative indices.</strong></dd>
      <dt>ind.col</dt>
      <dd>An optional vector of the column indices (SNPs) that are used.
If not specified, all columns are used.
<strong>Don&#39;t use negative indices.</strong></dd>
      <dt>fun.scaling</dt>
      <dd>A function that returns a named list of
<code>mean</code> and <code>sd</code> for every column, to scale each of their elements
such as followed: $$\frac{X_{i,j} - mean_j}{sd_j}.$$
Default is <code>snp_scaleBinom()</code>.</dd>
      <dt>thr.r2.init</dt>
      <dd>Threshold on squared correlation between SNPs.
Default is <code>0.2</code>.</dd>
      <dt>thr.r2.step</dt>
      <dd>Coefficient of division of <code>thr.r2</code> at each step.
Default is <code>2</code> which means that at iteration 2, <code>thr.r2 = 0.1</code>, at iteration
3, <code>thr.r2 = 0.05</code>, etc.</dd>
      <dt>size</dt>
      <dd>Radius of the window&#39;s size for the LD evaluations of the initial
step of clumping. Default is <code>500</code>.</dd>
      <dt>k</dt>
      <dd>Number of singular vectors/values to compute.
Default is <code>10</code>. <strong>This algorithm should be used to compute only
a few singular vectors/values.</strong></dd>
      <dt>roll.size</dt>
      <dd>Radius of rolling windows to smooth log-p-values.</dd>
      <dt>int.min.size</dt>
      <dd>Minimum size of intervals of consecutive significant
indices.</dd>
      <dt>ncores</dt>
      <dd>Number of cores used. Default doesn&#39;t use parallelism.</dd>
      <dt>verbose</dt>
      <dd>Output some information on the iterations? Default is <code>TRUE</code>.</dd>
    </dl>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A named list (an S3 class &quot;big_SVD&quot;) of
<ul>
<li> <code>d</code>, the singular values,
</li>
<li> <code>u</code>, the left singular vectors,
</li>
<li> <code>v</code>, the right singular vectors,
</li>
<li> <code>niter</code>, the number of the iteration of the algorithm,
</li>
<li> <code>nops</code>, number of Matrix-Vector multiplications used,
</li>
<li> <code>means</code>, the centering vector,
</li>
<li> <code>sds</code>, the scaling vector.
</li>
</ul></p>
    <p>Note that to obtain the Principal Components, you must use
predict on the result. See examples.</p>
    
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>Improvements will come in the future, as for example, warm starts in order
to make the SVD computations faster.</p>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='no'>ex</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='snp_attachExtdata.html'>snp_attachExtdata</a></span>()

<span class='no'>obj.svd</span> <span class='kw'>&lt;-</span> <span class='fu'>snp_clumpedSVD</span>(<span class='kw'>G</span> <span class='kw'>=</span> <span class='no'>ex</span>$<span class='no'>genotypes</span>,
                          <span class='kw'>infos.chr</span> <span class='kw'>=</span> <span class='no'>ex</span>$<span class='no'>map</span>$<span class='no'>chromosome</span>)</div><div class='output co'>#&gt; First phase of clumping at r2 &gt; 0.2.. keep 4270 SNPs.
#&gt; 
#&gt; Iteration 1:
#&gt; Computing SVD..
#&gt; 
#&gt; Converged!</div><div class='input'>
<span class='fu'>str</span>(<span class='no'>obj.svd</span>)</div><div class='output co'>#&gt; List of 7
#&gt;  $ d    : num [1:10] 235.4 148 105.5 96.4 94.9 ...
#&gt;  $ u    : num [1:517, 1:10] 0.0801 0.0798 0.0646 0.0781 0.0818 ...
#&gt;  $ v    : num [1:4270, 1:10] 0.00174 -0.03142 0.01527 -0.0132 -0.0154 ...
#&gt;  $ niter: int 10
#&gt;  $ nops : int 170
#&gt;  $ means: num [1:4270] 1.59 1.53 1.63 1.09 1.29 ...
#&gt;  $ sds  : num [1:4270] 0.572 0.601 0.549 0.704 0.677 ...
#&gt;  - attr(*, &quot;class&quot;)= chr &quot;big_SVD&quot;
#&gt;  - attr(*, &quot;subset&quot;)= int [1:4270] 2 3 4 5 6 7 8 9 10 11 ...</div><div class='input'>
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#details">Details</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Florian Privé.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  </body>
</html>
