<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Manhattan plot — snp_manhattan • bigsnpr</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://maxcdn.bootstrapcdn.com/bootswatch/3.3.7/readable/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">

<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">


<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script>
<script src="../pkgdown.js"></script>

<!-- mathjax -->
<script src='https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">bigsnpr</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../reference/index.html">Manual</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/pruning-vs-clumping.html">Why clumping should be preferred over pruning</a>
    </li>
    <li>
      <a href="../articles/demo.html">Small demo</a>
    </li>
    <li>
      <a href="../articles/bcm-seminar-packages.html">Two example analyses</a>
    </li>
  </ul>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://privefl.github.io/about.html">
    <span class="fa fa-user"></span>
     
    About
  </a>
</li>
<li>
  <a href="https://github.com/privefl/bigsnpr">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

      <div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Manhattan plot</h1>
    </div>

    
    <p>Creates a manhattan plot.</p>
    

    <pre><span class='fu'>snp_manhattan</span>(<span class='no'>gwas</span>, <span class='no'>infos.chr</span>, <span class='no'>infos.pos</span>, <span class='kw'>colors</span> <span class='kw'>=</span> <span class='fu'>c</span>(<span class='st'>"black"</span>, <span class='st'>"grey60"</span>),
  <span class='kw'>dist.sep.chrs</span> <span class='kw'>=</span> <span class='fl'>1e+07</span>, <span class='kw'>ind.highlight</span> <span class='kw'>=</span> <span class='fu'>integer</span>(<span class='fl'>0</span>),
  <span class='kw'>col.highlight</span> <span class='kw'>=</span> <span class='st'>"red"</span>, <span class='kw'>labels</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>npoints</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>coeff</span> <span class='kw'>=</span> <span class='fl'>1</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a> Arguments</h2>
    <dl class="dl-horizontal">
      <dt>gwas</dt>
      <dd>A <code>mhtest</code> object with the p-values associated with each SNP.
Typically, the output of big_univLinReg, big_univLogReg or <a href='snp_pcadapt.html'>snp_pcadapt</a>.</dd>
      <dt>infos.chr</dt>
      <dd>Vector of integers specifying each SNP&#39;s chromosome.
Typically <code>&lt;bigSNP&gt;$map$chromosome</code>.</dd>
      <dt>infos.pos</dt>
      <dd>Vector of integers specifying the physical position
on a chromosome (in base pairs) of each SNP.
Typically <code>&lt;bigSNP&gt;$map$physical.pos</code>.</dd>
      <dt>colors</dt>
      <dd>Colors that used for each chromosome (they are recycled).
Default is an alternance of black and grey.</dd>
      <dt>dist.sep.chrs</dt>
      <dd>&quot;Physical&quot; distance that separates two chromosomes.
Default is 10 Mbp.</dd>
      <dt>ind.highlight</dt>
      <dd>Indices of SNPs you want to highlight (of interest).
Default doesn&#39;t highlight any SNPs.</dd>
      <dt>col.highlight</dt>
      <dd>Color used for highlighting SNPs. Default uses red.</dd>
      <dt>labels</dt>
      <dd>Labels of the x axis. Default uses the number of the
chromosome there are in <code>infos.chr</code>(<code>sort(unique(infos.chr))</code>). This may be
useful to restrict the number of labels so that they are not overlapping.</dd>
      <dt>npoints</dt>
      <dd>Number of points to keep (ranked by p-value) in order to get
a lighter object (and plot). Default doesn&#39;t cut anything.
If used, the resulting object will have an attribute called <code>subset</code>
giving the indices of the kept points.</dd>
      <dt>coeff</dt>
      <dd>Relative size of text. Default is <code>1</code>.</dd>
    </dl>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A <code>ggplot2</code> object. You can plot it using the <code>print</code> method.
You can modify it as you wish by adding layers. You might want to read
<a href = 'this chapter'>http://r4ds.had.co.nz/data-visualisation.html</a>
to get more familiar with the package <strong>ggplot2</strong>.</p>
    
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>If you don&#39;t have infos of chromosome and position, you should simply
use <code>plot</code> instead.</p>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='fu'>set.seed</span>(<span class='fl'>9</span>)

<span class='no'>test</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='snp_attachExtdata.html'>snp_attachExtdata</a></span>()
<span class='no'>G</span> <span class='kw'>&lt;-</span> <span class='no'>test</span>$<span class='no'>genotypes</span>
<span class='no'>y</span> <span class='kw'>&lt;-</span> <span class='fu'>rnorm</span>(<span class='fu'>nrow</span>(<span class='no'>G</span>))

<span class='no'>gwas</span> <span class='kw'>&lt;-</span> <span class='fu'>big_univLinReg</span>(<span class='no'>G</span>, <span class='no'>y</span>)
<span class='fu'><a href='snp_qq.html'>snp_qq</a></span>(<span class='no'>gwas</span>)</div><img src='snp_manhattan-2.png' alt='' width='540' height='400' /><div class='input'><span class='no'>gwas_gc</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='snp_gc.html'>snp_gc</a></span>(<span class='no'>gwas</span>) <span class='co'># change attr(gwas_gc, "transfo")</span>

<span class='fu'><a href='snp_qq.html'>snp_qq</a></span>(<span class='no'>gwas_gc</span>)</div><img src='snp_manhattan-4.png' alt='' width='540' height='400' /><div class='input'><span class='co'># The next plot should be prettier with a real dataset</span>
<span class='fu'>snp_manhattan</span>(<span class='no'>gwas_gc</span>,
              <span class='kw'>infos.chr</span> <span class='kw'>=</span> <span class='no'>test</span>$<span class='no'>map</span>$<span class='no'>chromosome</span>,
              <span class='kw'>infos.pos</span> <span class='kw'>=</span> <span class='no'>test</span>$<span class='no'>map</span>$<span class='no'>physical.pos</span>)</div><img src='snp_manhattan-6.png' alt='' width='540' height='400' /><div class='input'>
<span class='no'>p</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='snp_qq.html'>snp_qq</a></span>(<span class='no'>gwas_gc</span>) + <span class='kw'>ggplot2</span><span class='kw'>::</span><span class='fu'>aes</span>(<span class='kw'>text</span> <span class='kw'>=</span> <span class='fu'>asPlotlyText</span>(<span class='no'>test</span>$<span class='no'>map</span>))
<span class='co'>## Not run: plotly::ggplotly(p, tooltip = &amp;quot;text&amp;quot;)</span></div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#details">Details</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Florian Privé.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  </body>
</html>
