<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>MAX3 statistic — snp_MAX3 • bigsnpr</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script>
<script src="../pkgdown.js"></script>



<meta property="og:title" content="MAX3 statistic — snp_MAX3" />

<meta property="og:description" content="Compute the MAX3 statistic, which tests for three genetic models
(additive, recessive and dominant)." />
<meta name="twitter:card" content="summary" />


<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">bigsnpr</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">0.3.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../reference/index.html">Manual</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/demo.html">Quick demo</a>
    </li>
    <li>
      <a href="../articles/how-to-PCA.html">How to capture Population Structure with PCA</a>
    </li>
    <li>
      <a href="../articles/pruning-vs-clumping.html">Why clumping should be preferred over pruning</a>
    </li>
    <li>
      <a href="../articles/dosage.html">Dosage data: proof of concept</a>
    </li>
  </ul>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://privefl.github.io/about.html">
    <span class="fa fa-user"></span>
     
    About
  </a>
</li>
<li>
  <a href="https://github.com/privefl/bigsnpr">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>MAX3 statistic</h1>
    
    <div class="hidden name"><code>snp_MAX3.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>Compute the MAX3 statistic, which tests for three genetic models
(additive, recessive and dominant).</p>
    
    </div>

    <pre class="usage"><span class='fu'>snp_MAX3</span>(<span class='no'>Gna</span>, <span class='no'>y01.train</span>, <span class='kw'>ind.train</span> <span class='kw'>=</span> <span class='fu'>rows_along</span>(<span class='no'>Gna</span>), <span class='kw'>val</span> <span class='kw'>=</span> <span class='fu'>c</span>(<span class='fl'>0</span>, <span class='fl'>0.5</span>, <span class='fl'>1</span>))</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>Gna</th>
      <td><p>A FBM.code256
(typically <code>&lt;bigSNP&gt;$genotypes</code>).<br />
You can have missing values in your data.</p></td>
    </tr>
    <tr>
      <th>y01.train</th>
      <td><p>Vector of responses, corresponding to <code>ind.train</code>.
<strong>Must be only 0s and 1s.</strong></p></td>
    </tr>
    <tr>
      <th>ind.train</th>
      <td><p>An optional vector of the row indices that are used,
for the training part. If not specified, all rows are used.
<strong>Don't use negative indices.</strong></p></td>
    </tr>
    <tr>
      <th>val</th>
      <td><p>Computing \(\smash{\displaystyle\max_{x \in val}}~Z_{CATT}^2(x)\).</p><ul>
<li><p>Default is <code>c(0, 0.5, 1)</code> and corresponds to the <em>MAX3</em> statistic.</p></li>
<li><p>Only <code>c(0, 1)</code> corresponds to <em>MAX2</em>.</p></li>
<li><p>And only <code>0.5</code> corresponds to the Armitage trend test.</p></li>
<li><p>Finally, <code>seq(0, 1, length.out = L)</code> corresponds to <em>MAXL</em>.</p></li>
</ul></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>An object of classes <code>mhtest</code> and <code>data.frame</code> returning one
score by SNP. See <code>methods(class = "mhtest")</code>.</p>
    
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p><strong>P-values associated with returned scores are in fact the minimum of the
p-values of each test separately. Thus, they are biased downward.</strong></p>
    
    <h2 class="hasAnchor" id="references"><a class="anchor" href="#references"></a>References</h2>

    <p>Zheng, G., Yang, Y., Zhu, X., &amp; Elston, R. (2012).
Robust Procedures. Analysis Of Genetic Association Studies, 151-206.
<a href='http://dx.doi.org/10.1007/978-1-4614-2245-7_6'>http://dx.doi.org/10.1007/978-1-4614-2245-7_6</a>.</p>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='fu'>set.seed</span>(<span class='fl'>1</span>)

<span class='co'># constructing a fake genotype big.matrix</span>
<span class='no'>N</span> <span class='kw'>&lt;-</span> <span class='fl'>50</span>; <span class='no'>M</span> <span class='kw'>&lt;-</span> <span class='fl'>1200</span>
<span class='no'>fake</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='snp_fake.html'>snp_fake</a></span>(<span class='no'>N</span>, <span class='no'>M</span>)
<span class='no'>G</span> <span class='kw'>&lt;-</span> <span class='no'>fake</span>$<span class='no'>genotypes</span>
<span class='no'>G</span>[] <span class='kw'>&lt;-</span> <span class='fu'>sample</span>(<span class='fu'>as.raw</span>(<span class='fl'>0</span>:<span class='fl'>3</span>), <span class='kw'>size</span> <span class='kw'>=</span> <span class='fu'>length</span>(<span class='no'>G</span>), <span class='kw'>replace</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>)
<span class='no'>G</span>[<span class='fl'>1</span>:<span class='fl'>8</span>, <span class='fl'>1</span>:<span class='fl'>10</span>]</div><div class='output co'>#&gt;      [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8] [,9] [,10]
#&gt; [1,]    1    0    2    0    0    2    0    0    0     2
#&gt; [2,]    2    0    0    0    2    0   NA    0    0     0
#&gt; [3,]   NA   NA    0    0   NA    0    0   NA    0     0
#&gt; [4,]    2   NA   NA    1    1    2    2    0   NA     1
#&gt; [5,]    2    1   NA    0    1   NA    0    2   NA    NA
#&gt; [6,]   NA    0   NA    2   NA    0    1   NA    0     2
#&gt; [7,]    2    1    0    2    1    2    2    1    1     0
#&gt; [8,]   NA    2    2    0    0    0    2   NA    1     2</div><div class='input'>
<span class='co'># Specify case/control phenotypes</span>
<span class='no'>fake</span>$<span class='no'>fam</span>$<span class='no'>affection</span> <span class='kw'>&lt;-</span> <span class='fu'>rep</span>(<span class='fl'>1</span>:<span class='fl'>2</span>, <span class='kw'>each</span> <span class='kw'>=</span> <span class='no'>N</span> / <span class='fl'>2</span>)

<span class='co'># Get MAX3 statistics</span>
<span class='no'>y01</span> <span class='kw'>&lt;-</span> <span class='no'>fake</span>$<span class='no'>fam</span>$<span class='no'>affection</span> - <span class='fl'>1</span>
<span class='fu'>str</span>(<span class='no'>test</span> <span class='kw'>&lt;-</span> <span class='fu'>snp_MAX3</span>(<span class='no'>fake</span>$<span class='no'>genotypes</span>, <span class='kw'>y01.train</span> <span class='kw'>=</span> <span class='no'>y01</span>))</div><div class='output co'>#&gt; Classes ‘mhtest’ and 'data.frame':	1200 obs. of  1 variable:
#&gt;  $ score: num  3.976 1.591 0.551 2.22 0.201 ...
#&gt;  - attr(*, "transfo")=function (x)  
#&gt;  - attr(*, "predict")=function (xtr)  
#&gt;   ..- attr(*, "srcref")=Class 'srcref'  atomic [1:8] 98 15 100 3 15 3 98 100
#&gt;   .. .. ..- attr(*, "srcfile")=Classes 'srcfilecopy', 'srcfile' &lt;environment: 0xd89b510&gt; </div><div class='input'><span class='co'># p-values are not well calibrated</span>
<span class='fu'><a href='snp_qq.html'>snp_qq</a></span>(<span class='no'>test</span>)</div><div class='img'><img src='snp_MAX3-1.png' alt='' width='700' height='433' /></div><div class='input'><span class='co'># genomic control is not of much help</span>
<span class='fu'><a href='snp_qq.html'>snp_qq</a></span>(<span class='fu'><a href='snp_gc.html'>snp_gc</a></span>(<span class='no'>test</span>))</div><div class='img'><img src='snp_MAX3-2.png' alt='' width='700' height='433' /></div><div class='input'>
<span class='co'># Armitage trend test (well calibrated because only one test)</span>
<span class='no'>test2</span> <span class='kw'>&lt;-</span> <span class='fu'>snp_MAX3</span>(<span class='no'>fake</span>$<span class='no'>genotypes</span>, <span class='kw'>y01.train</span> <span class='kw'>=</span> <span class='no'>y01</span>, <span class='kw'>val</span> <span class='kw'>=</span> <span class='fl'>0.5</span>)
<span class='fu'><a href='snp_qq.html'>snp_qq</a></span>(<span class='no'>test2</span>)</div><div class='img'><img src='snp_MAX3-3.png' alt='' width='700' height='433' /></div><div class='input'>
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#details">Details</a></li>

      <li><a href="#references">References</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Florian Privé.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  

  </body>
</html>

