<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>PRS — snp_PRS • bigsnpr</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://maxcdn.bootstrapcdn.com/bootswatch/3.3.7/readable/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">

<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">


<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script>
<script src="../pkgdown.js"></script>

<!-- mathjax -->
<script src='https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">bigsnpr</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../reference/index.html">Manual</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/pruning-vs-clumping.html">Why clumping should be preferred over pruning</a>
    </li>
    <li>
      <a href="../articles/bcm-seminar-packages.html">Two example analyses</a>
    </li>
  </ul>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://privefl.github.io/about.html">
    <span class="fa fa-user"></span>
     
    About
  </a>
</li>
<li>
  <a href="https://github.com/privefl/bigsnpr">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

      <div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>PRS</h1>
    </div>

    
    <p>Polygenic Risk Scores with possible clumping and thresholding.</p>
    

    <pre><span class='fu'>snp_PRS</span>(<span class='no'>G</span>, <span class='no'>betas</span>, <span class='no'>ind.test</span>, <span class='kw'>ind.keep</span> <span class='kw'>=</span> <span class='fu'>cols_along</span>(<span class='no'>G</span>), <span class='kw'>lpS</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>thr.list</span> <span class='kw'>=</span> <span class='fl'>0</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a> Arguments</h2>
    <dl class="dl-horizontal">
      <dt>G</dt>
      <dd>A BM.code.descriptor
(typically <code>&lt;bigSNP&gt;$genotypes</code>).
<strong>You shouldn&#39;t have missing values in your data or SNPs with 0 MAF.</strong></dd>
      <dt>betas</dt>
      <dd>Numeric vector of weights associated with each SNP.
You may want to see big_univLinReg or big_univLogReg.</dd>
      <dt>ind.test</dt>
      <dd>The individuals on whom to project the scores.</dd>
      <dt>ind.keep</dt>
      <dd>Column (SNP) indices to use (if using clumping, the
output of <a href='pruning-clumping.html'>snp_clumping</a>). Default doesn&#39;t clump.</dd>
      <dt>lpS</dt>
      <dd>Numeric vector of <code>-log10(p.value)</code> associated with <code>betas</code>.
Default doesn&#39;t use thresholding.</dd>
      <dt>thr.list</dt>
      <dd>Threshold vector on <code>lpS</code> at which SNPs are excluded if
they are not significant enough. Default doesn&#39;t use thresholding.</dd>
    </dl>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A matrix of scores, where rows correspond to <code>ind.test</code> and
columns correspond to <code>thr.list</code>.</p>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='no'>test</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='snp_attachExtdata.html'>snp_attachExtdata</a></span>()
<span class='no'>G</span> <span class='kw'>&lt;-</span> <span class='no'>test</span>$<span class='no'>genotypes</span>
<span class='no'>n</span> <span class='kw'>&lt;-</span> <span class='fu'>nrow</span>(<span class='no'>G</span>)
<span class='no'>m</span> <span class='kw'>&lt;-</span> <span class='fu'>ncol</span>(<span class='no'>G</span>)

<span class='co'># get some phenotypes (LTM)</span>
<span class='fu'>set.seed</span>(<span class='fl'>5</span>)
<span class='no'>X</span> <span class='kw'>&lt;-</span> <span class='fu'>attach.BM</span>(<span class='no'>G</span>)
<span class='no'>M</span> <span class='kw'>&lt;-</span> <span class='fl'>100</span> <span class='co'># number of causal variants</span>
<span class='no'>K</span> <span class='kw'>&lt;-</span> <span class='fl'>0.3</span> <span class='co'># prevalence</span>
<span class='no'>ind.causal</span> <span class='kw'>&lt;-</span> <span class='fu'>sample</span>(<span class='no'>m</span>, <span class='no'>M</span>)
<span class='no'>betas</span> <span class='kw'>&lt;-</span> <span class='fu'>rnorm</span>(<span class='no'>M</span>, <span class='kw'>sd</span> <span class='kw'>=</span> <span class='fu'>sqrt</span>(<span class='fl'>0.8</span> / <span class='no'>M</span>))
<span class='no'>y.g</span> <span class='kw'>&lt;-</span> <span class='fu'>drop</span>(<span class='fu'>scale</span>(<span class='no'>X</span>[, <span class='no'>ind.causal</span>]) <span class='kw'>%*%</span> <span class='no'>betas</span>)
<span class='fu'>rm</span>(<span class='no'>X</span>)
(<span class='no'>v</span> <span class='kw'>&lt;-</span> <span class='fu'>var</span>(<span class='no'>y.g</span>))</div><div class='output co'>#&gt; [1] 0.7538336</div><div class='input'><span class='no'>y.e</span> <span class='kw'>&lt;-</span> <span class='fu'>rnorm</span>(<span class='no'>n</span>, <span class='kw'>sd</span> <span class='kw'>=</span> <span class='fu'>sqrt</span>(<span class='fl'>1</span> - <span class='no'>v</span>))
(<span class='no'>y01</span> <span class='kw'>&lt;-</span> <span class='fu'>as.numeric</span>((<span class='no'>y.g</span> + <span class='no'>y.e</span>) <span class='kw'>&gt;</span> <span class='fu'>qnorm</span>(<span class='fl'>1</span> - <span class='no'>K</span>)))</div><div class='output co'>#&gt;   [1] 1 1 0 0 0 1 0 1 0 0 0 1 1 1 0 1 0 0 0 0 0 0 0 0 0 1 1 0 0 0 1 1 0 0 0 0 1
#&gt;  [38] 0 1 0 0 0 0 0 1 1 0 0 1 0 1 1 1 0 0 0 1 1 0 1 0 1 0 0 1 1 0 0 1 0 0 0 1 0
#&gt;  [75] 0 1 0 0 0 0 1 0 0 0 0 0 1 0 0 1 1 0 0 0 0 0 1 0 0 1 0 1 1 1 0 0 0 0 0 0 0
#&gt; [112] 1 1 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 1 0 1 0 0 1 0 0 1 0 0 0 1 1 1
#&gt; [149] 0 1 0 0 1 1 0 0 1 0 0 0 1 0 0 0 1 0 1 0 0 1 0 1 0 1 0 1 1 1 0 0 0 0 0 0 0
#&gt; [186] 0 0 0 1 1 0 0 0 0 0 0 0 0 0 0 0 0 1 1 0 1 1 1 1 0 0 1 0 0 0 1 0 0 1 1 0 0
#&gt; [223] 0 0 1 0 0 0 0 1 1 0 1 1 1 1 1 1 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 1 1 1 0
#&gt; [260] 0 1 1 1 1 1 1 0 0 0 0 0 0 0 0 1 0 0 0 0 1 0 0 0 1 0 1 1 1 1 0 1 0 0 1 1 0
#&gt; [297] 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 0 0 0 1 0 0 0 1 0 0 0
#&gt; [334] 1 0 0 0 0 0 0 0 0 0 0 1 0 0 1 0 0 0 0 1 1 0 0 1 0 0 1 0 0 1 0 0 0 0 0 0 1
#&gt; [371] 0 0 1 0 1 1 0 1 0 1 1 1 0 0 0 1 0 0 0 1 0 1 0 0 0 0 0 0 0 0 0 0 1 1 0 0 1
#&gt; [408] 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
#&gt; [445] 0 0 0 0 0 0 1 0 1 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 1 1 1 0 1 1 0 0 1 0
#&gt; [482] 0 1 1 1 0 1 0 0 0 0 1 1 0 0 0 0 0 0 1 0 0 1 0 0 0 0 1 0 1 1 0 1 0 0 0 0</div><div class='input'>
<span class='co'># PCA -&gt; covariables</span>
<span class='no'>obj.svd</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='snp_clumpedSVD.html'>snp_clumpedSVD</a></span>(<span class='no'>G</span>, <span class='kw'>infos.chr</span> <span class='kw'>=</span> <span class='no'>test</span>$<span class='no'>map</span>$<span class='no'>chromosome</span>)</div><div class='output co'>#&gt; First phase of clumping at r2 &gt; 0.2.. keep 4270 SNPs.
#&gt; 
#&gt; Iteration 1:
#&gt; Computing SVD..
#&gt; 
#&gt; Converged!</div><div class='input'>
<span class='co'># train and test set</span>
<span class='no'>ind.train</span> <span class='kw'>&lt;-</span> <span class='fu'>sort</span>(<span class='fu'>sample</span>(<span class='no'>n</span>, <span class='fl'>400</span>))
<span class='no'>ind.test</span> <span class='kw'>&lt;-</span> <span class='fu'>setdiff</span>(<span class='fu'>rows_along</span>(<span class='no'>G</span>), <span class='no'>ind.train</span>) <span class='co'># 117</span>

<span class='co'># GWAS</span>
<span class='no'>gwas.train</span> <span class='kw'>&lt;-</span> <span class='fu'>big_univLogReg</span>(<span class='no'>G</span>, <span class='kw'>y01.train</span> <span class='kw'>=</span> <span class='no'>y01</span>[<span class='no'>ind.train</span>],
                             <span class='kw'>ind.train</span> <span class='kw'>=</span> <span class='no'>ind.train</span>,
                             <span class='kw'>covar.train</span> <span class='kw'>=</span> <span class='no'>obj.svd</span>$<span class='no'>u</span>[<span class='no'>ind.train</span>, ])
<span class='co'># clumping</span>
<span class='no'>ind.keep</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='pruning-clumping.html'>snp_clumping</a></span>(<span class='no'>G</span>, <span class='kw'>infos.chr</span> <span class='kw'>=</span> <span class='no'>test</span>$<span class='no'>map</span>$<span class='no'>chromosome</span>,
                         <span class='kw'>ind.row</span> <span class='kw'>=</span> <span class='no'>ind.train</span>,
                         <span class='kw'>S</span> <span class='kw'>=</span> <span class='fu'>abs</span>(<span class='no'>gwas.train</span>$<span class='no'>score</span>))
<span class='co'># -log10(p-values) and thresolding</span>
<span class='fu'>summary</span>(<span class='no'>lpS</span> <span class='kw'>&lt;-</span> -<span class='fu'>predict</span>(<span class='no'>gwas.train</span>))</div><div class='output co'>#&gt;     Min.  1st Qu.   Median     Mean  3rd Qu.     Max. 
#&gt; 0.000019 0.129900 0.317300 0.456500 0.641400 3.584000 </div><div class='input'><span class='no'>thrs</span> <span class='kw'>&lt;-</span> <span class='fu'>seq</span>(<span class='fl'>0</span>, <span class='fl'>3.5</span>, <span class='kw'>by</span> <span class='kw'>=</span> <span class='fl'>0.5</span>)
<span class='no'>nb.pred</span> <span class='kw'>&lt;-</span> <span class='fu'>sapply</span>(<span class='no'>thrs</span>, <span class='kw'>function</span>(<span class='no'>thr</span>) <span class='fu'>sum</span>(<span class='no'>lpS</span>[<span class='no'>ind.keep</span>] <span class='kw'>&gt;</span> <span class='no'>thr</span>))

<span class='co'># PRS</span>
<span class='no'>prs</span> <span class='kw'>&lt;-</span> <span class='fu'>snp_PRS</span>(<span class='no'>G</span>, <span class='kw'>betas</span> <span class='kw'>=</span> <span class='no'>gwas.train</span>$<span class='no'>estim</span>,
               <span class='kw'>ind.test</span> <span class='kw'>=</span> <span class='no'>ind.test</span>,
               <span class='kw'>ind.keep</span> <span class='kw'>=</span> <span class='no'>ind.keep</span>,
               <span class='kw'>lpS</span> <span class='kw'>=</span> <span class='no'>lpS</span>,
               <span class='kw'>thr.list</span> <span class='kw'>=</span> <span class='no'>thrs</span>)

<span class='co'># AUC as a function of the number of predictors</span>
<span class='no'>aucs</span> <span class='kw'>&lt;-</span> <span class='fu'>apply</span>(<span class='no'>prs</span>, <span class='fl'>2</span>, <span class='no'>AUC</span>, <span class='kw'>target</span> <span class='kw'>=</span> <span class='no'>y01</span>[<span class='no'>ind.test</span>])
<span class='fu'>library</span>(<span class='no'>ggplot2</span>)
<span class='kw'>bigstatsr</span><span class='kw'>:::</span><span class='fu'>MY_THEME</span>(<span class='fu'>qplot</span>(<span class='no'>nb.pred</span>, <span class='no'>aucs</span>)) +
  <span class='fu'>geom_line</span>() +
  <span class='fu'>scale_x_log10</span>(<span class='kw'>breaks</span> <span class='kw'>=</span> <span class='no'>nb.pred</span>) +
  <span class='fu'>labs</span>(<span class='kw'>x</span> <span class='kw'>=</span> <span class='st'>"Number of predictors"</span>, <span class='kw'>y</span> <span class='kw'>=</span> <span class='st'>"AUC"</span>)</div><img src='snp_PRS-10.png' alt='' width='540' height='400' /></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Florian Privé.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  </body>
</html>
