<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Subset — subset.bigSNP • bigsnpr</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script>
<script src="../pkgdown.js"></script>
<meta property="og:title" content="Subset — subset.bigSNP" />

<meta property="og:description" content="A method to get a subset (copy) of a bigSNP.
This new bigSNP will also be backed by files (in the same directory)." />
<meta name="twitter:card" content="summary" />


<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->




  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">bigsnpr</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">0.3.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../reference/index.html">Manual</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/demo.html">Quick demo</a>
    </li>
    <li>
      <a href="../articles/how-to-PCA.html">How to capture Population Structure with PCA</a>
    </li>
    <li>
      <a href="../articles/pruning-vs-clumping.html">Why clumping should be preferred over pruning</a>
    </li>
    <li>
      <a href="../articles/dosage.html">Dosage data: proof of concept</a>
    </li>
  </ul>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://privefl.github.io/about.html">
    <span class="fa fa-user"></span>
     
    About
  </a>
</li>
<li>
  <a href="https://github.com/privefl/bigsnpr">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Subset</h1>
    
    <div class="hidden name"><code>subset.bigSNP.Rd</code></div>
    </div>

    
    <p>A method to get a subset (copy) of a <code>bigSNP</code>.
This new <code>bigSNP</code> will also be backed by files (in the same directory).</p>
    

    <pre class="usage"><span class='co'># S3 method for bigSNP</span>
<span class='fu'>subset</span>(<span class='no'>x</span>, <span class='kw'>ind.row</span> <span class='kw'>=</span> <span class='fu'>rows_along</span>(<span class='no'>G</span>), <span class='kw'>ind.col</span> <span class='kw'>=</span> <span class='fu'>cols_along</span>(<span class='no'>G</span>),
  <span class='no'>...</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>x</th>
      <td><p>A <a href='bigSNP-class.html'>bigSNP</a>.</p></td>
    </tr>
    <tr>
      <th>ind.row</th>
      <td><p>Indices of the rows (individuals) to keep.
Negative indices <strong>can</strong> be used to exclude row indices.
Default: keep them all.</p></td>
    </tr>
    <tr>
      <th>ind.col</th>
      <td><p>Indices of the columns (SNPs) to keep.
Negative indices <strong>can</strong> be used to exclude column indices.
Default: keep them all.</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>Not used.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>The path to the RDS file that stores the <code>bigSNP</code> object.<br />
Note that this function creates two files whose names have been
automatically chosen by appending "_sub" and a number to the prefix of the
input bigSNP backing files.</p>
    
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><a href='bigSNP-class.html'>bigSNP</a></p></div>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='fu'>str</span>(<span class='no'>test</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='snp_attachExtdata.html'>snp_attachExtdata</a></span>())</div><div class='output co'>#&gt; List of 3
#&gt;  $ genotypes:Reference class 'FBM.code256' [package "bigstatsr"] with 7 fields
#&gt;   ..$ extptr     :&lt;externalptr&gt; 
#&gt;   ..$ nrow       : int 517
#&gt;   ..$ ncol       : int 4542
#&gt;   ..$ type       : Named int 1
#&gt;   .. ..- attr(*, "names")= chr "unsigned char"
#&gt;   ..$ backingfile: chr "/tmp/Rtmp5GaOe4/file4c675b3d745.bk"
#&gt;   ..$ address    :&lt;externalptr&gt; 
#&gt;   ..$ code256    : num [1:256] 0 1 2 NA NA NA NA NA NA NA ...
#&gt;   ..and 21 methods, of which 7 are  possibly relevant:
#&gt;   ..  as.FBM, copy#envRefClass, initialize, initialize#FBM, save,
#&gt;   ..  show#envRefClass, show#FBM
#&gt;  $ fam      :'data.frame':	517 obs. of  6 variables:
#&gt;   ..$ family.ID  : chr [1:517] "POP1" "POP1" "POP1" "POP1" ...
#&gt;   ..$ sample.ID  : chr [1:517] "IND0" "IND1" "IND2" "IND3" ...
#&gt;   ..$ paternal.ID: int [1:517] 0 0 0 0 0 0 0 0 0 0 ...
#&gt;   ..$ maternal.ID: int [1:517] 0 0 0 0 0 0 0 0 0 0 ...
#&gt;   ..$ sex        : int [1:517] 0 0 0 0 0 0 0 0 0 0 ...
#&gt;   ..$ affection  : int [1:517] 1 1 2 1 1 1 1 1 1 1 ...
#&gt;  $ map      :'data.frame':	4542 obs. of  6 variables:
#&gt;   ..$ chromosome  : int [1:4542] 1 1 1 1 1 1 1 1 1 1 ...
#&gt;   ..$ marker.ID   : chr [1:4542] "SNP0" "SNP1" "SNP2" "SNP3" ...
#&gt;   ..$ genetic.dist: int [1:4542] 0 0 0 0 0 0 0 0 0 0 ...
#&gt;   ..$ physical.pos: int [1:4542] 112 1098 2089 3107 4091 5091 6107 7103 8090 9074 ...
#&gt;   ..$ allele1     : chr [1:4542] "A" "T" "T" "T" ...
#&gt;   ..$ allele2     : chr [1:4542] "T" "A" "A" "A" ...
#&gt;  - attr(*, "class")= chr "bigSNP"</div><div class='input'>
<span class='co'># keep only first 50 samples and SNPs</span>
<span class='no'>rdsfile</span> <span class='kw'>&lt;-</span> <span class='fu'>subset</span>(<span class='no'>test</span>, <span class='kw'>ind.row</span> <span class='kw'>=</span> <span class='fl'>1</span>:<span class='fl'>50</span>, <span class='kw'>ind.col</span> <span class='kw'>=</span> <span class='fl'>1</span>:<span class='fl'>50</span>)
<span class='fu'>str</span>(<span class='fu'><a href='snp_attach.html'>snp_attach</a></span>(<span class='no'>rdsfile</span>))</div><div class='output co'>#&gt; List of 3
#&gt;  $ genotypes:Reference class 'FBM.code256' [package "bigstatsr"] with 7 fields
#&gt;   ..$ extptr     :&lt;externalptr&gt; 
#&gt;   ..$ nrow       : int 50
#&gt;   ..$ ncol       : int 50
#&gt;   ..$ type       : Named int 1
#&gt;   .. ..- attr(*, "names")= chr "unsigned char"
#&gt;   ..$ backingfile: chr "/tmp/Rtmp5GaOe4/file4c675b3d745_sub1.bk"
#&gt;   ..$ address    :&lt;externalptr&gt; 
#&gt;   ..$ code256    : num [1:256] 0 1 2 NA NA NA NA NA NA NA ...
#&gt;   ..and 21 methods, of which 7 are  possibly relevant:
#&gt;   ..  as.FBM, copy#envRefClass, initialize, initialize#FBM, save,
#&gt;   ..  show#envRefClass, show#FBM
#&gt;  $ fam      :'data.frame':	50 obs. of  6 variables:
#&gt;   ..$ family.ID  : chr [1:50] "POP1" "POP1" "POP1" "POP1" ...
#&gt;   ..$ sample.ID  : chr [1:50] "IND0" "IND1" "IND2" "IND3" ...
#&gt;   ..$ paternal.ID: int [1:50] 0 0 0 0 0 0 0 0 0 0 ...
#&gt;   ..$ maternal.ID: int [1:50] 0 0 0 0 0 0 0 0 0 0 ...
#&gt;   ..$ sex        : int [1:50] 0 0 0 0 0 0 0 0 0 0 ...
#&gt;   ..$ affection  : int [1:50] 1 1 2 1 1 1 1 1 1 1 ...
#&gt;  $ map      :'data.frame':	50 obs. of  6 variables:
#&gt;   ..$ chromosome  : int [1:50] 1 1 1 1 1 1 1 1 1 1 ...
#&gt;   ..$ marker.ID   : chr [1:50] "SNP0" "SNP1" "SNP2" "SNP3" ...
#&gt;   ..$ genetic.dist: int [1:50] 0 0 0 0 0 0 0 0 0 0 ...
#&gt;   ..$ physical.pos: int [1:50] 112 1098 2089 3107 4091 5091 6107 7103 8090 9074 ...
#&gt;   ..$ allele1     : chr [1:50] "A" "T" "T" "T" ...
#&gt;   ..$ allele2     : chr [1:50] "T" "A" "A" "A" ...
#&gt;  - attr(*, "class")= chr "bigSNP"</div><div class='input'>
<span class='co'># remove only first 50 samples and SNPs</span>
<span class='no'>rdsfile2</span> <span class='kw'>&lt;-</span> <span class='fu'>subset</span>(<span class='no'>test</span>, <span class='kw'>ind.row</span> <span class='kw'>=</span> -(<span class='fl'>1</span>:<span class='fl'>50</span>), <span class='kw'>ind.col</span> <span class='kw'>=</span> -(<span class='fl'>1</span>:<span class='fl'>50</span>))
<span class='fu'>str</span>(<span class='fu'><a href='snp_attach.html'>snp_attach</a></span>(<span class='no'>rdsfile2</span>))</div><div class='output co'>#&gt; List of 3
#&gt;  $ genotypes:Reference class 'FBM.code256' [package "bigstatsr"] with 7 fields
#&gt;   ..$ extptr     :&lt;externalptr&gt; 
#&gt;   ..$ nrow       : int 467
#&gt;   ..$ ncol       : int 4492
#&gt;   ..$ type       : Named int 1
#&gt;   .. ..- attr(*, "names")= chr "unsigned char"
#&gt;   ..$ backingfile: chr "/tmp/Rtmp5GaOe4/file4c675b3d745_sub2.bk"
#&gt;   ..$ address    :&lt;externalptr&gt; 
#&gt;   ..$ code256    : num [1:256] 0 1 2 NA NA NA NA NA NA NA ...
#&gt;   ..and 21 methods, of which 7 are  possibly relevant:
#&gt;   ..  as.FBM, copy#envRefClass, initialize, initialize#FBM, save,
#&gt;   ..  show#envRefClass, show#FBM
#&gt;  $ fam      :'data.frame':	467 obs. of  6 variables:
#&gt;   ..$ family.ID  : chr [1:467] "POP2" "POP2" "POP2" "POP2" ...
#&gt;   ..$ sample.ID  : chr [1:467] "IND50" "IND51" "IND52" "IND53" ...
#&gt;   ..$ paternal.ID: int [1:467] 0 0 0 0 0 0 0 0 0 0 ...
#&gt;   ..$ maternal.ID: int [1:467] 0 0 0 0 0 0 0 0 0 0 ...
#&gt;   ..$ sex        : int [1:467] 0 0 0 0 0 0 0 0 0 0 ...
#&gt;   ..$ affection  : int [1:467] 1 1 2 1 1 1 1 1 1 2 ...
#&gt;  $ map      :'data.frame':	4492 obs. of  6 variables:
#&gt;   ..$ chromosome  : int [1:4492] 1 1 1 1 1 1 1 1 1 1 ...
#&gt;   ..$ marker.ID   : chr [1:4492] "SNP50" "SNP51" "SNP52" "SNP53" ...
#&gt;   ..$ genetic.dist: int [1:4492] 0 0 0 0 0 0 0 0 0 0 ...
#&gt;   ..$ physical.pos: int [1:4492] 50100 51109 52101 53106 54101 55097 56103 57110 58101 59088 ...
#&gt;   ..$ allele1     : chr [1:4492] "A" "A" "A" "T" ...
#&gt;   ..$ allele2     : chr [1:4492] "T" "T" "T" "A" ...
#&gt;  - attr(*, "class")= chr "bigSNP"</div><div class='input'>
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#see-also">See also</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Florian Privé.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  

  </body>
</html>

