<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Subset — subset.bigSNP • bigsnpr</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://maxcdn.bootstrapcdn.com/bootswatch/3.3.7/readable/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">

<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">


<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script>
<script src="../pkgdown.js"></script>

<!-- mathjax -->
<script src='https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">bigsnpr</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../reference/index.html">Manual</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/pruning-vs-clumping.html">Why clumping should be preferred over pruning</a>
    </li>
    <li>
      <a href="../articles/demo.html">Small demo</a>
    </li>
  </ul>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://privefl.github.io/about.html">
    <span class="fa fa-user"></span>
     
    About
  </a>
</li>
<li>
  <a href="https://github.com/privefl/bigsnpr">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

      <div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Subset</h1>
    </div>

    
    <p>A method to get a subset (copy) of a <code>bigSNP</code>.
This new <code>bigSNP</code> will also be backed by files (in the same directory).</p>
    

    <pre><span class='co'># S3 method for bigSNP</span>
<span class='fu'>subset</span>(<span class='no'>x</span>, <span class='kw'>ind.row</span> <span class='kw'>=</span> <span class='fu'>rows_along</span>(<span class='no'>X</span>), <span class='kw'>ind.col</span> <span class='kw'>=</span> <span class='fu'>cols_along</span>(<span class='no'>X</span>),
  <span class='no'>...</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a> Arguments</h2>
    <dl class="dl-horizontal">
      <dt>x</dt>
      <dd>A <a href='bigSNP-class.html'>bigSNP</a>.</dd>
      <dt>ind.row</dt>
      <dd>Indices of the rows (individuals) to keep.
Negative indices <strong>can</strong> be used to exclude row indices.
Default: keep them all.</dd>
      <dt>ind.col</dt>
      <dd>Indices of the columns (SNPs) to keep.
Negative indices <strong>can</strong> be used to exclude column indices.
Default: keep them all.</dd>
      <dt>...</dt>
      <dd>Not used.</dd>
    </dl>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>The path to the RDS file that stores the <code>bigSNP</code> object.
Note that this function creates two files whose names have been
automatically chosen by appending &quot;_sub&quot; and a number to the prefix of the
input bigSNP backing files.</p>
    
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <p><a href='bigSNP-class.html'>bigSNP</a></p>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='fu'>str</span>(<span class='no'>test</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='snp_attachExtdata.html'>snp_attachExtdata</a></span>())</div><div class='output co'>#&gt; List of 4
#&gt;  $ genotypes:Formal class &#39;BM.code.descriptor&#39; [package &quot;bigstatsr&quot;] with 2 slots
#&gt;   .. ..@ code       : num [1:256] 0 1 2 NA NA NA NA NA NA NA ...
#&gt;   .. ..@ description:List of 13
#&gt;   .. .. ..$ sharedType: chr &quot;FileBacked&quot;
#&gt;   .. .. ..$ filename  : chr &quot;file1171350d7d2e.bk&quot;
#&gt;   .. .. ..$ dirname   : chr &quot;/tmp/RtmpQaJJOw/&quot;
#&gt;   .. .. ..$ totalRows : num 517
#&gt;   .. .. ..$ totalCols : num 4542
#&gt;   .. .. ..$ rowOffset : num 0
#&gt;   .. .. ..$ colOffset : num 0
#&gt;   .. .. ..$ nrow      : num 517
#&gt;   .. .. ..$ ncol      : num 4542
#&gt;   .. .. ..$ rowNames  : NULL
#&gt;   .. .. ..$ colNames  : NULL
#&gt;   .. .. ..$ type      : chr &quot;raw&quot;
#&gt;   .. .. ..$ separated : logi FALSE
#&gt;  $ fam      :&#39;data.frame&#39;:	517 obs. of  6 variables:
#&gt;   ..$ family.ID  : chr [1:517] &quot;POP1&quot; &quot;POP1&quot; &quot;POP1&quot; &quot;POP1&quot; ...
#&gt;   ..$ sample.ID  : chr [1:517] &quot;IND0&quot; &quot;IND1&quot; &quot;IND2&quot; &quot;IND3&quot; ...
#&gt;   ..$ paternal.ID: int [1:517] 0 0 0 0 0 0 0 0 0 0 ...
#&gt;   ..$ maternal.ID: int [1:517] 0 0 0 0 0 0 0 0 0 0 ...
#&gt;   ..$ sex        : int [1:517] 0 0 0 0 0 0 0 0 0 0 ...
#&gt;   ..$ affection  : int [1:517] 1 1 2 1 1 1 1 1 1 1 ...
#&gt;  $ map      :&#39;data.frame&#39;:	4542 obs. of  6 variables:
#&gt;   ..$ chromosome  : int [1:4542] 1 1 1 1 1 1 1 1 1 1 ...
#&gt;   ..$ marker.ID   : chr [1:4542] &quot;SNP0&quot; &quot;SNP1&quot; &quot;SNP2&quot; &quot;SNP3&quot; ...
#&gt;   ..$ genetic.dist: int [1:4542] 0 0 0 0 0 0 0 0 0 0 ...
#&gt;   ..$ physical.pos: int [1:4542] 112 1098 2089 3107 4091 5091 6107 7103 8090 9074 ...
#&gt;   ..$ allele1     : chr [1:4542] &quot;A&quot; &quot;T&quot; &quot;T&quot; &quot;T&quot; ...
#&gt;   ..$ allele2     : chr [1:4542] &quot;T&quot; &quot;A&quot; &quot;A&quot; &quot;A&quot; ...
#&gt;  $ savedIn  : chr &quot;/tmp/RtmpQaJJOw/file1171350d7d2e.rds&quot;
#&gt;  - attr(*, &quot;class&quot;)= chr &quot;bigSNP&quot;</div><div class='input'>
<span class='co'># keep only first 50 samples and SNPs</span>
<span class='no'>rdsfile</span> <span class='kw'>&lt;-</span> <span class='fu'>subset</span>(<span class='no'>test</span>, <span class='kw'>ind.row</span> <span class='kw'>=</span> <span class='fl'>1</span>:<span class='fl'>50</span>, <span class='kw'>ind.col</span> <span class='kw'>=</span> <span class='fl'>1</span>:<span class='fl'>50</span>)
<span class='fu'>str</span>(<span class='fu'><a href='snp_attach.html'>snp_attach</a></span>(<span class='no'>rdsfile</span>))</div><div class='output co'>#&gt; List of 4
#&gt;  $ genotypes:Formal class &#39;BM.code.descriptor&#39; [package &quot;bigstatsr&quot;] with 2 slots
#&gt;   .. ..@ code       : num [1:256] 0 1 2 NA NA NA NA NA NA NA ...
#&gt;   .. ..@ description:List of 13
#&gt;   .. .. ..$ sharedType: chr &quot;FileBacked&quot;
#&gt;   .. .. ..$ filename  : chr &quot;file1171350d7d2e_sub1.bk&quot;
#&gt;   .. .. ..$ dirname   : chr &quot;/tmp/RtmpQaJJOw/&quot;
#&gt;   .. .. ..$ totalRows : num 50
#&gt;   .. .. ..$ totalCols : num 50
#&gt;   .. .. ..$ rowOffset : num 0
#&gt;   .. .. ..$ colOffset : num 0
#&gt;   .. .. ..$ nrow      : num 50
#&gt;   .. .. ..$ ncol      : num 50
#&gt;   .. .. ..$ rowNames  : NULL
#&gt;   .. .. ..$ colNames  : NULL
#&gt;   .. .. ..$ type      : chr &quot;raw&quot;
#&gt;   .. .. ..$ separated : logi FALSE
#&gt;  $ fam      :&#39;data.frame&#39;:	50 obs. of  6 variables:
#&gt;   ..$ family.ID  : chr [1:50] &quot;POP1&quot; &quot;POP1&quot; &quot;POP1&quot; &quot;POP1&quot; ...
#&gt;   ..$ sample.ID  : chr [1:50] &quot;IND0&quot; &quot;IND1&quot; &quot;IND2&quot; &quot;IND3&quot; ...
#&gt;   ..$ paternal.ID: int [1:50] 0 0 0 0 0 0 0 0 0 0 ...
#&gt;   ..$ maternal.ID: int [1:50] 0 0 0 0 0 0 0 0 0 0 ...
#&gt;   ..$ sex        : int [1:50] 0 0 0 0 0 0 0 0 0 0 ...
#&gt;   ..$ affection  : int [1:50] 1 1 2 1 1 1 1 1 1 1 ...
#&gt;  $ map      :&#39;data.frame&#39;:	50 obs. of  6 variables:
#&gt;   ..$ chromosome  : int [1:50] 1 1 1 1 1 1 1 1 1 1 ...
#&gt;   ..$ marker.ID   : chr [1:50] &quot;SNP0&quot; &quot;SNP1&quot; &quot;SNP2&quot; &quot;SNP3&quot; ...
#&gt;   ..$ genetic.dist: int [1:50] 0 0 0 0 0 0 0 0 0 0 ...
#&gt;   ..$ physical.pos: int [1:50] 112 1098 2089 3107 4091 5091 6107 7103 8090 9074 ...
#&gt;   ..$ allele1     : chr [1:50] &quot;A&quot; &quot;T&quot; &quot;T&quot; &quot;T&quot; ...
#&gt;   ..$ allele2     : chr [1:50] &quot;T&quot; &quot;A&quot; &quot;A&quot; &quot;A&quot; ...
#&gt;  $ savedIn  : chr &quot;/tmp/RtmpQaJJOw/file1171350d7d2e_sub1.rds&quot;
#&gt;  - attr(*, &quot;class&quot;)= chr &quot;bigSNP&quot;</div><div class='input'>
<span class='co'># remove only first 50 samples and SNPs</span>
<span class='no'>rdsfile2</span> <span class='kw'>&lt;-</span> <span class='fu'>subset</span>(<span class='no'>test</span>, <span class='kw'>ind.row</span> <span class='kw'>=</span> -(<span class='fl'>1</span>:<span class='fl'>50</span>), <span class='kw'>ind.col</span> <span class='kw'>=</span> -(<span class='fl'>1</span>:<span class='fl'>50</span>))
<span class='fu'>str</span>(<span class='fu'><a href='snp_attach.html'>snp_attach</a></span>(<span class='no'>rdsfile2</span>))</div><div class='output co'>#&gt; List of 4
#&gt;  $ genotypes:Formal class &#39;BM.code.descriptor&#39; [package &quot;bigstatsr&quot;] with 2 slots
#&gt;   .. ..@ code       : num [1:256] 0 1 2 NA NA NA NA NA NA NA ...
#&gt;   .. ..@ description:List of 13
#&gt;   .. .. ..$ sharedType: chr &quot;FileBacked&quot;
#&gt;   .. .. ..$ filename  : chr &quot;file1171350d7d2e_sub2.bk&quot;
#&gt;   .. .. ..$ dirname   : chr &quot;/tmp/RtmpQaJJOw/&quot;
#&gt;   .. .. ..$ totalRows : num 467
#&gt;   .. .. ..$ totalCols : num 4492
#&gt;   .. .. ..$ rowOffset : num 0
#&gt;   .. .. ..$ colOffset : num 0
#&gt;   .. .. ..$ nrow      : num 467
#&gt;   .. .. ..$ ncol      : num 4492
#&gt;   .. .. ..$ rowNames  : NULL
#&gt;   .. .. ..$ colNames  : NULL
#&gt;   .. .. ..$ type      : chr &quot;raw&quot;
#&gt;   .. .. ..$ separated : logi FALSE
#&gt;  $ fam      :&#39;data.frame&#39;:	467 obs. of  6 variables:
#&gt;   ..$ family.ID  : chr [1:467] &quot;POP2&quot; &quot;POP2&quot; &quot;POP2&quot; &quot;POP2&quot; ...
#&gt;   ..$ sample.ID  : chr [1:467] &quot;IND50&quot; &quot;IND51&quot; &quot;IND52&quot; &quot;IND53&quot; ...
#&gt;   ..$ paternal.ID: int [1:467] 0 0 0 0 0 0 0 0 0 0 ...
#&gt;   ..$ maternal.ID: int [1:467] 0 0 0 0 0 0 0 0 0 0 ...
#&gt;   ..$ sex        : int [1:467] 0 0 0 0 0 0 0 0 0 0 ...
#&gt;   ..$ affection  : int [1:467] 1 1 2 1 1 1 1 1 1 2 ...
#&gt;  $ map      :&#39;data.frame&#39;:	4492 obs. of  6 variables:
#&gt;   ..$ chromosome  : int [1:4492] 1 1 1 1 1 1 1 1 1 1 ...
#&gt;   ..$ marker.ID   : chr [1:4492] &quot;SNP50&quot; &quot;SNP51&quot; &quot;SNP52&quot; &quot;SNP53&quot; ...
#&gt;   ..$ genetic.dist: int [1:4492] 0 0 0 0 0 0 0 0 0 0 ...
#&gt;   ..$ physical.pos: int [1:4492] 50100 51109 52101 53106 54101 55097 56103 57110 58101 59088 ...
#&gt;   ..$ allele1     : chr [1:4492] &quot;A&quot; &quot;A&quot; &quot;A&quot; &quot;T&quot; ...
#&gt;   ..$ allele2     : chr [1:4492] &quot;T&quot; &quot;T&quot; &quot;T&quot; &quot;A&quot; ...
#&gt;  $ savedIn  : chr &quot;/tmp/RtmpQaJJOw/file1171350d7d2e_sub2.rds&quot;
#&gt;  - attr(*, &quot;class&quot;)= chr &quot;bigSNP&quot;</div><div class='input'>
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#see-also">See also</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Florian Privé.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  </body>
</html>
